---
title: 'Microbiome Preprocessing'
output:
  html_document
---


# BACTERIA
## Import Data

```{r}
drive_path <- "/Users/giac0002/Library/CloudStorage/GoogleDrive-giulia.iacono1@monash.edu/Shared drives/Lung Tx SOPs/Biomarker 2/Microbiome/Dada2/cluster_output/"

## Load bacteria dada generated with DADA2-pipeline and taxonomy_species.rds for 16S taxonomy classifications
bact.seqtab <- t(data.frame(readRDS("16S/20220727/seqtab_nochim.rds")))
bact.taxonomy <- data.frame(readRDS("16S/20220727/taxonomy_species.rds"))
bact.fitGTR <- readRDS('16S/20220727/tree.rds')

# Remove _7 from Prevotella genus column
bact.taxonomy$Genus <- gsub("_7", "", bact.taxonomy$Genus)

## Identify each ASV by highest level of classification followed by the first letter of taxa level, then an identifying number, e.g. Muribaculaceae|f-1
bact.taxonomy <- add_column(bact.taxonomy, "ASV" = highest_ClassID(bact.taxonomy), .after = 0)

## Check matching ASV names and table dimensions 
dim(bact.seqtab) #6708  647
dim(bact.taxonomy) #6708    8

# Import tab-delimited metadata file
seq.metadata <- read_csv('Microbiome/Data/seq-metadata.csv', col_types = 'ccncccncncllllccccllllccc') %>% 
  dplyr::select(!c("Date", "Number", "Plate", "Sample", "PCR_box", "In_box", "DNA_extraction_rerun", "Seq_to_bis", "QCpassed", "Sample finished", "DNA_box/comment")) %>%
  filter(!is.na(Batch), Sequenced == TRUE, grepl("16S.", SampleID)) %>% 
  mutate(Patient_days = paste0(Record.ID, "_", Days, "_", Lobe),
         Patient_days = gsub("_NA", "", Patient_days)) %>%
  dplyr::select(!c("PCR_contam", "Sequenced", "Sample_excluded")) %>%
  as.data.frame()
rownames(seq.metadata) <- seq.metadata$SampleID
dim(seq.metadata) #581 12 less because not sequenced were excluded

# Match bacteria metadata to sequencing data
bact.seq.metadata <- seq.metadata[rownames(seq.metadata) %in% colnames(bact.seqtab), ]

# Match sequencing data to bacteria metadata
bact.seqtab <- bact.seqtab[ , colnames(bact.seqtab) %in% rownames(bact.seq.metadata)]

dim(bact.seq.metadata) #581  12
dim(bact.seqtab) #6708    581

# Order sample_ID by numerical order and check that the order is the same in all tables
bact.seqtab <- bact.seqtab[,sort(colnames(bact.seqtab))]
bact.seq.metadata <- bact.seq.metadata[sort(rownames(bact.seq.metadata)),]

identical(rownames(bact.seqtab), rownames(bact.taxonomy)) # all tables are in the same order
identical(colnames(bact.seqtab), rownames(bact.seq.metadata))
```

## Create Phyloseq object

```{r}
# Convert Bacteria and Fungi data to a Phyloseq object
bact.raw <- phyloseq(otu_table(bact.seqtab, taxa_are_rows = TRUE), 
                 sample_data(bact.seq.metadata), 
                 tax_table(as.matrix(bact.taxonomy)),
                 phy_tree(bact.fitGTR$tree))

identical(rownames(bact.raw@otu_table), rownames(bact.seqtab)) #TRUE
taxa_names(bact.raw) <- data.frame(tax_table(bact.raw))$ASV 

rm(bact.fitGTR)
rm(bact.seqtab)
rm(bact.taxonomy)

# Get raw counts
write.csv(data.frame('Reads' = colSums(otu_table(bact.raw))), 
          file = paste0("Microbiome/Data/16S/", Sys.Date(), "-bact-raw-read-counts.txt"), row.names = TRUE)

saveRDS(bact.raw, paste0("Microbiome/Data/16S/bact.raw.rds"))
bact.raw <- readRDS("Microbiome/Data/16S/bact.raw.rds")

# write.csv(data.frame(tax_table(bact.raw)) %>% filter(ASV == "Prevotellaceae|f-42"), "pre.csv")
```

## Preprocessing

```{r}
## Minimum read count
# Removes samples, all PCR negative controls and some negative extraction controls
bact.min.filt <- prune_samples(colSums(otu_table(bact.raw)) > 200, bact.raw)
# Before removal
dim(sample_data(bact.raw))[1] #581 | 318
# After removal
dim(sample_data(bact.min.filt))[1] #553 | 293

## Unclassified ASV removal
bact.tax.filt <- prune_taxa(!is.na(data.frame(tax_table(bact.min.filt))$Phylum), bact.min.filt)
dim(tax_table(bact.min.filt))[1] #6708
# After removal
dim(tax_table(bact.tax.filt))[1] #6612

## Positive controls removal
bact.pos.ctrl <- data.frame(sample_data(bact.tax.filt)) %>%
  filter(Type == "ExPosCtrl") %>%
  pull(SampleID) # 10

bact.pos.filt <- prune_samples(data.frame(sample_data(bact.tax.filt)) %>%
  filter(Type != "ExPosCtrl") %>%
  pull(SampleID), bact.tax.filt)
# After removal
dim(sample_data(bact.pos.filt))[1] #543 | 283
```

## Check negative controls

```{r}
bact.ctrls <- data.frame(sample_data(bact.raw)) %>%
  filter(grepl("Pos", Type) | Days < 90) %>%
  pull(SampleID) # 57

b <- prune_samples(bact.ctrls, bact.raw) %>%
  #microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt()

b$colour <- ifelse(b$OTU == "Enterococcus faecium|s-210", "red", "black")
b$colour <- ifelse(b$OTU == "Staphylococcus|g-16", "blue", b$colour)

unique(b$colour)

unique(b$OTU)[grepl("Staph", unique(b$OTU))]

b %>% filter(OTU == "Staphylococcus|g-16", Abundance > 0)
b %>% filter(OTU == "Enterococcus faecium|s-210", Abundance > 0)

b <- filter(b, )


ggplot(b, aes(x=Abundance, y=SampleID)) + 
  geom_bar(aes(fill = colour), stat = 'identity', size = 0.01) + 
  labs(x = "Relative abundance", y = " ", title = "Bacterial negative controls - taxonomy composition") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 6),
        strip.text = element_text(size = 8),
        legend.text = element_text(size = 8)) +
  guides(fill = guide_legend(ncol = 1)) +
  facet_wrap(Type~.) +
  scale_fill_manual(values = c(`red` = "red", `black` = "black", blue = "blue"))

```


## Check controls

```{r}
zymo.composition <- c("Bacillus", "Enterococcus", "Escherichia", "Lactobacillus", "Listeria", "Pseudomonas", "Salmonella", "Staphylococcus")

bact.ctrls <- data.frame(sample_data(bact.tax.filt)) %>%
  filter(Type != "Sample") %>%
  pull(SampleID) # 56

bact.ctrls.filt <- prune_samples(bact.ctrls, bact.tax.filt) %>%
  core(detection = 10, prevalence = 0.05) %>% 
  tax_glom(taxrank = 'Genus', NArm = T) %>%
  microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt()

bact.ctrls.filt %>% 
  #filter(Type == "ExPosCtrl") %>%

ggplot(aes(x=Abundance, y=SampleID)) + 
  geom_bar(aes(fill = Genus), stat = 'identity', size = 0.01) + 
  labs(x = "Relative abundance", y = " ", title = "Bacterial controls - taxonomy composition") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 6),
        strip.text = element_text(size = 8),
        legend.text = element_text(size = 8)) +
  guides(fill = guide_legend(ncol = 1)) +
  facet_grid(Type ~ ., scales = "free", space = "free")
```

### Check Cellulosimicrobium

```{r}
cell <- bact.raw %>%
  psmelt() %>%
  filter(ASV == "Cellulosimicrobium|g-1") 
cell$Status <- ifelse(cell$SampleID %in% c(bact.dup.df %>% filter(Type == "Sample") %>% pull(SampleID)), "Pass", "Fail")
  
cell %>%
  mutate(Days_post_transplant = cut(Days, breaks = c(-1, 42, 183, 365, 2000),
                          labels = c("<42", "42-183", "184-365", ">365")),
         Days_post_transplant2 = case_when(Status == "Pass" & !is.na(Days_post_transplant) ~ as.character(Days_post_transplant),
                                          Status == "Fail"  ~ "NA")) %>%
  filter(!SampleID %in% (cell %>% filter(Type == "Sample" & Status == "Fail") %>% pull(SampleID))) %>%

  ggplot(aes(x = Type, y = Abundance)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(colour = Days_post_transplant2, shape = Status), width = 0.1) +
  labs(y = "Log raw abundance", title = "Cellulosimicrobium|g-1") +
  scale_y_log10() +
  scale_colour_viridis(direction = -1, discrete = TRUE) 

ggplot(cell, aes(x = Days, y = Abundance)) +
  geom_point() +
  labs(y = "Log raw abundance", title = "Cellulosimicrobium|g-1") +
  scale_y_log10() +
  geom_smooth(se = FALSE)
```

### Total read counts

```{r}
counts <- data.frame(colSums(otu_table(bact))) %>%
  set_colnames("Count") %>%
  rownames_to_column("Patient_days_lobe") %>%
  left_join(data.frame(sample_data(bact)), by = "Patient_days_lobe") 

ggplot(counts, aes(x = reorder(Patient_days_lobe, Count), y = Count)) +
  geom_col(aes(fill = Days_post_transplant)) +
  theme +
  plot_fill(Days_post_transplant = TRUE) +
  scale_y_log10()

ggplot(counts, aes(x = Days_post_transplant, y = Count)) +
  geom_boxplot(aes(fill = Days_post_transplant)) +
  geom_point() +
  theme +
  plot_fill(Days_post_transplant = TRUE) +
  scale_y_log10()
```

## Decontam

```{r}
## Prevalence method
sample_data(bact.pos.filt)$is.neg <- grepl(pattern = 'Neg', sample_data(bact.pos.filt)$Type)
contamdf.prev <- isContaminant(bact.pos.filt, method="prevalence", neg="is.neg", batch = sample_data(bact.pos.filt)$MiseqID, threshold = 0.5) 
 
hist(contamdf.prev$p, 100)
table(contamdf.prev$contaminant) #53
rownames(contamdf.prev)[which(contamdf.prev$contaminant)]
```

```{r}
# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(bact.pos.filt, function(abund) 1*(abund>0))

ps.pa.neg <- prune_samples(grepl(pattern = 'Neg', sample_data(bact.pos.filt)$Type), ps.pa) 
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Type == "Sample", ps.pa)

# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf.prev$contaminant)
p <- 
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + 
  geom_point(size = 1, alpha = 0.8) +
  #theme(legend.position = "right", legend.justification = c("left")) +
  #geom_text_repel(label = ifelse(df.pa$contaminant == TRUE, rownames(df.pa), ""), max.overlaps = 6, cex=1.5) + 
  labs(title = "Bacteria decontam",
       colour = "Contaminant") +
  xlab("Prevalence (Negative Controls)") + 
  ylab("Prevalence (True Samples)") +
  scale_colour_manual(values = c("TRUE" = "#E60031", "FALSE" = "grey")) +
  scale_y_continuous(limits = c(0, 500)) +
  scale_x_continuous(limits = c(0, 25))

ggsave(paste0("Microbiome/Data/16S/contaminant/bact-decontam-legend.png"), p, width = 12, height = 8, units = 'cm')

rm(ps.pa)
rm(ps.pa.neg)
rm(ps.pa.pos)
rm(df.pa)
```

```{r}
# Remove contaminants from phyloseq
bact.decon <- prune_taxa(!contamdf.prev$contaminant, bact.pos.filt)
#bact.decon <- readRDS("Microbiome/Data/16S/bact.decon.rds")
```

## Check dataset post contamination filtering

```{r}
bact.decon.df <- data.frame("Reads" = colSums(otu_table(bact.decon)),
                             "Type" = sample_data(bact.decon)$Type)
bact.decon.df %>% 
  filter(Reads < 2000) %>%
ggplot(aes(x = rownames(.), y = Reads)) +
  geom_col(aes(fill = Type)) +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 3, angle = 90)) 
```

## Tip glom 

```{r}
## Agglomerate ASVs at the phylogenetic 
## Cluster
#c <- c(0.015, 0.02, 0.05, 0.1, 0.2)
#for (i in 1:length(c)){
#  h <- c[i]
#  bact.glom[[as.character(h)]] <- tip_glom(bact.decon, h = h) # smaller h equals to less cutting
#}
#saveRDS(bact.glom, "bact.glom.rds")
bact.glom <- readRDS("Microbiome/Data/16S/bact.glom.rds")

bact.decon # 6559
length(rownames(tax_table(bact.glom$`0.02`)) ) # 2008

plot_tree(bact.decon, "treeonly", title="Before tip_glom()")
plot_tree(bact.glom$`0.02`, "treeonly",  title="After tip_glom()")
plot_tree(bact.glom$`0.1`, "treeonly",  title="After tip_glom()")
plot_tree(bact.glom$`0.2`, "treeonly",  title="After tip_glom()")

# Add taxas back
tax_table(bact.glom$`0.02`) <- tax_table(bact.decon)[rownames(tax_table(bact.glom$`0.02`)),]
```

## Prevalence filter and duplicate removal

```{r}
# Prevalence filtering
bact.prev.filtered <- core(bact.glom$`0.02`, detection = 1, prevalence = 0, include.lowest = FALSE)

# Remove duplicate samples and keep the ones with the highest reads + remove R healthy
bact.summary <- data.frame('Reads' = colSums(otu_table(bact.prev.filtered)), 
                           'Patient_days' = sample_data(bact.prev.filtered)$Patient_days,
                           "Type" = sample_data(bact.prev.filtered)$Type)

bact.keep <- bact.summary %>%
  mutate(SampleID = rownames(bact.summary)) 
duplicates <- bact.keep %>% group_by(Patient_days) %>% filter(n() > 1) %>% group_by(Patient_days) %>% top_n(-1, Reads) %>% pull(SampleID)
bact.keep <-  bact.keep[!(bact.keep$SampleID %in% duplicates),] %>% pull(SampleID)
 
# Filtered phyloseq
bact.dup.filt <- prune_samples(bact.keep, bact.prev.filtered) 
```


## Check after prevalence filtering

```{r}
other.controls <- data.frame("SampleID" = sample_data(bact.raw)$SampleID,
                             "Reads" = colSums(otu_table(bact.raw)),
                             "Type" = sample_data(bact.raw)$Type) %>%
  filter(Reads <= 200, grepl("Neg", Type))

bact.dup.df <- data.frame("SampleID" = sample_data(bact.dup.filt)$SampleID,
                           "Reads" = colSums(otu_table(bact.dup.filt)),
                           "Type" = sample_data(bact.dup.filt)$Type) %>%
  full_join(other.controls)

bact.dup.df$Type <- ifelse(bact.dup.df$Type == "Sample" & bact.dup.df$Reads < 2000, "Excluded", 
                              ifelse(bact.dup.df$Type == "Sample" & bact.dup.df$Reads >= 2000, "Sample", bact.dup.df$Type))

p <- 
  
  bact.dup.df %>%
  mutate(Type = factor(Type, levels = c("BronchNegCtrl", "ExNegCtrl", "PCRNegCtrl", "Excluded", "Sample"))) %>%
  
ggplot(aes(x = Type, y = Reads)) +
 # geom_violin(aes(fill = Type)) +
  geom_boxplot(fill = "grey", outlier.shape = NA, alpha = 0.8) +
  geom_jitter(width = 0.2) +
  #geom_dotplot(binaxis = 'y', stackdir = 'center', dotsize = 0.5, width = 0.0002) +
  labs(title = "Bacteria read count filter",
       y = "Log ASV Reads",
       x  = " ") +
  geom_hline(aes(yintercept = 2000), linetype = "dashed", color = "grey") +
  scale_y_log10() +
  theme(legend.position = "none") 

ggsave(paste0("Microbiome/Data/16S/contaminant/bact-read-filter.png"), p, width = 12, height = 12, units = 'cm')
```


```{r}
bact.dup.df %>% group_by(Type) %>% count() %>% filter(Type %in% c("Sample", "Excluded")) %>%
  mutate(Percentage = round(100*n/sum(.$n))) %>%
  ungroup() %>%
  add_row(Type = "Total", n = sum(.$n), Percentage = NA)
```


## Composition comparison

```{r}
b.raw <- bact.raw %>%
core(detection = 10, prevalence = 0.05) %>% 
  tax_glom(taxrank = 'Genus', NArm = T) %>%
  microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt()

b.fil <- bact.prev.filtered %>%
core(detection = 10, prevalence = 0.05) %>% 
  tax_glom(taxrank = 'Genus', NArm = T) %>%
  microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt()
```

## Read threshold

```{r}
# Read threshold filtering
threshold <- 2000
bact.filt <- prune_samples(colSums(otu_table(bact.dup.filt)) >= threshold, bact.dup.filt)

data.frame('Summary' = c('ASVs/genera', 'Samples'),
           'Bacteria_before_ASV_filtering' = dim(otu_table(bact.raw)), # 581 6708
           'Bacteria_after_contam_filtering' = dim(otu_table(bact.decon)), # 543 6559 
           'Bacteria_after_prev/thresh_filtering' = dim(otu_table(bact.filt))) # 445 1992
```


## Save filtered library size

```{r}
# Estimate Shannon index and retrieve the number of ASVs (observed) 
sample_data(bact.filt)$Diversity <- estimate_richness(bact.filt, split = TRUE, measures = c('Shannon'))$Shannon
sample_names(bact.filt) <- sample_data(bact.filt)$Patient_days
saveRDS(bact.filt, "Microbiome/Data/16S/bact.filt.rds")
# bact.filt <- readRDS("Microbiome/Data/16S/bact.filt.rds")

bact_seq_metadata <- data.frame(sample_data(bact.filt), row.names = NULL) %>%
  dplyr::select(Record.ID, Days, Patient_days, Lobe, Diversity) 
saveRDS(bact_seq_metadata, "Microbiome/Data/16S/bact_seq_metadata.rds")
# Now check patient selection before next chunk

rm(bact.decon, bact.dup.filt, bact.dup.df, bact.glom, bact.min.filt, bact.pos.ctrl, bact.pos.filt, bact.prev.filtered, bact.summary, bact.raw, bact.tax.filt, contamdf.prev, other.controls)
```


# Create complete phyloseq object

```{r}
# Select columns to be matched to metadata tables
b <- fev_at_bronch(bact_seq_metadata, 
                      spirometry_metadata,
                      unique(bact_seq_metadata$Record.ID)) # need record.id, days and patient days

bact_metadata <- bact_seq_metadata %>%
  left_join(master_metadata$bronch_metadata[, c("Record.ID", "Patient_days", "Days")], by = c("Record.ID", "Days"), suffix = c("_lobe", "")) %>%
  left_join(master_metadata$meds_metadata, by = c("Patient_days")) %>%
  left_join(master_metadata$general_metadata, by = "Record.ID") %>% 
  mutate(Patient_days = gsub("_Left|_Right", "", Patient_days_lobe)) %>%
  
  left_join(b, by = c("Patient_days")) %>%
  left_join(master_metadata$donor, by = "Record.ID") %>%
  left_join(master_metadata$bal_class_hosp) %>%
  left_join(clad_assessment[,c("Record.ID", "Transplanted")], by = "Record.ID") %>%
  mutate(Lobe = case_when(Lobe == Transplanted ~ "transplanted",
                          Lobe != Transplanted ~ "native",
                          is.na(Lobe) ~ "transplanted")) 
bact_metadata <- annotate_clad(bact_metadata)
bact_metadata <- classify_dosage(bact_metadata)

bact_metadata <- bact_metadata %>%
  mutate(Time_interval_detailed = case_when(Days <= 42 ~ "<1.5 months",
                                         Days > 42 & Days <= 92 ~ "1.5-3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_detailed = factor(Time_interval_detailed, c("<1.5 months", "1.5-3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Time_interval_combined = case_when(Days <= 92 ~ "<3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_combined = factor(Time_interval_combined, c("<3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Days_interval = case_when(Days <= 183 ~ "<6 months",
                                   Days > 183 ~ ">6 months"), 
         Days_interval = factor(Days_interval, c("<6 months", ">6 months")),
         
         Days_interval2 = case_when(Days <= 92 ~ "<3",
                                   Days > 92 ~ ">3"), 
         Days_interval2 = factor(Days_interval2, c("<3", ">3")),
         
         Immunosuppression = case_when(Days <= 183 ~ "High",
                                         Days > 183 & Days <= 365 ~ "Mid",
                                         Days > 365 ~ "Low"), 
         Immunosuppression = factor(Immunosuppression, c("High", "Mid", "Low")),
         
         Months_grouped = case_when(Days <= 42 ~ "<1.5",
                                    Days > 42 & Days <= 92 ~ "1.5-3",
                                    Days > 92 & Days <= 183 ~ "4-6",
                                    Days > 183 & Days <= 273 ~ "7-9",
                                    Days > 273 & Days <= 365 ~ "10-12",
                                    Days > 365 & Days <= 547 ~ "13-18",
                                    Days > 547 ~ ">18"), 
         Months_grouped = factor(Months_grouped, c("<1.5", "1.5-3", "4-6", "7-9", "10-12", "13-18", ">18")),
         
         Months = round(Days/30.417, 1))

rownames(bact_metadata) <- bact_metadata$Patient_days_lobe

# Add new metadata to phyloseq. Rows need to match ps row order
bact_metadata <- bact_metadata[sample_names(bact.filt),]
identical(rownames(bact_metadata), sample_names(bact.filt)) #TRUE
identical(colnames(otu_table(bact.filt)), rownames(bact_metadata)) #TRUE
bact <- phyloseq::phyloseq(otu_table(bact.filt), tax_table(bact.filt), sample_data(bact_metadata), phy_tree(bact.filt))

# Subset phyloseq object for study
bact <- phyloseq::subset_samples(bact, Record.ID %in% unique(c(stable_patients, biomarker_patients)) ) # patients
bact <- phyloseq::subset_samples(bact, Patient_days != "44_182") # remove sex mismatched sample
bact <- phyloseq::subset_samples(bact, Lobe == "transplanted" & Months < 26)
bact <- phyloseq::subset_samples(bact, is.na(days_to_clad) | days_to_clad >= -30) # only tx lobes, and samples less than 30M

sample_data(bact)$Total_ant_cat_split <- ifelse(sample_data(bact)$Total_ant_num == 0, "0",
                                          ifelse(sample_data(bact)$Total_ant_num > 0 & sample_data(bact)$Total_ant_num <= 2, "1-2", "3-5"))
sample_data(bact)$Total_ant_cat_split <- factor(sample_data(bact)$Total_ant_cat_split, levels = c("0", "1-2", "3-5"))
#sample_data(bact)$GIT_ischemic_time_hours <- round(as.numeric(sample_data(bact)$GIT_ischemic_time/60), 1)

#sample_data(bact)$Ischemic_time_cat <- ifelse(sample_data(bact)$GIT_ischemic_time <= 300, "<300",
#                                          ifelse(sample_data(bact)$GIT_ischemic_time > 300 & sample_data(bact)$GIT_ischemic_time <= 400, "300-400", ">400"))
#sample_data(bact)$Ischemic_time_cat <- factor(sample_data(bact)$Ischemic_time_cat, levels = c("<300", "300-400", ">400"))
sample_names(bact) <- sample_data(bact)$Patient_days

# Checks
length(unique(c(stable_patients, biomarker_patients)) ) # 69 unique patients
length(unique(sample_data(bact)$Record.ID)) # 69

length(all_selected_samples_collected_all_datasets %>% filter(Dataset == "Bacteria") %>% pull(Patient_days) ) #365
length(data.frame(sample_data(bact))$Patient_days ) # 365

setdiff(all_selected_samples_collected_all_datasets %>% filter(Dataset == "Bacteria") %>% pull(Patient_days), data.frame(sample_data(bact))$Patient_days ) 

bact_metadata <- data.frame(sample_data(bact))
colnames(bact_metadata) <- ifelse(grepl("Piperacillin.Tazobactam", colnames(bact_metadata)), "Piperacillin-Tazobactam", colnames(bact_metadata)  )
colnames(bact_metadata) <- ifelse(grepl("Mycophenolate.Mofetil", colnames(bact_metadata)), "Mycophenolate-Mofetil", colnames(bact_metadata)  )

saveRDS(bact, "Microbiome/Data/16S/bact.rds")
saveRDS(bact_metadata, "Microbiome/Data/16S/bact_metadata.rds")

rm(b, bact.filt)

# for AI
b <- bact
otu_table(b) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(b), p = 0.5)), taxa_are_rows = TRUE)
b <- microbiome::transform(b, transform = 'log') 

write.csv(otu_table(b), "bact_otu_table.csv")
write.csv(tax_table(b), "bact_tax_table.csv")

write.csv(bact_metadata %>% dplyr::select(!c("ParticipantInitials", "Gender", "Age", "transplant_date", "Donor_gender", "Donor_age")), "bact_metadata.csv")
```

```{r}
# summarize_phyloseq(bact.filt)

# Number of Genera and Phyla
data.frame("genus" = data.frame(tax_table(bact))$Genus) %>% group_by(genus) %>% dplyr::count() %>% arrange(desc(n))
head(names(base::sort(desc(table(data.frame(tax_table(bact))$Genus)))))
data.frame("phylum" = data.frame(tax_table(bact))$Phylum) %>% group_by(phylum) %>% dplyr::count() %>% arrange(desc(n))

# Number of ASVs and samples
dim(otu_table(bact)) 
# 1992 365

view(tax_table(bact))
```


# ------------------------------

# FUNGI
# Import data

```{r}
# Check setups
concat_only <- data.frame(read.delim("~/Desktop/R/Biomarker2/Dada2/cluster_output/ITS/20220727_concat_only/total_counts.txt", sep = "\t"))
pe_concat <- data.frame(read.delim("~/Desktop/R/Biomarker2/Dada2/cluster_output/ITS/20220727_pe_concat/total_counts.txt", sep = "\t"))

concat.df <- concat_only %>%
  inner_join(pe_concat, by = "SampleID", suffix = c("_concat_only", "_pe_concat")) %>%
  filter(abs(Total.count_concat_only - Total.count_pe_concat) > 10)

concat.df[which(concat.df$Total.count_concat_only > concat.df$Total.count_pe_concat),] #23 higher in concat only
table(concat.df$Total.count_pe_concat > concat.df$Total.count_concat_only) #92 higher in pe concat
```


```{r}
## Import
fungi.seqtab <- t(data.frame(readRDS("~/Desktop/R/Biomarker2/Dada2/cluster_output/ITS/20220727_concat_only/seqtab_nochim.rds")))
fungi.taxonomy <- data.frame(readRDS("~/Desktop/R/Biomarker2/Dada2/cluster_output/ITS/20220727_concat_only/taxonomy.rds"))
fungi.fitGTR <- readRDS('~/Desktop/R/Biomarker2/Dada2/cluster_output/ITS/20220727_concat_only/tree.rds')

## Reshape taxonomy to be comparable to bacteria
fungi.taxonomy[] <- lapply(fungi.taxonomy, gsub, pattern='[a-z]__', replacement='')
fungi.taxonomy <- add_column(fungi.taxonomy, "ASV" = highest_ClassID(fungi.taxonomy), .after = 0)

## Check matching ASV names and table dimensions 
dim(fungi.seqtab) #1901  480
dim(fungi.taxonomy) #1901    8

# Import tab-delimited metadata file
seq.metadata <- read_csv('Data/seq-metadata.csv', col_types = 'ccncccncncllllccccllllccc') %>% 
  dplyr::select(!c("Date", "Number", "Plate", "Sample", "PCR_box", "In_box", "DNA_extraction_rerun", "Seq_to_bis", "QCpassed", "Sample finished", "DNA_box/comment")) %>%
  filter(!is.na(Batch), Sequenced == TRUE, grepl("ITS.", SampleID)) %>% 
  mutate(Patient_days = paste0(Record.ID, "_", Days, "_", Lobe),
         Patient_days = gsub("_NA", "", Patient_days)) %>%
  dplyr::select(!c("PCR_contam", "Sequenced", "Sample_excluded")) %>%
  as.data.frame()
rownames(seq.metadata) <- seq.metadata$SampleID
dim(seq.metadata) #359 12 less because not sequenced were excluded

# Match fungi metadata to sequencing data
fungi.seq.metadata <- seq.metadata[rownames(seq.metadata) %in% colnames(fungi.seqtab), ]

# Match sequencing data to fungi metadata
fungi.seqtab <- fungi.seqtab[ ,colnames(fungi.seqtab) %in% rownames(fungi.seq.metadata)]

# Order sample_ID by numerical order and check that the order is the same in all tables
fungi.seqtab <- fungi.seqtab[,sort(colnames(fungi.seqtab))]
fungi.seq.metadata <- fungi.seq.metadata[sort(rownames(fungi.seq.metadata)),]

identical(rownames(fungi.seqtab), rownames(fungi.taxonomy)) #TRUE
identical(colnames(fungi.seqtab), rownames(fungi.seq.metadata)) #TRUE
```


## Create Phyloseq object

```{r}
# Convert Bacteria and Fungi data to a Phyloseq object
fungi.raw <- phyloseq(otu_table(fungi.seqtab, taxa_are_rows = TRUE), 
                 sample_data(fungi.seq.metadata), 
                 tax_table(as.matrix(fungi.taxonomy)),
                 phy_tree(fungi.fitGTR$tree))

# Check order
identical(rownames(otu_table(fungi.raw)), rownames(tax_table(fungi.raw))) #TRUE
taxa_names(fungi.raw) <- data.frame(tax_table(fungi.raw))$ASV 

rm(fungi.seqtab)
rm(fungi.taxonomy)
rm(fungi.fitGTR)

# Get raw counts
write.csv(data.frame('Reads' = colSums(otu_table(fungi.raw))), 
          file = paste0("Data/ITS/", Sys.Date(), "-fungi-raw-read-counts.csv"), row.names = TRUE)
saveRDS(fungi.raw, "Microbiome/Data/ITS/fungi.raw.rds")
fungi.raw <- readRDS("Microbiome/Data/ITS/fungi.raw.rds")
```

## Preprocessing

```{r}
## Minimum read count
# Removes samples, all PCR negative controls and some negative extraction controls
fungi.min.filt <- prune_samples(colSums(otu_table(fungi.raw)) > 200, fungi.raw)
# Before removal
dim(sample_data(fungi.raw))[1] #357
# After removal
dim(sample_data(fungi.min.filt))[1] #320 

## Unclassified ASV removal
fungi.tax.filt <- prune_taxa(!is.na(data.frame(tax_table(fungi.min.filt))$Phylum), fungi.min.filt)
dim(tax_table(fungi.min.filt))[1] #1901
# After removal
dim(tax_table(fungi.tax.filt))[1] #1724

## Positive controls removal
fungi.pos.ctrl <- data.frame(sample_data(fungi.tax.filt)) %>%
  filter(Type == "ExPosCtrl") %>%
  pull(SampleID) # 10

fungi.pos.filt <- prune_samples(data.frame(sample_data(fungi.tax.filt)) %>%
  filter(Type != "ExPosCtrl") %>%
  pull(SampleID), fungi.tax.filt)
# After removal
dim(sample_data(fungi.pos.filt))[1] #310 

rm(fungi.min.filt)
rm(fungi.tax.filt)
```

## Check controls

```{r}
zymo.composition <- c("Cryptococcus", "Saccharomyces")

fungi.ctrls <- data.frame(sample_data(fungi.tax.filt)) %>%
  filter(Type != "Sample") %>%
  pull(SampleID) # 56

fungi.ctrls.filt <- prune_samples(fungi.ctrls, fungi.tax.filt) %>%
  core(detection = 10, prevalence = 0.05) %>% 
  tax_glom(taxrank = 'Species', NArm = F) %>%
  #microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt() 
  #filter(Type == "ExNegCtrl") %>%

ggplot(fungi.ctrls.filt, aes(x=Abundance, y=SampleID)) + 
  geom_bar(aes(fill = Genus), stat = 'identity', size = 0.01) + 
  labs(x = "Relative abundance", y = " ", title = "Fungal controls - taxonomy composition") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 6),
        strip.text = element_text(size = 8),
        legend.text = element_text(size = 8)) +
  guides(fill = guide_legend(ncol = 1)) +
  facet_grid(Type ~ ., scales = "free", space = "free")
```

### Candida check

```{r}
data.frame(otu_table(fungi.raw)) %>%
  dplyr::select(data.frame(sample_data(fungi.ctrls.filt)) %>%
  filter(Type == "ExPosCtrl") %>% pull(SampleID))

cand <- fungi.raw %>%
  psmelt() %>%
  filter(ASV == "Saccharomyces|g-5") 
cand$Status <- ifelse(cand$SampleID %in% c(fungi.dup.df %>% filter(Type == "Sample") %>% pull(SampleID)), "Pass", "Fail")
  
cand %>%
  mutate(Days_post_transplant = cut(Days, breaks = c(-1, 42, 183, 365, 2000),
                          labels = c("<42", "42-183", "184-365", ">365")),
         Days_post_transplant2 = case_when(Status == "Pass" & !is.na(Days_post_transplant) ~ as.character(Days_post_transplant),
                                          Status == "Fail"  ~ "NA")) %>%
  filter(!SampleID %in% (cand %>% filter(Type == "Sample" & Status == "Fail") %>% pull(SampleID))) %>%

  ggplot(aes(x = Type, y = Abundance)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(colour = Days_post_transplant2, shape = Status), width = 0.1) +
  labs(y = "Log raw abundance", title = "Candida albicans|s-1") +
  scale_y_log10() +
  scale_colour_viridis(direction = -1, discrete = TRUE) 
```

### Total read counts

```{r}
counts <- data.frame(colSums(otu_table(fungi))) %>%
  set_colnames("Count") %>%
  rownames_to_column("Patient_days_lobe") %>%
  left_join(data.frame(sample_data(fungi)), by = "Patient_days_lobe") 

ggplot(counts, aes(x = reorder(Patient_days_lobe, Count), y = Count)) +
  geom_col(aes(fill = Days_post_transplant)) +
  theme +
  plot_fill(Days_post_transplant = TRUE) +
  scale_y_log10()

ggplot(counts, aes(x = Days_post_transplant, y = Count)) +
  geom_boxplot(aes(fill = Days_post_transplant)) +
  geom_point() +
  theme +
  plot_fill(Days_post_transplant = TRUE) +
  scale_y_log10()

```

## Decontam

```{r}
## Prevalence method
sample_data(fungi.pos.filt)$is.neg <- grepl(pattern = 'Neg', sample_data(fungi.pos.filt)$Type)
fungi.contamdf.prev <- isContaminant(fungi.pos.filt, method="prevalence", neg="is.neg", batch = sample_data(fungi.pos.filt)$MiseqID, threshold = 0.3) 

hist(fungi.contamdf.prev$p, 100)
table(fungi.contamdf.prev$contaminant) #27
rownames(fungi.contamdf.prev)[which(fungi.contamdf.prev$contaminant)]
```

```{r}
# Make phyloseq object of presence-absence in negative controls and true samples
fungi.ps.pa <- transform_sample_counts(fungi.pos.filt, function(abund) 1*(abund>0))

fungi.ps.pa.neg <- prune_samples(grepl(pattern = 'Neg', sample_data(fungi.pos.filt)$Type), fungi.ps.pa) 
fungi.ps.pa.pos <- prune_samples(sample_data(fungi.ps.pa)$Type == "Sample", fungi.ps.pa)

# Make data.frame of prevalence in positive and negative samples
fungi.df.pa <- data.frame(fungi.pa.pos=taxa_sums(fungi.ps.pa.pos), fungi.pa.neg=taxa_sums(fungi.ps.pa.neg),
                      contaminant=fungi.contamdf.prev$contaminant)

p <- 
ggplot(data=fungi.df.pa, aes(x=fungi.pa.neg, y=fungi.pa.pos)) + 
  geom_point(aes(color=contaminant), size = 1, alpha = 0.8) +
  theme(legend.position = "none", legend.justification = c("left")) +
  #geom_text_repel(label = ifelse(df.pa$contaminant == TRUE, rownames(df.pa), ""), max.overlaps = 6, cex=1.5) + 
  labs(title = "Fungi decontam",
       colour = "Contaminant") +
  xlab("Prevalence (Negative Controls)") + 
  ylab("Prevalence (True Samples)") +
  scale_colour_manual(values = c("TRUE" = "#E60031", "FALSE" = "grey")) +
  scale_y_continuous(limits = c(0, 500)) +
  scale_x_continuous(limits = c(0, 25))

ggsave("Microbiome/Data/ITS/contaminant/fungi-decontam.png", p, width = 8, height = 8, units = 'cm')

rm(fungi.ps.pa)
rm(fungi.ps.pa.neg)
rm(fungi.ps.pa.pos)
```

```{r}
# Remove contaminants from phyloseq
fungi.decon <- prune_taxa(!fungi.contamdf.prev$contaminant, fungi.pos.filt)
#fungi.decon <- prune_taxa(data.frame(tax_table(fungi.decon))$ASV != "Sarocladium kiliense|s-14", fungi.decon)
fungi.decon <- readRDS("Microbiome/Data/ITS/fungi.decon.rds")
```

## Check dataset post contamination filtering

```{r}
fungi.decon.df <- data.frame("Reads" = colSums(otu_table(fungi.decon)),
                             "Type" = sample_data(fungi.decon)$Type) 

fungi.decon.df %>%
  filter(Reads > 2000, Type == "ExNegCtrl") %>%

ggplot(aes(x = rownames(.), y = Reads)) +
  geom_col(aes(fill = Type)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  geom_hline(aes(yintercept = 2000), linetype = "dashed", color = "red")
```

## Tip glom 

```{r}
# smux n --time=1-00:00:00 -J tip --mem 16GB --ntasks 4

# module load R/4.1.0-mkl

## Agglomerate ASVs at the phylogenetic 
## Cluster
#fungi.glom <- list()
#c <- c(0.015, 0.02, 0.05, 0.1, 0.2)
#for (i in 1:length(c)){
#  h <- c[i]
#  fungi.glom[[as.character(h)]] <- tip_glom(fungi.decon, h = h) # smaller h equals to less cutting
#}
#saveRDS(fungi.glom, "fungi.glom.rds")
fungi.glom <- readRDS("Microbiome/Data/ITS/fungi.glom.rds")

fungi.decon # 1697
length(rownames(tax_table(fungi.glom$`0.02`)) ) #615

plot_tree(fungi.decon, "treeonly", title="Before tip_glom()")
plot_tree(fungi.glom$`0.05`, "treeonly",  title="After tip_glom()")
plot_tree(fungi.glom$`0.1`, "treeonly",  title="After tip_glom()")
plot_tree(fungi.glom$`0.2`, "treeonly",  title="After tip_glom()")


# Add taxas back
tax_table(fungi.glom$`0.02`) <- tax_table(fungi.decon)[rownames(tax_table(fungi.glom$`0.02`)),]

# metric for best tip glom
```

## Prevalence filter and duplicate removal

```{r}
# Prevalence filtering
fungi.prev.filtered <- core(fungi.glom$`0.02`, detection = 1, prevalence = 0, include.lowest = FALSE)

# Remove duplicate samples and keep the ones with the highest reads + remove R healthy
fungi.summary <- data.frame('Reads' = colSums(otu_table(fungi.prev.filtered)), 
                           'Patient_days' = sample_data(fungi.prev.filtered)$Patient_days,
                           "Type" = sample_data(fungi.prev.filtered)$Type)

fungi.keep <- fungi.summary %>%
  mutate(SampleID = rownames(fungi.summary)) #%>%
  #filter(!SampleID %in% c("16S.B0003682", "16S.B0003680")) # remove R healthy lung of patient 19 BGP, (left tx one is 3712)
duplicates <- fungi.keep %>% group_by(Patient_days) %>% filter(n() > 1) %>% group_by(Patient_days) %>% top_n(-1, Reads) %>% pull(SampleID)
fungi.keep <-  fungi.keep[!(fungi.keep$SampleID %in% duplicates),] %>% pull(SampleID)
 
# Filtered phyloseq
fungi.dup.filt <- prune_samples(fungi.keep, fungi.prev.filtered) 
```


## Check after prevalence filtering

```{r}
other.controls <- data.frame("SampleID" = sample_data(fungi.raw)$SampleID,
                             "Reads" = colSums(otu_table(fungi.raw)),
                             "Type" = sample_data(fungi.raw)$Type) %>%
  filter(Reads <= 200, grepl("Neg", Type))

fungi.dup.df <- data.frame("SampleID" = sample_data(fungi.dup.filt)$SampleID,
                           "Reads" = colSums(otu_table(fungi.dup.filt)),
                           "Type" = sample_data(fungi.dup.filt)$Type) %>%
  full_join(other.controls)

fungi.dup.df$Type <- ifelse(fungi.dup.df$Type == "Sample" & fungi.dup.df$Reads < 2000, "Excluded", 
                              ifelse(fungi.dup.df$Type == "Sample" & fungi.dup.df$Reads >= 2000, "Sample", fungi.dup.df$Type))

p <- 
  
  fungi.dup.df %>%
  mutate(Type = factor(Type, levels = c("BronchNegCtrl", "ExNegCtrl", "PCRNegCtrl", "Excluded", "Sample"))) %>%
  
ggplot(aes(x = Type, y = Reads)) +
 # geom_violin(aes(fill = Type)) +
  geom_boxplot(fill = "grey", outlier.shape = NA, alpha = 0.8) +
  geom_jitter(width = 0.2) +
  #geom_dotplot(binaxis = 'y', stackdir = 'center', dotsize = 0.5, width = 0.0002) +
  labs(title = "Fungi read count filter",
       y = "Log ASV Reads",
       x  = " ") +
  geom_hline(aes(yintercept = 2000), linetype = "dashed", color = "grey") +
  scale_y_log10() +
  theme(legend.position = "none") 

ggsave("Microbiome/Data/ITS/contaminant/fungi-read-filter.png", p, width = 12, height = 12, units = 'cm')
```


```{r}
fungi.dup.df %>% group_by(Type) %>% count() %>% filter(Type %in% c("Sample", "Excluded")) %>%
  mutate(Percentage = round(100*n/sum(.$n))) %>%
  ungroup() %>%
  add_row(Type = "Total", n = sum(.$n), Percentage = NA)
```


## Read threshold

```{r}
# Read threshold filtering
threshold <- 2000
fungi.filt <- prune_samples(colSums(otu_table(fungi.dup.filt)) >= threshold, fungi.dup.filt)

# remove control
fungi.filt <- prune_samples(sample_names(fungi.filt) != "ITS.NegCtrlB3", fungi.filt)

data.frame('Summary' = c('ASVs/genera', 'Samples'),
           'Fungi_before_ASV_filtering' = dim(otu_table(fungi.raw)), # 1901 357 
           'Fungi_after_contam_filtering' = dim(otu_table(fungi.decon)), # 1691 310 
           'Fungi_after_prev/thresh_filtering' = dim(otu_table(fungi.filt))) # 603 235
```


## Save filtered library size

```{r}
# Estimate Shannon index and retrieve the number of ASVs (observed) 
sample_data(fungi.filt)$Diversity <- estimate_richness(fungi.filt, split = TRUE, measures = c('Shannon'))$Shannon
# Change sample names
sample_names(fungi.filt) <- sample_data(fungi.filt)$Patient_days
saveRDS(fungi.filt, "Microbiome/Data/ITS/fungi.filt.rds")
# fungi.filt <- readRDS("Microbiome/Data/ITS/fungi.filt.rds")

fungi_seq_metadata <- data.frame(sample_data(fungi.filt), row.names = NULL) %>%
  dplyr::select(Record.ID, Days, Patient_days, Lobe, Diversity)
saveRDS(fungi_seq_metadata, "Microbiome/Data/ITS/fungi_seq_metadata.rds")
# Now check patient selection before next chunk

rm(fungi.decon.df, fungi.ps.pa.pos, fungi.ps.pa.neg, fungi.ps.pa, fungi.ctrls.filt, fungi.ctrls.filt, fungi.ctrls, fungi.tax.filt, fungi.min.filt, other.controls, fungi.contamdf.prev, fungi.dup.filt, fungi.df.pa, fungi.dup.df, fungi.summary, fungi.raw, fungi.glom, fungi.pos.ctrl, fungi.pos.filt, fungi.prev.filtered, fungi.decon) 
```


# Create complete phyloseq object

```{r}
f <- fev_at_bronch(fungi_seq_metadata, 
                      spirometry_metadata,
                      unique(fungi_seq_metadata$Record.ID)) # need record.id, days and patient days

fungi_metadata <- fungi_seq_metadata %>%
  left_join(master_metadata$bronch_metadata[, c("Record.ID", "Patient_days", "Days")], by = c("Record.ID", "Days"), suffix = c("_lobe", "")) %>%
  left_join(master_metadata$meds_metadata, by = c("Patient_days")) %>%
  left_join(master_metadata$general_metadata, by = "Record.ID") %>% 
  mutate(Patient_days = gsub("_Left|_Right", "", Patient_days_lobe)) %>%
  
  left_join(f, by = c("Patient_days")) %>%
  left_join(master_metadata$donor, by = "Record.ID") %>%
  left_join(master_metadata$bal_class_hosp) %>%
  left_join(clad_assessment[,c("Record.ID", "Transplanted")], by = "Record.ID") %>%
  mutate(Lobe = case_when(Lobe == Transplanted ~ "transplanted",
                          Lobe != Transplanted ~ "native",
                          is.na(Lobe) ~ "transplanted"))
fungi_metadata <- annotate_clad(fungi_metadata)
fungi_metadata <- classify_dosage(fungi_metadata)

fungi_metadata <- fungi_metadata %>%
  mutate(Time_interval_detailed = case_when(Days <= 42 ~ "<1.5 months",
                                         Days > 42 & Days <= 92 ~ "1.5-3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_detailed = factor(Time_interval_detailed, c("<1.5 months", "1.5-3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Time_interval_combined = case_when(Days <= 92 ~ "<3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_combined = factor(Time_interval_combined, c("<3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Days_interval = case_when(Days <= 183 ~ "<6 months",
                                   Days > 183 ~ ">6 months"), 
         Days_interval = factor(Days_interval, c("<6 months", ">6 months")),
         
         Days_interval2 = case_when(Days <= 92 ~ "<3",
                                   Days > 92 ~ ">3"), 
         Days_interval2 = factor(Days_interval2, c("<3", ">3")),
         
         Immunosuppression = case_when(Days <= 183 ~ "High",
                                         Days > 183 & Days <= 365 ~ "Mid",
                                         Days > 365 ~ "Low"), 
         Immunosuppression = factor(Immunosuppression, c("High", "Mid", "Low")),
         
         Months_grouped = case_when(Days <= 42 ~ "<1.5",
                                    Days > 42 & Days <= 92 ~ "1.5-3",
                                    Days > 92 & Days <= 183 ~ "4-6",
                                    Days > 183 & Days <= 273 ~ "7-9",
                                    Days > 273 & Days <= 365 ~ "10-12",
                                    Days > 365 & Days <= 547 ~ "13-18",
                                    Days > 547 ~ ">18"), 
         Months_grouped = factor(Months_grouped, c("<1.5", "1.5-3", "4-6", "7-9", "10-12", "13-18", ">18")),
         
         Months = round(Days/30.417, 1) )

rownames(fungi_metadata) <- fungi_metadata$Patient_days_lobe

# Add new metadata to phyloseq. Rows need to match ps row order
fungi_metadata <- fungi_metadata[sample_names(fungi.filt),]
identical(rownames(fungi_metadata), sample_names(fungi.filt)) #TRUE
identical(colnames(otu_table(fungi.filt)), rownames(fungi_metadata)) #TRUE
fungi <- phyloseq::phyloseq(otu_table(fungi.filt), tax_table(fungi.filt), sample_data(fungi_metadata), phy_tree(fungi.filt))

# Subset phyloseq object for study
fungi <- phyloseq::subset_samples(fungi, Record.ID %in% unique(c(stable_patients, biomarker_patients)) ) # patients
fungi <- phyloseq::subset_samples(fungi, Patient_days != "44_182") # remove sex mismatched sample
fungi <- phyloseq::subset_samples(fungi, Lobe == "transplanted" & Months < 26)
fungi <- phyloseq::subset_samples(fungi, is.na(days_to_clad) | days_to_clad >= -30) # only tx lobes, and samples less than 30M
sample_data(fungi)$GIT_ischemic_time_hours <- round(as.numeric(sample_data(fungi)$GIT_ischemic_time/60), 1)

#sample_data(fungi)$Ischemic_time_cat <- ifelse(sample_data(fungi)$GIT_ischemic_time <= 250, "<250",
#                                          ifelse(sample_data(fungi)$GIT_ischemic_time > 250 & sample_data(fungi)$GIT_ischemic_time <= 400, "250-400", ">400"))
sample_names(fungi) <- sample_data(fungi)$Patient_days

# Checks
length(unique(c(stable_patients, biomarker_patients)) ) # 69 unique patients
length(unique(sample_data(fungi)$Record.ID)) # 56

length(all_selected_samples_collected_all_datasets %>% filter(Dataset == "Fungi") %>% pull(Patient_days) ) #200
length(data.frame(sample_data(fungi))$Patient_days ) # 200

setdiff(all_selected_samples_collected_all_datasets %>% filter(Dataset == "Fungi") %>% pull(Patient_days), data.frame(sample_data(fungi))$Patient_days ) 

fungi_metadata <- data.frame(sample_data(fungi))
colnames(fungi_metadata) <- ifelse(grepl("Piperacillin.Tazobactam", colnames(fungi_metadata)), "Piperacillin-Tazobactam", colnames(fungi_metadata)  )
colnames(fungi_metadata) <- ifelse(grepl("Mycophenolate.Mofetil", colnames(fungi_metadata)), "Mycophenolate-Mofetil", colnames(fungi_metadata)  )
rownames(fungi_metadata) <- fungi_metadata$Patient_days

saveRDS(fungi, "Microbiome/Data/ITS/fungi.rds")
saveRDS(fungi_metadata, "Microbiome/Data/ITS/fungi_metadata.rds")

rm(f, fungi.filt)

#fungi <- readRDS("Data/16S/Microbiome/bact.rds")
#fungi_metadata <- readRDS("Data/16S/Microbiome/bact_metadata.rds")
```

```{r}
# summarize_phyloseq(fungi) 

# Number of Genera and Phyla
data.frame("genus" = data.frame(tax_table(fungi))$Genus) %>% group_by(genus) %>% dplyr::count() %>% arrange(desc(n))
head(names(sort(desc(table(data.frame(tax_table(fungi))$Genus)))))
data.frame("phylum" = data.frame(tax_table(fungi))$Phylum) %>% group_by(phylum) %>% dplyr::count() %>% arrange(desc(n))

# Number of ASVs and samples
dim(otu_table(fungi)) # 603 200

# Check Aspergillus
aspergillus <- as.data.frame(fungi@otu_table) %>% filter(grepl(c("Aspergillus"), rownames(.)))
aspergillus[,colSums(aspergillus) > 0]
rm(aspergillus)
```

## Check dataset post prevalence and read filtering

```{r}
# Compare controls to all other samples
fung.compare <- c(fung.keep, fung.ctrl.list) #bind two sample lists (includes samples above threshold and all controls)

#filter pyloseq object according to selected samples
fung.compare.p <- prune_samples(fung.compare, fungi.data.removed) # 113 samples
```

```{r}
fung.compare.p <- prune_samples(colSums(otu_table(fung.compare.p)) > 0, fung.compare.p) # 314 samples

# Perform Cumulative Sum Scaling transformation
fung.compare.CSS <- fung.compare.p
otu_table(fung.compare.CSS) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(fung.compare.CSS), p = 0.5)), taxa_are_rows = TRUE)

# Perform Cumulative Sum Scaling transformation followed by log
fung.compare.logCSS <- microbiome::transform(fung.compare.CSS, transform = 'log')

# About 20 overlapping
ggsave("ITS/All-pcoa-uni-logcss.pdf",
  
  plot_ordination(fung.compare.logCSS, ordinate(fung.compare.logCSS, method="PCoA", distance="unifrac", weighted=TRUE), label = "Sample_ID") +
  #geom_path(aes(group = Patient), col = 'grey90') +
  stat_ellipse(aes(fill = Type), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
  scale_shape_identity() +
  geom_point(aes(fill = Type), shape = 21, size = 2, stroke = 0.25) + # Days_post_transplant
  theme_minimal() +
  scale_fill_manual(values = c("black", "red", "green", "orange", "blue"), aesthetics = "fill")

, width = 29.7, height = 21, units = 'cm') 
```


## Comparison between samples with repeated sequencing

```{r}
comp <- bact.seq.metadata %>%
  filter(Code %in% c("DOM","WIM","GJH","PEB","BGP","MGQ")) %>%
  pull(Sample_ID)

p.comp <- prune_samples(comp, bact.data.raw)

p <- p.comp %>%
  core(detection = 100, prevalence = 0.15) %>%
  tax_glom(taxrank = 'Genus', NArm = T) %>%
  microbiome::transform(transform = 'compositional', target = 'OTU') %>%
  psmelt()

ggplot(p, aes(x=Sample_ID, y=Abundance, fill=Genus, NArm = F)) + # Patient_days, Days_post_transplant
  geom_bar(color= 'black' , stat = 'identity', position = 'fill', size = 0.01) + #remove position = fill and the bars will not touch the top
  labs(x = "Days", y = "Relative abundance", title = "Bacterial taxonomy composition") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 6, angle = 90),
        strip.text = element_text(size = 10),
        legend.key.size = unit(0.1, "cm")) +
  guides(fill = guide_legend(ncol = 1)) +
  facet_wrap(~Code, scales = "free") #space = "free_x"
```


