---
title: "Transcriptomics"
author: "Giulia"
date: "24/07/2024"
output:
---

# Samplesheet

```{r}
# Select samples of interest
m1 <- read.csv("Transcriptomics/Data/novogene/Run1/sample-list-run1.csv") %>% dplyr::select(SampleID, Patient_days, Run) 
m2 <- read.csv("Transcriptomics/Data/novogene/Run2/sample-list-run2.csv") %>% dplyr::select(SampleID, Patient_days, Run)
m3 <- read.csv("Transcriptomics/Data/novogene/Run3/sample-list-run3.csv") %>% dplyr::select(SampleID, Patient_days, Run)
qc_report <- read.csv("Transcriptomics/Data/novogene/qc-report-all-runs.csv")
sequenced <- read.csv("Transcriptomics/Data/nf-core/samples-sequenced.csv") %>% pull(SampleID)

a <- rbind(m1, m2, m3) %>% 
  mutate(SampleID = as.character(SampleID),
         Record.ID = as.numeric(gsub("_.*", "", Patient_days)),
         Days = as.numeric(gsub(".*_", "", Patient_days))) %>%
  rename(Batch = Run) %>%
  #filter(SampleID %in% sequenced) %>% # those that were actually sequenced
  left_join(qc_report) %>%
  mutate(Batch = factor(Batch, levels = c("1", "2", "3")))
rownames(a) <- a$SampleID
dim(a) #264 12

# Check multiqc report for sample to exclude
#multiqc_report <- read.delim("Transcriptomics/Data/Laxy/laxyrun3/75NDVoxSyuTccJ0ZxTjiDt/output/sikRun/fastqcReport/multiqc_data/multiqc_fastqc.txt", header = TRUE) %>% mutate(SampleID = gsub("_1|_2|R", "", Sample)) %>% filter(SampleID %in% a$SampleID)
# multiqc_report %>% filter(per_tile_sequence_quality == "fail" | 
#                          per_base_sequence_content == "fail" | per_sequence_gc_content == "fail" | 
#                          adapter_content == "fail" ) %>%
#  dplyr::select(Sample, per_tile_sequence_quality, per_base_sequence_content, per_sequence_gc_content, adapter_content)

samples_sequenced_detailed <- data.frame(SampleID = sequenced) %>% 
  mutate(SampleID = as.character(SampleID)) %>% 
  left_join(a) %>% 
  left_join(qc_report[, c("SampleID", "Sample.QC.Results", "Sample.QC.Memo")]) %>%
  left_join(clad_assessment[, c("Record.ID", "clad_clinical")]) %>%
  mutate(Group = ifelse(clad_clinical == TRUE, "CLAD", "CLAD-free"),
         Group = factor(Group, levels = c("CLAD-free", "CLAD")),
         sample = paste(Group, Patient_days, sep = "_"),
         sample = gsub("-", "", sample),
         sample = ifelse(SampleID == "5134", "CLADfree_22_382", sample), # change duplicate
         sample = ifelse(SampleID == "3672", "CLADfree_19_300R", sample),
         sample = ifelse(SampleID == "3704", "CLADfree_19_300L", sample)) %>%
  dplyr::select(!clad_clinical)
dim(samples_sequenced_detailed) # 264

rm(m1, m2, m3, a, qc_report, sequenced)

length(unique(samples_sequenced_detailed$Record.ID)) # 54

# Samples sequenced vs those that failed
samples_sequenced_detailed %>% group_by(Group) %>% dplyr::count() # 196 68
```

# Import count tables

```{r}
# Counts
nf_core_raw <- read.delim(paste0(preprocessing_path, "Transcriptomics/Data/nf-core/rnaseq_output/star_salmon/counts/salmon.merged.gene_counts_length_scaled.tsv"), sep = "\t", header = TRUE) 
dim(nf_core_raw) # 63241 266

nf_core_raw <- nf_core_raw %>%
  rename(ENSEMBL = gene_id, SYMBOL = gene_name) %>%
  mutate(ENTREZID = mapIds(org.Hs.eg.db, keys = ENSEMBL, keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>%
  relocate(ENSEMBL, SYMBOL, ENTREZID)
dim(nf_core_raw) # 63241   
length(nf_core_raw$ENSEMBL) #63241

ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
biotype <- getBM(attributes = c("ensembl_gene_id","gene_biotype", "description", "chromosome_name"),
                 filters = c("ensembl_gene_id"), 
                 values = nf_core_raw$ENSEMBL, 
                 mart = ensembl)
colnames(biotype)[1] <- c("ENSEMBL")

# saveRDS(biotype, "Transcriptomics/Data/biotype.rds")
biotype <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/biotype.rds"))
# rm(ensembl)

length(unique(biotype$ENSEMBL)) #63114
dim(biotype) #63114

nf_core_raw <- nf_core_raw %>%
  left_join(biotype) %>% 
  relocate(SYMBOL, ENTREZID, colnames(biotype)) 
dim(nf_core_raw) #63241

length(nf_core_raw$ENSEMBL) #63241
length(nf_core_raw$SYMBOL) #63241
length(unique(nf_core_raw$SYMBOL)) #61617

colnames(nf_core_raw) <- ifelse(colnames(nf_core_raw) == "CLADfree_5_12", "CLADfree_5_16", colnames(nf_core_raw))
colnames(nf_core_raw) <- ifelse(colnames(nf_core_raw) == "CLAD_10_265", "CLAD_10_267", colnames(nf_core_raw))

saveRDS(nf_core_raw, paste0(preprocessing_path, "Transcriptomics/Data/nf_core_raw.rds"))
#nf_core_raw <- readRDS(preprocessing_path, "Transcriptomics/Data/nf_core_raw.rds"))
```

```{r}
rna_seq_metadata <- samples_sequenced_detailed %>%
  filter(!SampleID %in% c("3672", "4138", "5134")) # 3672 is 19BGP right native, 4138 is sex mismatched, and 5134 is duplicate

rna_seq_metadata <- rna_seq_metadata %>%
  mutate(Days = ifelse(Record.ID == 5 & Days == 12, 16, Days),
         Patient_days = ifelse(Patient_days == "5_12", "5_16", Patient_days),
         sample = ifelse(sample == "CLADfree_5_12", "CLADfree_5_16", sample),
         Days = ifelse(Record.ID == 10 & Days == 265, 267, Days),
         Patient_days = ifelse(Patient_days == "10_265", "10_267", Patient_days),
         sample = ifelse(sample == "CLAD_10_265", "CLAD_10_267", sample)) # change mislabelled samples
saveRDS(rna_seq_metadata, "Transcriptomics/Data/rna_seq_metadata.rds")
write.csv(rna_seq_metadata, "Transcriptomics/Data/rna_seq_metadata.csv")

rna_metadata <- rna_seq_metadata %>% 
  filter(Patient_days %in% c(all_selected_samples_collected_all_datasets %>% filter(Dataset == "Transcriptomics") %>% pull(Patient_days)) )
rownames(rna_metadata) <- rna_metadata$Patient_days

transcriptomics_counts <- nf_core_raw %>% dplyr::select(c(colnames(nf_core_raw)[c(1:6)], rna_metadata$sample)) # samples of interest
colnames(transcriptomics_counts)[-c(1:6)] <- gsub("CLAD_|CLADfree_|R|L", "", colnames(transcriptomics_counts)[-c(1:6)])
rownames(transcriptomics_counts) <- make.unique(transcriptomics_counts$SYMBOL)
dim(rna_metadata) #234 14
identical(rna_metadata$Patient_days, colnames(transcriptomics_counts)[-c(1:6)]) #TRUE

length(unique(transcriptomics_counts$SYMBOL)) #61617

saveRDS(transcriptomics_counts, "Transcriptomics/Data/transcriptomics_counts.rds")
transcriptomics_counts <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/transcriptomics_counts.rds"))
saveRDS(rna_metadata, "Transcriptomics/Data/rna_metadata.rds")
rna_metadata <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/rna_metadata.rds"))
```

## Process for DESEQ2

```{r}
dge <- DGEList(counts = round(transcriptomics_counts[,-c(1:6)]), 
               genes = transcriptomics_counts[,c(1:6)], 
               group = rna_metadata$Group,
               samples = rna_metadata,
               remove.zeros = TRUE)
saveRDS(dge, "Transcriptomics/Data/dge.rds")
# dge <- readRDS("Transcriptomics/Data/dge.rds")

# Filter lowly expressed genes
d <- model.matrix(~Group, data = dge$samples)
dge_filtered <- dge[filterByExpr(dge, group = d), , keep.lib.sizes = FALSE]
saveRDS(dge_filtered, "Transcriptomics/Data/dge_filtered.rds")
# dge_filtered <- readRDS("Transcriptomics/Data/dge_filtered.rds")

rna_combat <- ComBat_seq(dge_filtered$counts, 
                             group = as.factor(dge_filtered$samples$Group),
                             batch = as.factor(dge_filtered$samples$Batch))
saveRDS(rna_combat, "Transcriptomics/Data/rna_combat.rds")
# rna_combat <- readRDS("Transcriptomics/Data/rna_combat.rds")

dge_filtered$samples$lib.size <- NULL
dge_filtered$samples$norm.factors <- NULL

rna_corrected <- DGEList(counts = rna_combat, 
                          genes = dge_filtered$genes, 
                          samples = dge_filtered$samples)

# Identify Y chromosome genes
sex_associated_genes <- rna_corrected$genes %>% 
  filter(chromosome_name %in% c("Y") | SYMBOL == "XIST") %>%
  pull(ENSEMBL)
rna_corrected <- rna_corrected[!rna_corrected$genes$ENSEMBL %in% sex_associated_genes, ]

protein_coding <- rna_corrected$genes %>% 
  filter(gene_biotype == "protein_coding") %>%
  pull(ENSEMBL)
rna_corrected <- rna_corrected[rna_corrected$genes$ENSEMBL %in% protein_coding, ]

dim(rna_corrected) #12707   234
#table(is.na(rna_corrected$counts))
#table(is.na(rownames(rna_corrected$counts)))
#rm(clad_combat)
#rm(clad_dge_filtered)

rna_corrected$samples$lib.size <- NULL
rna_corrected$samples$norm.factors <- NULL

rm(protein_coding)
```

## Process for Limma

```{r}
# Changed order so that it is done before filtering for lowly expressed genes
genes <- transcriptomics_counts[,c(1:6)]
transcriptomics_counts <- transcriptomics_counts[,-c(1:6)]

dge <- DGEList(counts = transcriptomics_counts, 
               genes = genes, 
               group = rna_metadata$Group,
               samples = rna_metadata,
               remove.zeros = TRUE)
saveRDS(dge, paste0(preprocessing_path, "Transcriptomics/Data/limma/dge.rds"))
#dge <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/limma/dge.rds"))

rna_combat <- ComBat_seq(dge$counts, 
                             group = as.factor(dge$samples$Group),
                             batch = as.factor(dge$samples$Batch))
#saveRDS(rna_combat, paste0(preprocessing_path, "Transcriptomics/Data/limma/rna_combat.rds"))
#rna_combat <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/limma/rna_combat.rds"))

identical(dge$samples$Patient_days, colnames(rna_combat)) # TRUE
identical(rownames(dge$genes), rownames(rna_combat)) # TRUE

# Filter lowly expressed genes
d <- model.matrix(~Group, data = dge$samples)
dge_combat_filtered <- rna_combat[filterByExpr(rna_combat, d), ]
#saveRDS(dge_combat_filtered, paste0(preprocessing_path, "Transcriptomics/Data/limma/dge_combat_filtered.rds"))
#dge_combat_filtered <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/limma/dge_combat_filtered.rds"))

identical(dge$samples$Patient_days, colnames(dge_combat_filtered)) # TRUE

g <- dge$genes %>% rownames_to_column("ID") %>% filter(ID %in% rownames(dge_combat_filtered)) %>% column_to_rownames("ID")
identical(rownames(g), rownames(dge_combat_filtered)) # TRUE

dge$samples$lib.size <- NULL
dge$samples$norm.factors <- NULL

dge_corrected <- DGEList(counts = dge_combat_filtered, 
               genes = g, 
               group = dge$samples$Group,
               samples = dge$samples,
               remove.zeros = TRUE)

# Identify Y chromosome genes
sex_associated_genes <- dge_corrected$genes %>% 
  filter(chromosome_name %in% c("Y") | SYMBOL == "XIST") %>%
  pull(ENSEMBL)
dge_corrected <- dge_corrected[!dge_corrected$genes$ENSEMBL %in% sex_associated_genes, ]

protein_coding <- dge_corrected$genes %>% 
  filter(gene_biotype == "protein_coding") %>%
  pull(ENSEMBL)
dge_corrected <- dge_corrected[dge_corrected$genes$ENSEMBL %in% protein_coding, ]
saveRDS(dge_corrected, paste0(preprocessing_path, "Transcriptomics/Data/limma/dge_corrected.rds"))

d <- model.matrix(~Group, data = dge_corrected$samples)
rna_corrected <- edgeR::calcNormFactors(dge_corrected)
```

## Metadata

```{r}
r <- fev_at_bronch(rna_corrected$samples, 
                      spirometry_metadata,
                      unique(rna_corrected$samples$Record.ID)) # need record.id, days and patient days

rna_corrected$samples <- rna_corrected$samples %>%
  dplyr::select(!c("Group", "group")) %>%
  left_join(master_metadata$meds_metadata, by = c("Patient_days")) %>%
  left_join(master_metadata$general_metadata, by = "Record.ID") %>% 
  left_join(r, by = c("Patient_days")) %>%
  left_join(master_metadata$donor, by = "Record.ID") %>% 
  left_join(master_metadata$bal_class_hosp)
rna_corrected$samples <- annotate_clad(rna_corrected$samples)
rna_corrected$samples <- classify_dosage(rna_corrected$samples)

rna_corrected$samples <- rna_corrected$samples %>%
  mutate(Time_interval_detailed = case_when(Days <= 42 ~ "<1.5 months",
                                         Days > 42 & Days <= 92 ~ "1.5-3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_detailed = factor(Time_interval_detailed, c("<1.5 months", "1.5-3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Time_interval_combined = case_when(Days <= 92 ~ "<3 months",
                                         Days > 92 & Days <= 183 ~ "3-6 months",
                                         Days > 183 & Days <= 365 ~ "6-12 months",
                                         Days > 365 ~ ">12 months"), 
         Time_interval_combined = factor(Time_interval_combined, c("<3 months", "3-6 months", "6-12 months", ">12 months")),
         
         Days_interval = case_when(Days <= 183 ~ "<6 months",
                                   Days > 183 ~ ">6 months"), 
         Days_interval = factor(Days_interval, c("<6 months", ">6 months")),
         
         Days_interval2 = case_when(Days <= 92 ~ "<3 months",
                                   Days > 92 ~ ">3 months"), 
         Days_interval2 = factor(Days_interval2, c("<3 months", ">3 months")),
         
         Immunosuppression = case_when(Days <= 183 ~ "High",
                                         Days > 183 & Days <= 365 ~ "Mid",
                                         Days > 365 ~ "Low"), 
         Immunosuppression = factor(Immunosuppression, c("High", "Mid", "Low")),
         Months = round(Days/30.417, 1),
         Months_grouped = case_when(Days <= 42 ~ "<1.5",
                                    Days > 42 & Days <= 92 ~ "1.5-3",
                                    Days > 92 & Days <= 183 ~ "4-6",
                                    Days > 183 & Days <= 273 ~ "7-9",
                                    Days > 273 & Days <= 365 ~ "10-12",
                                    Days > 365 & Days <= 547 ~ "13-18",
                                    Days > 547 ~ ">18"), 
         Months_grouped = factor(Months_grouped, c("<1.5", "1.5-3", "4-6", "7-9", "10-12", "13-18", ">18")),
         Total_ant_cat_split = ifelse(Total_ant_num == 0, "0",
                                          ifelse(Total_ant_num > 0 & Total_ant_num <= 2, "1-2", "3-5")),
         Total_ant_cat_split = ifelse(Total_ant_num == 0, "0",
                                          ifelse(Total_ant_num > 0 & Total_ant_num <= 2, "1-2", "3-5")),
         Total_ant_cat_split = factor(Total_ant_cat_split, levels = c("0", "1-2", "3-5"))
)

rownames(rna_corrected$samples) <- rna_corrected$samples$Patient_days
identical(colnames(rna_corrected$counts), rownames(rna_corrected$samples)) # TRUE
```


```{r}
saveRDS(rna_corrected, paste0(preprocessing_path, "Transcriptomics/Data/limma/rna_corrected.rds"))
rna_corrected <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/deseq/rna_corrected.rds"))
dim(rna_corrected) #13796   234


r <- vst(rna_corrected$counts)
write.csv(r, "rna_counts_table.csv")
write.csv(rna_corrected$samples %>% dplyr::select(!c("ParticipantInitials", "Gender", "Age", "transplant_date", "Donor_gender", "Donor_age")), "rna_metadata_table.csv")

# rna_corrected$samples %>% group_by(Months_grouped) %>% dplyr::count()

rna_corrected_normalised <- voom(rna_corrected, d, plot = FALSE)
saveRDS(rna_corrected_normalised, paste0(preprocessing_path, "Transcriptomics/Data/limma/rna_corrected_normalised.rds"))
rna_corrected_normalised <- readRDS(paste0(preprocessing_path, "Transcriptomics/Data/limma/rna_corrected_normalised.rds"))

dim(rna_corrected_normalised) #13796   234
#table(is.na(rna_corrected$counts))
#table(is.na(rownames(rna_corrected$counts)))

rm(r, nf_core_raw, dge, dge_filtered, rna_combat, transcriptomics_counts, dge_combat_filtered, dge_corrected, protein_coding)
```


# Check for batch effects
## PCA

```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_corrected$counts,
                              colData = rna_corrected$samples,
                              design = ~1)  
vsd <- vst(dds, blind=FALSE)

pca_list <- pca_process(as.data.frame(assay(vsd)), as.data.frame(colData(vsd)))

p <-
  
  ggplot(pca_list$data.PCA, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(group = Batch, fill = Batch, colour = Batch), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) +
  geom_point(aes(fill = Batch), shape = 21, size = 2) +
  theme +
  labs(title = 'PCA ordination',
       fill = "Time interval",
       colour = "Time interval",
       shape = "BAL class",
       x = paste0('PC1 ', round(pca_list$varexp[1], 2), '%'),
       y = paste0('PC2 ', round(pca_list$varexp[2], 2), '%')) 

ggsave("pca-batch.pdf", p, width = 9, height = 7, units = 'cm')

adonis2(formula = t(countdata) ~ Batch,
                  data = pca.list$metadata, 
                  permutations = 999,
                  method='eu')
#          Df SumOfSqs    R2      F Pr(>F)  
#Batch      2   100863 0.17945 17.277  0.001 *** vsn
#Batch      2   195906 0.1339 12.213  0.001 *** cpm best

rm(dds, vsd, pca_list)
```

```{r}
# LogCPM total counts
c <- data.frame(countdata, check.names = FALSE) %>% rownames_to_column("Gene") %>% pivot_longer(., `61_781`:`66_719`, names_to = "Patient_days", values_to = "logCPM") %>% left_join(metadata[, c("Patient_days", "Batch")]) %>% arrange(Batch) %>%
  mutate(Batch = factor(Batch))

name <- "rna-counts-logcpm"

p <- ggplot(c, aes(x = factor(Patient_days, levels = unique(Patient_days)), y = logCPM)) +
  geom_boxplot(aes(fill = Batch)) +
  theme +
  theme(axis.text.x = element_blank()) +
  labs(title = "Raw counts",
       x = "Samples")

ggsave(paste0(folder, name, ".png"), p, width = 20, height = 10, units = 'cm')

hist(c$logCPM)
```

