---
title: "Library"
author: "Giulia"
date: "27/09/2021"
output:
  html_document: default
  word_document: default
---

# Libraries

```{r, echo=FALSE, results='hide'}
suppressPackageStartupMessages({ 

# Data wrangling and visualisation
library(ggplotify) #turn plot() output into ggplot
library(ggsci)
library(stringr)
library(knitr)
library(magrittr)
library(ggpubr)
library(gsubfn)
library(viridis)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggalluvial)
library(dendsort)
library(tinytex)
library(ggVennDiagram)
library(reshape2)
library(ggrepel)
library(gplots)
library(gridExtra)
library(lubridate)
library(readr)
library(janitor)
#library(ggbeeswarm)
library(circlize) # chorddiagram
#library(ggridges) # ridge plot
library(gt) # tables
library(ggdendro) # dendrograms
library(strex) # stringr manipulation

## Networks
library(igraph)
library(intergraph)
library(GGally) # networks
library(sna)
#library(SpiecEasi) 
#library(NetCoMi)
#library(NetSwan) # network attack
#library(asnipe)
library(rnetcarto) # network participation and role
library(enrichplot)
#library(RCy3) # cytoscape

# Analysis and statistics  
library(statmod)
library(lme4)
library(lmerTest)
library(longpower)
library(cluster)
library(factoextra)
library(fpc)
library(rstatix)
library(broom)
#library(pairwiseAdonis)
library(S4Vectors)
library(data.table)
#library(devtools)
#library(usedist)
library(strucchange) #time split
#library(drc) # curve modelling
library(sjmisc)
library(ggeffects) # curve modelling

## ms pre-processing
#library(pmp)
#library(MSPrep)
#library(impute)
#library(missForest)
library(sva) # batch correction
#library(DEP)
#library(proBatch)
library(SummarizedExperiment)
#library(KEGGREST)
library(FELLA)
library(vsn)
#library(OlinkAnalyze)

## microbiome
#library(doParallel)
#library(Maaslin2)
library(phyloseq)
library(metagenomeSeq)
library(MASS)
library(vegan)
library(microbiome)
#library(microbiomeMarker) DESeq2 wrapper for microbiome
#library(MicrobiomeProfiler)
#library(decontam)
#library(MetaLonDA) # significant intervals over time
#library(syncomR) # Community stability
#library(BacDive) # Aerobe/anaerobe status
library(dysbiosisR)  

library(biomaRt) #microarray
#library(hgu95av2.db) #microarray interlung
#library(hgu133plus2.db) # microarray weigt
# library(GEOquery)

# Mofa
library(MultiAssayExperiment)
#library(psych)
library(MOFA2)
library(survival) # COX factor modelling
library(survminer) # COX factor modelling visualisation
library(corrplot) # correlation plot

  ## rnaseq
#library(RUVSeq)
library(limma)
library(variancePartition)
library(formatR)
library(splines)
library(edgeR)
library(org.Hs.eg.db) # biomart
#library(GO.db)
#library(GeneTonic)
#library(topGO)
#library(pcaExplorer)
#library(ReactomePA) # pathway analysis
library(msigdbr) # molecular signature database (immune)
#library(specmine)
#library(WGCNA)
library(caret) # ML
library(MLeval) # ML
library(caTools) # ML
library(MLmetrics) # ML
#library(BisqueRNA)
#library(Biobase)
#library(seurat)
library(clusterProfiler) # pathway analysis needs tidytree installed
#library(pathfindR) # pathway analysis
library(DESeq2)
library(ashr) # DE shrinkage for DESeq2
library(qusage) # pathway analysis
library(TMixClust) # soft gene trend clustering
  
library(tidyverse)
library(ggforce)
})
```

```{r}
theme <- theme_classic() + # light
         theme(plot.title = element_text(face = "bold", size = 11),
               strip.background = element_rect(fill = "white"),
               strip.text = element_text(face = "bold", colour = "black"),
               legend.position = "right", legend.justification = "left")
```


```{r, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

# Install

Make sure you installed all required libraries **after** launching the `.Rproj` file and initialization with packrat using `renv::init()`. Use `renv::snapshot()` to save changes (eg. if you insttalled a new library).  The chunk below only needs to be run the very first time (may take a while). Answer **no** to *'Do you want to install from sources the package which needs compilation? (Yes/no/cancel)'* and *'Update all/some/none? [a/s/n]'*.

```{r, eval=FALSE, include=FALSE}
## Install renv if needed
if (!requireNamespace("remotes"))
  install.packages("remotes")
remotes::install_github("rstudio/renv")

# Initialise 
renv::init(project = getwd(), bare = TRUE)
```

```{r, eval=FALSE, include=FALSE}
# Install required packages via Bioconductor
if (!requireNamespace('BiocManager'))
    install.packages('BiocManager')

install.packages(c('rmarkdown',
                   'kableExtra', 
                   'knitr',
                   'magrittr',
                   'reshape2',
                   'ggpubr',
                   'tidyverse',   
                   'circlize', 
                   'gsubfn', 
                   'lmerTest', 
                   'lme4',
                   'doParallel',
                   'statmod',
                   'viridis',
                   'RColorBrewer',
                   'longpower',
                   'factoextra',
                   'dendsort',
                   'doParallel',
                   'formatR',
                   'igraph',
                   'missForest',
                   'vegan',
                   'compositions', 
                   'ruv',
                   'GUniFrac', 
                   'zCompositions',
                   'pander',
                   'randomcoloR',
                   'lilikoi',
                   'alluvial'
                   ))

# dbnorm package dependencies
#install.packages(c("parallel", "plyr", "installr","fs","rmarkdown","processx","backports",
#"bootstrap","boot","caret","dplyr","stringr","ggfortify","RCurl","data.table","tidyr","scales", "e1071","fpc","rlang","glue","digest"))
#BiocManager::install(c("pcaMethods","impute","sva","BiocParallel","genefilter","Biobase","mixOmics","statTarget", "multtest"))
#require(usethis)
#require(devtools)
#devtools::install_version("ber", version = "4.0", repos = "http://cran.us.r-project.org")
#devtools::install_version("NormalizeMets", version = "0.25", repos = "http://cran.us.r-project.org")
#devtools::install_version("metabolomics", version = "0.1.4", repos = "http://cran.us.r-project.org")
#library(ber)
#library(NormalizeMets)
#library(Rcpp)
#library(metabolomics)

BiocManager::install(c('Maaslin2',
                       'ComplexHeatmap', 
                       'FELLA',
                       'KEGGREST',
                       'limma',
                       'biomaRt',
                       'ggbiplot',
                       'devtools',
                       'phyloseq',
                       'pmp',
                       'apeglm',
                       'vsn',
                       'DESeq2',
                       'ggpmisc',
                       'microbiome',
                       'phyloseq', 
                        'DECIPHER', 
                        'philr',
                        'metagenomeSeq',
                       'decontam',
                       'impute',
                       'lipidr',
                       'Mfuzz',
                       'Trendy',
                       'RUVSeq',
                       'DEP',
                       'GeneTonic',
                       'topGO',
                       'MOFA2',
                       'MultiAssayExperiment',
                       "hgu95av2.db"
                       ))

#
#install_github("zdk123/SpiecEasi")
#library(remotes)
#remotes::install_github("ricoderks/Rcpm")
```

```{r}
# Pro Batch dependencies
bioc_deps <- c("GO.db", "impute", "preprocessCore", "pvca","sva")
cran_deps <- c("corrplot", "data.table", "ggplot2", "ggfortify","lazyeval",
"lubridate", "pheatmap", "reshape2","readr", "rlang",
"tibble", "dplyr", "tidyr", "wesanderson","WGCNA")
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(bioc_deps)
install.packages(cran_deps)
```

```{r}
# Install required packages via CRAN
install.packages(c('tidyverse', 
                   'data.table', 
                   'gridExtra', 
                   'vegan', 
                   'randomcoloR', 
                   'compositions', 
                   'ruv', 
                   'ggpubr', 
                   'pander', 
                   'kableExtra', 
                   'rmarkdown', 
                   'knitr', 
                   'doParallel', 
                   'GUniFrac', 
                   'zCompositions', 
                   "dunn.test",
                   'viridis'))
```

```{r}
# NEtcomi
devtools::install_github("stefpeschel/NetCoMi", 
                         dependencies = c("Depends", "Imports", "LinkingTo"),
                         repos = c("https://cloud.r-project.org/",
                                   BiocManager::repositories()))

install.packages("BacDive", repos = "http://R-Forge.R-project.org")
#  bacdive <- open_bacdive("giulia.iacono@monash.edu", "Tellmeaboutmybacteria")
```


```{r}
# Molecular signature database immune signatures
msigdb_df <- msigdbr(species = "Homo sapiens", category = "C7") %>% dplyr::select(gs_name, entrez_gene)
head(msigdb_df)
#enricher(gene.df$ENTREZID, TERM2GENE = msigdb_df)
#em2 <- GSEA(geneList, TERM2GENE = msigdb_df)
```

