---
title: "library2"
author: "Giulia"
date: '2022-11-08'
output: html_document
---

# Scrap code

```{r}
lm_test <- function(SEexperiment, group_factors=c('factor_main','factor_random_tissue', 'factor_random_patient')) {
  
  data <- as.data.frame(cbind(factor_main=SEexperiment[[group_factors[1]]], factor_random_tissue=SEexperiment[[group_factors[2]]], factor_random_patient=SEexperiment[[group_factors[3]]]))
  
  coeff <- sapply(1:nrow(assays(SEexperiment)[['counts']]), function(x)summary(glht(lmer(assays(SEexperiment)[['counts']][x,]~factor_main+factor_random_tissue+(1|factor_random_patient), data=data)))$coef)
  
  pval <- sapply(1:nrow(assays(SEexperiment)[['counts']]), function(x)summary(glht(lmer(assays(SEexperiment)[['counts']][x,]~factor_main+factor_random_tissue+(1|factor_random_patient), data=data)))$test$pvalues)
  
  res_df <- data.frame(pval=pval[2,], adj_pval=p.adjust(pval[2,], method = 'fdr'), dm=coeff[2,])
  
  rownames(res_df) <- rownames(SEexperiment)
  res_df$AlignmentID <- rowData(SEexperiment)$info.Alignment.ID
  res_df$Ion <- rowData(SEexperiment)$info.Adduct.type
  
  return(res_df)}
```

```{r}
## Average path length
# Generate 1000 random graphs
gl <- vector("list", 1000)

for(i in 1:1000){
  # n = number of vertices, gpn = use density of the graph to generate random graphs
  gl[[i]] <- erdos.renyi.game(n = gorder(cross.mb.ig$mid), p.or.m = edge_density(cross.mb.ig$mid), type = "gnp")
}

# Calcumid average path length of 1000 random graphs
gl.apls <- unlist(lapply(gl, mean_distance, directed = FALSE))

hist(gl.apls, breaks = 20)
abline(
  v = mean_distance(cross.mb.ig$mid, directed = FALSE),
  col = "red",
  lty = 3,
  lwd = 2,)
```



```{r}
# Michaelis-Menten (MM) diversity establishment modelling
mm_data <- data.frame(sample_data(fungi.logcss))

# Run dose-response function
mm_model <- drm(Diversity ~ Days, data = mm_data, fct = MM.2())

# Check residuals
fit_res <- data.frame(fitted = fitted(mm_model), residuals = residuals(mm_model), condition = "CLAD-free")

ggplot(fit_res, aes(x = fitted, y = residuals)) +
    geom_hline(yintercept = 0, linetype = 2, color = 'grey70') +
    geom_point(aes(fill = condition), shape = 21) +
    facet_grid(cols = vars(condition), scales = 'free') +
    labs(title = 'Model residuals plot',
         x = 'Fitted (model)',
         y = 'Residuals (model)')

qqplot <- ggqqplot(residuals(mm_model)) +
    theme_grey() +
    labs(title = 'CLAD-free',
         subtitle = paste0('Shapiro-Wilk test: p = ', 
                           round(shapiro.test(residuals(mm_model))$p.value, 5)),
         x = 'Theoretical Quantiles',
         y = 'Sample Quantiles')


# Predict diversity based on age
simage <- seq(0, max(mm_data$Days), length.out = 1000)
mm <- data.frame(Days = simage)
mm <- cbind(mm, predict(mm_model, mm, interval = 'confidence', level = 0.95))
mm$condition <- 'CLAD-free'

# Calculate breakpoints
bp <- breakpoints(mm$Prediction ~ Days, data = mm, breaks = 1)
mm$breakpoint <- bp$breakpoints

name <- "fungi-predicted-shannon"
p <-
  
ggplot(mm, aes(x = Days, y = Prediction)) +
    geom_ribbon(aes(ymin = Lower, ymax = Upper), alpha = 0.2) +
    geom_line() +
    geom_vline(aes(xintercept = breakpoint), linetype = 2) +
    theme +
    labs(title = 'Post-transplant diversity establishment',
         #subtitle = 'Michaelis-Menten equation modelling',
         x = 'Days post-transplant',
         y = 'Predicted Shannon Diversity') +
    geom_text(x = mm$breakpoint[1] + 60, y = 0.1, size = 3,
            label = paste0('Breakpoint:\n', mm$breakpoint[1], ' days'))

ggsave(paste0(folder, name, ".png"), p, width = 10, height = 8, units = 'cm') #150
```

```{r}
# Michaelis-Menten (MM) diversity establishment modelling
mm_data <- data.frame(sample_data(groups.bact.logcss))
mm_data_clad <- filter(mm_data, Group == "CLAD") 
mm_data_cladfree <- filter(mm_data, Group == "CLAD-free") 

# Run dose-response function
mm_model_clad <- lm(Diversity ~ ns(Days, df = 3), data = mm_data_clad) #drm(Diversity ~ Days, data = mm_data_clad, fct = MM.2())
mm_model_cladfree <- lm(Diversity ~ ns(Days, df = 3), data = mm_data_cladfree)#drm(Diversity ~ Days, data = mm_data_cladfree, fct = MM.2())

# Check residuals
fit_res_clad <- data.frame(fitted = fitted(mm_model_clad), residuals = residuals(mm_model_clad), Group = "CLAD")
fit_res_cladfree <- data.frame(fitted = fitted(mm_model_cladfree), residuals = residuals(mm_model_cladfree), Group = "CLAD-free")
fit_res_combined <- rbind(fit_res_clad, fit_res_cladfree)

ggplot(fit_res_combined, aes(x = fitted, y = residuals)) +
    geom_hline(yintercept = 0, linetype = 2, color = 'grey70') +
    geom_point(aes(fill = Group), shape = 21) +
    #scale_fill_jama(name = 'Group') +
    facet_grid(cols = vars(Group), scales = 'free') +
    labs(title = 'Model residuals plot',
         x = 'Fitted (model)',
         y = 'Residuals (model)')

# Check normality of residuals
qqplot_clad <- ggqqplot(residuals(mm_model_clad)) +
    theme_grey() +
    labs(title = 'CF',
         subtitle = paste0('Shapiro-Wilk test: p = ', 
                           round(shapiro.test(residuals(mm_model_clad))$p.value, 5)),
         x = 'Theoretical Quantiles',
         y = 'Sample Quantiles')


qqplot_cladfree <- ggqqplot(residuals(mm_model_cladfree)) +
    theme_grey() +
    labs(title = 'Healthy',
         subtitle = paste0('Shapiro-Wilk test: p = ', 
                           round(shapiro.test(residuals(mm_model_cladfree))$p.value, 5)),
         x = 'Theoretical Quantiles',
         y = 'Sample Quantiles')

qq_plots <- ggarrange(qqplot_clad, qqplot_cladfree, nrow = 1)
qq_plots <- annotate_figure(qq_plots, top = text_grob(label = 'Normal Q-Q plots', size = 16))
  

# Predict diversity based on age
simage_clad <- seq(0, max(mm_data_clad$Days), length.out = 1000)
simage_cladfree <- seq(0, max(mm_data_cladfree$Days), length.out = 1000)

mm_clad <- data.frame(Days = simage_clad)
mm_clad <- cbind(mm_clad, predict(mm_model_clad, mm_clad, interval = 'confidence', level = 0.95))
mm_clad$Group <- 'CLAD'

mm_cladfree <- data.frame(Days = simage_cladfree)
mm_cladfree <- cbind(mm_cladfree, predict(mm_model_cladfree, mm_cladfree, interval = 'confidence', level = 0.95))
mm_cladfree$Group <- 'CLAD-free'

# Calculate breakpoints
bp_clad <- breakpoints(mm_clad$fit ~ Days, data = mm_clad, breaks = 2)
bp_cladfree <- breakpoints(mm_cladfree$fit ~ Days, data = mm_cladfree, breaks = 2)

#fstatclad <- Fstats(mm_clad$Prediction ~ Days, data = mm_clad)
#sctest(fstatclad, type = "supF")
#plot(fstatclad)

bp_data <- data.frame(Group = c('CLAD', 'CLAD-free'),
                      breakpoint = c(bp_clad$breakpoints, bp_cladfree$breakpoints))
mm_clad$breakpoint <- bp_clad$breakpoints
mm_cladfree$breakpoint <- bp_cladfree$breakpoints

# Combined CF and healthy
mm_combined <- rbind(mm_clad, mm_cladfree)

# Shannon diversity - days overlayed with predicted
name <- "sh-bact-time-with-predicted"
p <-
  
ggplot() +
  geom_point(aes(x = Days_after_max_last_day, y = Diversity, colour = Status), data = data.frame(sample_data(groups.bact.logcss))) +
  #geom_path(aes(x = Days, y = Diversity, colour = Group, group = Record.ID), data = data.frame(sample_data(groups.bact.logcss))) +
  theme +
  labs(x = "Days post-transplant", 
       y = "Shannon Index", 
       title = "Bacterial Diversity",
       subtitle = 'Overlayed with predicted Shannon diversity') +
  facet_wrap(~Group) +

  geom_ribbon(aes(x = Days, y = fit, ymin = lwr, ymax = upr), data = mm_combined %>% filter(Group == "CLAD-free"), alpha = 0.2) +
  geom_ribbon(aes(x = Days, y = fit, ymin = lwr, ymax = upr), data = mm_combined %>% filter(Group == "CLAD"), alpha = 0.2) +
  geom_line(aes(x = Days, y = fit, colour = Group), data = mm_combined) +
  geom_vline(aes(xintercept = breakpoint+3, colour = Group), data = mm_combined %>% filter(Group == "CLAD-free"), linetype = 2) +
  geom_vline(aes(xintercept = breakpoint, colour = Group), data = mm_combined %>% filter(Group == "CLAD"), linetype = 2) +
  plot_colours(Status = TRUE) 

ggsave(paste0("clad/", name, ".png"), p, width = 12, height = 8, units = 'cm')
# 150 567 CLAD
# 150 527
```

## Cell deconvolution with Bisque

```{r}
# Raw count matrix as input, logCPM normalisation performed by function
bisque.res <- BisqueRNA::MarkerBasedDecomposition(Biobase::ExpressionSet(assayData = gene.corrected$counts), markers)

sc.eset.metadata <- read.delim("/Users/giac0001/Desktop/R/Biomarker2/Multiomics/RNAseq/Data/bisque/GSE136831_AllCells.Samples.CellType.MetadataTable.txt") %>% 
  dplyr::filter(Disease_Identity == "Control")
sc.eset <- Matrix::readMM("GSE136831_RawCounts_Sparse.mtx")
```

## Tables templates

```{r}
for (i in c(1:length(datasets))){
  
  table <- list()

  ## Patients
  col <- unique(c(na.omit(datasets[[i]][, c("Record.ID")])))
  table[["Record.ID"]] <-  data.frame(cbind("Patients n" , paste0(length(col)), length(col))) %>% unname()

  
## Recipient Age
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, Age) %>%
    distinct() %>%
    pull(Age)
  table[["Recipient_age"]] <-  data.frame(cbind("Recipient age at Tx (range)" , paste0(round(mean(col), 1), " (", paste0(min(col), "-", max(col)), ")"), round(mean(col), 1))) %>% unname()
  
  
  ## Female patients
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, Gender) %>%
    distinct() %>%
    group_by(Gender) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    filter(Gender == "Female")
  table[["Gender"]] <-  data.frame(cbind("Female patients n (%)" , paste0(col$n, " (", col$Percentage, "%)"), col$n)) %>% unname()
  
  
  ## Donor Age
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, Donor_age) %>%
    distinct() %>%
    pull(Donor_age)
  table[["Donor_age"]] <-  data.frame(cbind("Donor age at Tx (range)" , paste0(round(mean(col), 1), " (", paste0(min(col), "-", max(col)), ")"), round(mean(col), 1))) %>% unname()
  
    ## Donor smoking
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, Donor_smoking) %>%
    distinct() %>%
    group_by(Donor_smoking) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    filter(Donor_smoking == "TRUE")
  table[["Smoking"]] <-  data.frame(cbind("Donor smoking n (%)" , paste0(col$n, " (", col$Percentage, "%)"), col$n)) %>% unname()
 
  ## Ischemic time
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, GIT_ischemic_time) %>%
    distinct() %>%
    na.omit() %>%
    pull(GIT_ischemic_time) 
  table[["GIT_ischemic_time"]] <-  data.frame(cbind("Ischemic time (min) (range)" , paste0(round(mean(col), 1), " (", paste0(min(col), "-", max(col)), ")"), round(mean(col), 1))) %>% unname() 
  

  
## Transplant type patients
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, TypeofTransplant) %>%
    distinct() %>%
    group_by(TypeofTransplant) %>%
    count() %>%
    rename(numb = n) %>%
    mutate(Percentage = round(numb/sum(.$numb)*100)) %>%
    rename(`Transplant type patients n (%)` = TypeofTransplant) %>%
    mutate(" " = paste0(numb, " (", Percentage, "%)")) 
  
  table[["TypeofTransplant"]] <-  rbind(gsub("numb", "", colnames(col)[c(1,4,2)]), col[, c(1,4,2)] %>% as.matrix() %>% unname())

  
## DSA at transplant patients
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, DSAatTransplant) %>%
    distinct() %>%
    group_by(DSAatTransplant) %>%
    count() %>%
    rename(numb = n) %>%
    na.omit() %>%
    mutate(Percentage = round(numb/sum(.$numb)*100)) %>%
    filter(DSAatTransplant == "DSAatTransplant") %>%
    mutate(DSAatTransplant = gsub("DSAatTransplant", "DSA at tx patients n (%)", DSAatTransplant)) %>%
    mutate("p" = paste0(numb, " (", Percentage, "%)")) %>%
    as.matrix() %>%
    unname()

  table[["DSAatTransplant"]] <-  data.frame(cbind(col[,1], col[,4], col[,2])) %>% unname()

  
## CMV mismatch
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, CMVMismatch) %>%
    distinct() %>%
    group_by(CMVMismatch) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    filter(CMVMismatch == "CMVMismatch") %>%
    mutate(CMVMismatch = gsub("CMVMismatch", "CMV mismatched patients n (%)", CMVMismatch)) %>%
    mutate("p" = paste0(n, " (", Percentage, "%)")) %>%
    as.matrix() %>%
    unname()

table[["CMVMismatch"]] <-  data.frame(cbind(col[,1], col[,4],col[,2])) %>% unname()


### Induction patients
  induction <- datasets[[i]] %>% 
  dplyr::select(Record.ID, Group, InductionTacrolimus, InductionAzathioprine, InductionBasilimax, InductionMycophenolate) %>%
    distinct()
  for (e in 3:dim(induction)[2]){induction[,e] <- ifelse(grepl("No", induction[,e]), 0, 1)}
 
induction <- induction %>%
  rowwise(Record.ID) %>%
  mutate(None = sum(InductionAzathioprine, InductionTacrolimus, InductionBasilimax, InductionMycophenolate)) 
colnames(induction) <- gsub("Induction", "", colnames(induction))

 col <- induction %>%
    pivot_longer(Tacrolimus:None, names_to = "Induction", values_to = "Count") %>%
    filter(Induction != "None" & Count != 0 | Induction == "None" & Count == 0) %>%
    group_by(Induction) %>%
    count() %>%
    rename(numb = n) %>%
    ungroup() %>%
    #arrange(desc(n)) %>%
    rename(`Induction patients n (%)` = Induction) %>%
    mutate(Percentage = round(numb/dim(induction)[1]*100), # corresponds to the total number of patients in that dataset
           " " = paste0(numb, " (", Percentage, "%)")) 
  table[["Induction"]] <-  rbind(gsub("numb", "", colnames(col)[c(1,4,2)]), col[, c(1,4,2)]) %>% as.matrix() %>% unname()

  
## ICUHours
  col <- datasets[[i]] %>%
    dplyr::select(Record.ID, ICUHours) %>%
    distinct() %>%
    pull(ICUHours)
  table[["ICUHours"]] <-  data.frame(cbind("ICU hours per patient (range)" , paste0(round(mean(col), 1), " (", paste0(min(col), "-", max(col)), ")"), round(mean(col), 1))) %>% unname()


## CLAD 
#  col <- datasets[[i]] %>%
#    dplyr::select(Record.ID, Group) %>%
#    distinct() %>%
#    group_by(Group) %>%
#    count() %>%
#    mutate(Percentage = round(n/sum(.$n)*100)) %>%
#    filter(Group == "CLAD") %>%
#    mutate(Group = gsub("CLAD", "CLAD at censor date patients n (%)", Group)) %>%
#    mutate("p" = paste0(n, " (", Percentage, "%)")) %>%
#    as.matrix() %>%
#    unname()
#  table[["Group"]] <-  data.frame(cbind("CLAD at censor date patients n (%)" , col[,4])) %>% unname()

  
## Samples
  col <- datasets[[i]][, c("Patient_days")] 
  table[["Samples"]] <-  data.frame(cbind("Samples n" , length(col),length(col))) %>% unname()
  
  
  ## Samples per reason
   col <- datasets[[i]] %>%
    group_by(BAL_class_combined) %>%
    count() %>%
    rename(numb = n) %>%
    mutate(Percentage = round(numb/sum(.$numb)*100)) %>%
    rename(`Bronch. reason samples n (%)` = BAL_class_combined) %>%
    mutate(" " = paste0(numb, " (", Percentage, "%)"))
  table[["Reason_samples"]] <-  rbind(gsub("numb", "", colnames(col)[c(1,4,2)]), col[, c(1,4,2)] %>% as.matrix() %>% unname())
  
  
  ## Samples per patient
   col <- datasets[[i]] %>%
    group_by(Record.ID) %>%
    count() 
  table[["Samples_patient"]] <-  data.frame(cbind("Samples per patient n (range)" , paste0(round(mean(col$n)), " (", paste0(min(col$n), "-", max(col$n)), ")"), round(mean(col$n)))) %>% unname()

  
## Average sampling in days
   col <- datasets[[i]] %>%
    pull(Days)
  table[["Samples_days"]] <-  data.frame(cbind("Average sampling days (range)" , paste0(round(mean(col), 1), " (", paste0(min(col), "-", max(col)), ")"), round(mean(col), 1))) %>% unname()
  
  
  
  
  ## Transplant indication
  ### patients
   col <- datasets[[i]] %>%
    dplyr::select(Record.ID, Transplant_indication) %>%
    distinct() %>%
    group_by(Transplant_indication) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    rename(`Tx indication samples n (%) | patients n (%)` = Transplant_indication) %>%
    mutate(`n (%)` = paste0(n, " (", Percentage, "%)"))
  table[["Transplant_indication_patients"]] <-  col[, c(1,4,2)] # correct

  
   ### samples
   col <- datasets[[i]] %>%
    group_by(Transplant_indication) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    rename(`Tx indication samples n (%) | patients n (%)` = Transplant_indication) %>%
    mutate(`n (%)` = paste0(n, " (", Percentage, "%)"))
  table[["Transplant_indication_samples"]] <-  col[, c(1,4,2)] # correct
  
  col <- table[["Transplant_indication_samples"]] %>% 
    full_join(table[["Transplant_indication_patients"]], by = "Tx indication samples n (%) | patients n (%)", suffix = c("_samp", "_pat")) %>%
    mutate(" " = paste0(`n (%)_samp`, " | ", `n (%)_pat`)) %>%
    dplyr::select(1,6,5)
  
  table[["Transplant_indication"]] <- rbind(gsub("n_pat", "", colnames(col), fixed  = TRUE), col 
                                                             %>% as.matrix() %>% unname()) # correct
  table[["Transplant_indication_samples"]] <- NULL
  table[["Transplant_indication_patients"]] <- NULL

  

## Immunosuppressants
  ### Samples
  immunos <- datasets[[i]] %>% 
  dplyr::select(Record.ID, Patient_days_lobe, Group, meds.list %>% filter(Type == "Immunosuppressant") %>% pull(Name)) %>%
    distinct()
  for (e in 4:dim(immunos)[2]){immunos[,e] <- ifelse(grepl("No", immunos[,e]), 0, 1)}
  
  col <- immunos %>%
    pivot_longer(Azathioprine:Tacrolimus, names_to = "Immunosuppressant", values_to = "Count") %>%
    filter(Count != 0) %>%
    group_by(Immunosuppressant) %>%
    count() %>%
    ungroup() %>%
    #arrange(desc(n)) %>%
    rename(`Immunosuppressant samples n (%) | patients n (%)` = Immunosuppressant) %>%
    mutate(Percentage = round(n/dim(immunos)[1]*100),
           `n (%)` = paste0(n, " (", Percentage, "%)")) 
  table[["Immunosuppressant_samples"]] <-  col[, c(1,4,2)] # correct
  
  ### Patients
  col <- immunos %>%
    pivot_longer(Azathioprine:Tacrolimus, names_to = "Immunosuppressant", values_to = "Count") %>%
    filter(Count != 0) %>%
    group_by(Immunosuppressant, Record.ID) %>%
    count() %>%
    group_by(Immunosuppressant) %>%
    count() %>%
    ungroup() %>%
    #arrange(desc(n)) %>%
    #add_row(Immunosuppressant = "Other", n = sum(.$n[.$n < 9])) %>%
    #filter(n >= 9) %>%
    rename(`Immunosuppressant samples n (%) | patients n (%)` = Immunosuppressant) %>%
    mutate(Percentage = round(n/length(unique(immunos$Record.ID))*100),
           `n (%)` = paste0(n, " (", Percentage, "%)")) 
  table[["Immunosuppressant_patients"]] <-  col[, c(1,4,2)] # correct
  
    col <- table[["Immunosuppressant_samples"]] %>% 
    full_join(table[["Immunosuppressant_patients"]], by = "Immunosuppressant samples n (%) | patients n (%)", suffix = c("_samp", "_pat")) %>%
    mutate(" " = paste0(`n (%)_samp`, " | ", `n (%)_pat`)) %>%
    dplyr::select(1,6,5)
    
  table[["Immunosuppressant_samples+patients"]] <- rbind(gsub("n_pat", "", colnames(col), fixed  = TRUE), col %>% as.matrix() %>% unname()) # correct
  table[["Immunosuppressant_samples"]] <- NULL
  table[["Immunosuppressant_patients"]] <- NULL

  
  
## Antibiotics
  ### Samples
  antibiotics <- datasets[[i]] %>% 
  dplyr::select(Record.ID, Patient_days_lobe, meds.list %>% filter(Type == "Antibiotic") %>% pull(Name)) %>%
    distinct()
  for (f in 3:dim(antibiotics)[2]){antibiotics[,f] <- ifelse(grepl("No", antibiotics[,f]), 0, 1)}
  
  table[["Antibiotics_samples"]] <- antibiotics %>%
    pivot_longer(Amoxycillin:Vancomycin, names_to = "Antibiotic", values_to = "Count") %>%
    filter(Count != 0) %>%
    group_by(Antibiotic) %>%
    count() %>%
    ungroup()
  
  if (i == 1){ # based on the total column to line up the rows, always 1
    ant.other <- table[["Antibiotics_samples"]] %>% filter(n <= 15) %>% pull(Antibiotic) 
    }

  ### Patients
  table[["Antibiotics_patients"]] <- antibiotics %>%
    pivot_longer(Amoxycillin:Vancomycin, names_to = "Antibiotic", values_to = "Count") %>%
    filter(Count != 0) %>%
    group_by(Antibiotic, Record.ID) %>%
    count() %>%
    group_by(Antibiotic) %>%
    count() %>%
    ungroup()
  
  col <- table[["Antibiotics_samples"]] %>% 
    full_join(table[["Antibiotics_patients"]], by = "Antibiotic", suffix = c("_samp", "_pat")) %>%
    add_row(`Antibiotic` = "Other", n_samp = sum(.$n_samp[.$Antibiotic %in% ant.other]), n_pat = sum(.$n_pat[.$Antibiotic %in% ant.other])) %>%
    filter(!Antibiotic %in% ant.other) %>%
    rename(`Antibiotics samples n (%) | patients n (%)` = Antibiotic) %>%
    mutate(Percentage_samp = round(n_samp/dim(antibiotics)[1]*100),
           Percentage_pat = round(n_pat/length(unique(antibiotics$Record.ID))*100),
           " " = paste0(`n_samp`, " (", Percentage_samp, "%)", " | ", `n_pat`, " (", Percentage_pat, "%)")) %>% 
    dplyr::select(1,6,3) # 3 correct n_pat
  
    table[["Antibiotics_samples_patients"]] <- rbind(gsub("n_pat", "", colnames(col), fixed  = TRUE), col %>% as.matrix() %>% unname()) # correct
    table[["Antibiotics_samples"]] <- NULL
    table[["Antibiotics_patients"]] <- NULL
  
  
  ## Positive microbiology
  ### patients
   col <- datasets[[i]] %>%
    dplyr::select(Record.ID, BAL_micro_results) %>%
    distinct() %>%
    group_by(BAL_micro_results) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    rename(`BAL microbiology samples n (%) | patients n (%)` = BAL_micro_results) %>%
    mutate(`n (%)` = paste0(n, " (", Percentage, "%)"))
  table[["BAL_micro_results_patients"]] <-  col[, c(1,4,2)] # correct

   ### samples
   col <- datasets[[i]] %>%
    group_by(BAL_micro_results) %>%
    count() %>%
    mutate(Percentage = round(n/sum(.$n)*100)) %>%
    rename(`BAL microbiology samples n (%) | patients n (%)` = BAL_micro_results) %>%
    mutate(`n (%)` = paste0(n, " (", Percentage, "%)"))
  table[["BAL_micro_results_samples"]] <-  col[, c(1,4,2)] # correct
  
  col <- table[["BAL_micro_results_samples"]] %>% 
    full_join(table[["BAL_micro_results_patients"]], by = "BAL microbiology samples n (%) | patients n (%)", suffix = c("_samp", "_pat")) %>%
    mutate(" " = paste0(`n (%)_samp`, " | ", `n (%)_pat`)) %>%
    dplyr::select(1,6,5) %>%
    na.omit()
  
  table[["BAL_micro_results"]] <- rbind(gsub("n_pat", "", colnames(col), fixed  = TRUE), col 
                                                             %>% as.matrix() %>% unname()) # correct
  table[["BAL_micro_results_samples"]] <- NULL
  table[["BAL_micro_results_patients"]] <- NULL
  
tab[[i]] <- plyr::ldply(table, data.frame)    

}

datasets.full.table <- tab[[1]] %>% dplyr::select("ID" = `.id`, "demo" = X1, "Total" = X2) %>%
  full_join(tab[[2]] %>% dplyr::select("ID" = `.id`, "demo" = X1, "16S Dataset" = X2)) %>%
  full_join(tab[[3]] %>% dplyr::select("ID" = `.id`, "demo" = X1, "RNA Dataset" = X2)) %>%
  full_join(tab[[4]] %>% dplyr::select("ID" = `.id`, "demo" = X1, "Metabolomics Dataset" = X2)) %>%
  full_join(tab[[5]] %>% dplyr::select("ID" = `.id`, "demo" = X1, "Lipidomics Dataset" = X2))
write.csv(datasets.full.table, "clinical-checks/3-1-datasets.csv", row.names = FALSE)
```

#### Heatmap pathways

```{r}
## Heatmap
# Select pathways
s <- list()

for (i in unique(deseq_list$`group*months`$GroupCLAD$go_summary$Gene)){ #rna.list$group_before_decline_new$go_summary
  s[[i]] <- deseq_list$`group*months`$GroupCLAD$go_summary %>% 
  filter(Gene == i) %>%
  slice_min(qvalue)}

heat.path <- bind_rows(s) %>%
  dplyr::select(Gene, Description) %>% 
  right_join(deseq_list$`group*months`$GroupCLAD$top_results_df, by = "Gene") %>%
  mutate(Pathway = ifelse(is.na(Description), "Other", Description)) %>%
  dplyr::select(Gene, Pathway)

data <- deseq_list$`group*days-all`$GroupCLAD.ns.Days..df...2.2$heat_sig#[d, ]
metadata <- deseq_list$`group*days-all`$metadata

name <- "rna-heat-top"
p <- 
  
      draw(Heatmap(deseq_list$`group*days-all`$GroupCLAD.free.ns.Days..df...2.2$heat_sig, # %>% filter(Group == "CLAD")
        name = "z-score",
        column_title = "",
        row_title = 'Genes',  
        show_column_names = FALSE,
        show_row_names = TRUE,
        row_names_gp = gpar(fontsize = 8),
        column_order = colnames(deseq_list$`group*days-all`$GroupCLAD.free.ns.Days..df...2.2$heat_sig[, metadata %>% arrange(Group, Days) %>% pull(Patient_days)]),
        top_annotation = annotate_heatmap(metadata, c("Group","Days", "BAL_class_combined")),
        #right_annotation = annotate_heatmap_row(metadata = heat.path, 
        #                                        row_id = Gene, 
        #                                        row_or = rownames(deseq_list$`group*ns(days)`$Group_CLAD_vs_CLAD.free$heat_top), 
        #                                        annotations = c("Pathway"), 
        #                                        palette = c("#00AFB5",  "#184E77","#FF6700", "#7AE7C7", "#D81159", "white")),
        #heatmap_legend_param = list(direction = "horizontal"),
        #col = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")), space = 'rgb')(50)
        #col = colorRamp2(c(-4, 0, 4), rev(brewer.pal(n = 7, name = "RdYlBu")))   
        ), 
        padding = unit(c(0, 0, 0, 0), "cm"),
     merge_legend = TRUE,
     heatmap_legend_side = "right")

p <- recordPlot()

ggsave(paste0("predictive-clad/deseq/", test_variable, "/", "GroupCLAD.ns.Months..df...2.2", "/heat-top.pdf"), p, width = 30, height = 15, units = 'cm')


pdf(paste0("predictive-clad/deseq/", test_variable, "/", "GroupCLAD.ns.Months..df...2.2", "/heat.pdf"), width = 20000, height = 6000, res = 800, bg = "transparent")
p
invisible(dev.off()) 

view(master.metadata$micro_metadata)
```

### FC Plot

Log2 fold changeover time for individuals with CLAD on the x axis and individuals without CLAD on the y axis with each point representing a gene. Genes showing an interaction between time and drug are coloured blue or gold depending on whether their fold change is greater post-rituximab (blue) or post-tocilizumab (gold). Genes without interaction, but changing significantly over time are coloured green and tend to lie along the line of identity. See the Longitudinal tab in https://r4ra.hpc.qmul.ac.uk for an interactive version of the above plot.

```{r}
fc <- deseq_list[[test_variable]][["res"]]$GroupCLAD.ns.Days. %>% as.data.frame() %>% mutate(Group = case_when(log2FoldChange > 0 ~ "CLAD",
                                                                                                                    log2FoldChange < 0 ~ "CLAD-free"))

fc %>% filter(Group == "CLAD") %>% dplyr::select(log2FoldChange, padj) %>% rownames_to_column("Gene") %>% rename(padj_interaction = padj) %>%
    full_join(fc %>% filter(Group == "CLAD-free") %>% dplyr::select(log2FoldChange) %>% rownames_to_column("Gene"), by = "Gene", suffix = c("CLAD_", "CLAD-free_"))


ggplot(fc, aes(x = fc %>% filter(Group == "CLAD") %>% pull(log2FoldChange),
               y = fc %>% filter(Group == "CLAD-free") %>% pull(log2FoldChange))) +
  geom_point

```

```{r}
time_hits <- deseq_list[[test_variable]][["sig.results.df"]]$ns.Days. %>% as.data.frame() %>% mutate(Time = case_when(log2FoldChange > 0 ~ "Days_up",
                                                                                                                    log2FoldChange < 0 ~ "Days_down")) %>%
  rownames_to_column("Gene.Name") %>%
  left_join(deseq_list[[test_variable]][["gene.df"]]$ns.Days.) %>%
  dplyr::select(ENTREZID, Time)

group_hits <- deseq_list[[test_variable]][["sig.results.df"]]$Group_CLAD_vs_CLAD.free %>% as.data.frame() %>% mutate(Group = case_when(log2FoldChange > 0 ~ "CLAD",
                                                                                                                    log2FoldChange < 0 ~ "CLAD-free")) %>%
  rownames_to_column("Gene.Name") %>%
  left_join(deseq_list[[test_variable]][["gene.df"]]$Group_CLAD_vs_CLAD.free) %>%
  dplyr::select(ENTREZID, Group)

hits <- time_hits %>% full_join(group_hits)

formula_res <- compareCluster(ENTREZID ~ Time+Group, data=hits, fun="enricher",
TERM2GENE=wp[,c("wpid", "gene")], TERM2NAME=wp[,c("wpid", "name")])


```

### GSEA

```{r}

#goplot(deseq_list[[test_variable]][[name]][["ego_simp"]] )

d <- deseq_list[[test_variable]][[name]]$sig_results_df[, c("Gene", "log2FoldChange")] %>%
  left_join(deseq_list[[test_variable]][[name]][["gene_df"]] %>% rename(Gene = Gene.Name))
genelist <- d[,2] # logfoldchange
names(genelist) <- d[,4] # entrezid
genelist <- sort(genelist, decreasing = TRUE)

gsego <- gseGO(geneList    = genelist,
              OrgDb        = org.Hs.eg.db,
              ont          = "BP",
              minGSSize    = 3,
              maxGSSize    = 20,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

#xx <- compareCluster(Gene âˆ¼ Group*Months, data=DE_GSE8057, fun = enricher,
#TERM2GENE=wp[,c("wpid", "gene")], TERM2NAME=wp[,c("wpid", "name")])
gsego@result

```

# Maaslin process

```{r}
extract_maaslin <- function(fit_data, value_interest, bind_id, interest_vars, input_data, input_metadata, multiple_features = FALSE, number = 2, max_significance = 0.05){ 
  
  #fit_data <- bact_maaslin[[id]]
  #value_interest <- "CLAD"
 # bind_id <- "Patient_days"
  #interest_vars <- c("Days", "Days_post_transplant", "Group") 
  #input_data <- input_data 
  #input_metadata <- input_metadata
  
  maaslin.out <- data.frame(fit_data$results) 
  
  if (isTRUE(multiple_features)){
      var_all <- maaslin.out %>% filter(qval < max_significance) %>% group_by(feature) %>% count() %>% filter(n > number) %>% pull(feature)
      maaslin.out <- maaslin.out %>% dplyr::select(feature) %>% filter(feature %in% var_all) %>% distinct()}
  
  if (isFALSE(multiple_features)){
      maaslin.out <- maaslin.out %>% filter(qval < max_significance & value %in% value_interest)}
    
  maaslin.out$feature <- gsub("[[:punct:]]", "", maaslin.out$feature)

  # Extract IDs from otu table
    mas <- data.frame(t(input_data), check.names = FALSE) %>% 
    rownames_to_column("feature") %>%
    mutate(Feature = feature) %>%
    relocate(Feature, .after = feature)
    mas$feature <- gsub("[[:punct:]]| ", "", mas$feature)
  
  # Merge features
    maaslin.out <- mas[,c("Feature", "feature")] %>% inner_join(maaslin.out, by = "feature")
    
    fit_data[["maaslin_out"]] <- maaslin.out
  
  # Counts over variable of interest
  count_table <- mas %>%
    inner_join(maaslin.out, by = c("Feature", "feature"))
  
  count_table <- count_table[, !colnames(count_table) %in% c("metadata", "value", "coef", "stderr", "name", "N", "N.not.zero", "pval", "qval")] %>% #metadata on purpose
    column_to_rownames("Feature") %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(bind_id) %>%
    inner_join(input_metadata[, c(bind_id, interest_vars)], by = bind_id)
  
  # Wide to long format
  var_counts <- reshape2::melt(count_table, id.vars = c(bind_id, interest_vars), variable.name = "ASV", value.name = "Abundance")
  var_counts$Abundance <- as.numeric(var_counts$Abundance)
  
  fit_data[["var_counts"]] <- var_counts
  
  return(fit_data)} 
```

```{r}
plot_time_counts <- function(de_list, metadata, bind_id,
                              top_hits = FALSE, sig_hits = FALSE, 
                              interest_vars, selected_fill = selected_fill, folder_id, 
                              correlation = NULL, correlation_id = NULL){
  
  #deseq2_list <- res.rna.clad
 # metadata <- rna.metadata
 # top_genes <- TRUE
  
  if (isTRUE(top_hits)){
    data <- as.matrix(de_list$heat_top_hits) }
  
  if (isTRUE(sig_hits)){
    data <- as.matrix(de_list$heat_sig_hits)}

plot.counts <- data %>% 
  t() %>%
  data.frame(., check.names = FALSE) %>%
  rownames_to_column(., var = "Patient_days") %>%
  inner_join(metadata[, interest_vars], by = "Patient_days") %>%
  reshape2::melt(., is.vars = interest_vars, measure.vars = rownames(data)) %>%
  arrange(Record.ID, Days) 

#var_counts <- melt(count_table, id.vars = interest_vars, variable.name = "ASV", value.name = "Abundance")

  dir.create(paste0("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id), "-", correlation_id))
  
  # Correlation
  if (!is.null(correlation_id)){
    
  for (a in 1:length(rownames(data))){
    
  var_select <- rownames(data)[a]
  
  g <- ggplot(plot.counts %>% filter(variable == var_select), aes(x = {{correlation}}, y = value)) + 
    geom_smooth(aes(colour = {{selected_fill}}, group = {{selected_fill}}), se = F, lty = "dashed", size = 0.5, method = lm) + #
    #stat_cor(method = "pearson") +
    geom_point(aes(colour = {{selected_fill}})) + #
    theme(plot.title = element_text(size = 8)) +
    labs(title = paste("Correlation between", var_select, "expression and", correlation_id)) +
    ylab(label = "Normalised log counts") +
    xlab(label = correlation_id) #+
    #scale_x_reverse() #+
    #scale_colour_manual(values = c("none" = "#240046", "clad" = "#FF6D00"))
    #scale_color_gradientn(colors = c("#F9E855", "#478E8C", "#3F1151"))

ggsave(paste("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id), "-", correlation_id, "/", gsub("/", "", var_select), ".png", sep = ""), g, width = 12, height = 8, units = 'cm')}}
  
  if (isTRUE(is.null(correlation_id))){
    # If clad clinical was selected 
  if (isTRUE(as.character(folder_id) == "clad_clinical")){
      
   for (a in 1:length(rownames(data))){
    
  var_select <- rownames(data)[a]
  
  g <- ggplot(plot.counts %>% filter(variable == var_select), aes(x = Days, y = value)) + 
  geom_smooth(aes(colour = {{selected_fill}}, group = {{selected_fill}}), se = F, lty = "dashed", size = 0.5, method = lm) +
  #geom_path(aes(group = Record.ID, colour = category_clinical)) +
  geom_point(aes(colour = {{selected_fill}}), size = 1.5, stroke = 0.3) +
  scale_colour_manual(values = c("none" = "#240046", "clad" = "#FF6D00")) + 
  labs(title = paste(var_select, "expression")) +
  ylab(label = "Normalised log counts") +
  xlab(label = "Time (days)") 

ggsave(paste("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id),"-", correlation_id, "/", gsub("/", "", var_select), ".png", sep = ""), g, width = 10, height = 7, units = 'cm')}}

  # If category was selected
  if (isTRUE(as.character(folder_id) == "clad_within")){
      
   for (a in 1:length(rownames(data))){
    
  var_select <- rownames(data)[a]
  
  g <- ggplot(plot.counts %>% filter(variable == var_select), aes(x = Days, y = value)) + 
  geom_smooth(aes(colour = {{selected_fill}}, group = {{selected_fill}}), se = F, lty = "dashed", size = 0.5, method = lm) +
  #geom_path(aes(group = Record.ID, colour = category_clinical)) +
  geom_point(aes(colour = {{selected_fill}}), size = 1.5, stroke = 0.3) +
  scale_colour_manual(values = c("Early_1.5" = "#7FCEA5", "Intermediate_1.5_to_2" = "#00BFB2", "Late_2_to_3" = "#126782", "None" = "#023047"
    #"none" = "#240046", "clad" = "#DA4167", "blad+clad" = "#FF6D00", "blad" = "#119DA4"
    )) +
  labs(title = paste(var_select, "expression")) +
  ylab(label = "Normalised log counts") +
  xlab(label = "Time (days)") 

ggsave(paste("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id),"-", correlation_id, "/", gsub("/", "", var_select), ".png", sep = ""), g, width = 10, height = 7, units = 'cm')}}
  
  # If recordID was selected
  if (isTRUE(as.character(folder_id) == "Record.ID")){
      
   for (a in 1:length(rownames(data))){
    
  var_select <- rownames(data)[a]
  
  g <- ggplot(plot.counts %>% filter(variable == var_select), aes(x = Days, y = value)) + 
  geom_path(aes(group = {{selected_fill}}, colour = {{selected_fill}})) +
  geom_point(aes(colour = {{selected_fill}}), size = 1.5, stroke = 0.3) +
  scale_colour_manual(values = c("11" = "#046C9A", "13" = "#ECCBAE", "19" = "#692895", "2" = "#08b6b6", "21" = "#ff8a5b", "22" = "#faa613", "23" = "#A6FFA1", "25" = "#EF233C",  "32" = "#A31621",  "6" = "#2d7dd2")) +
  labs(title = paste(var_select, "expression")) +
  ylab(label = "Normalised log counts") +
  xlab(label = "Time (days)") 

ggsave(paste("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id),"-", correlation_id, "/", gsub("/", "", var_select), ".png", sep = ""), g, width = 10, height = 7, units = 'cm')}}
    
 #else for (a in 1:length(rownames(data))){
    
  #var_select <- rownames(data)[a]
  
  #g <- ggplot(plot.counts %>% filter(variable == var_select), aes(x = Days, y = value)) + 
  #geom_smooth(aes(colour = {{selected_fill}}, group = {{selected_fill}}), se = F, lty = "dashed", size = 0.5, method = lm) +
  #geom_path(aes(group = Record.ID, colour = category_clinical)) +
  #geom_point(aes(colour = {{selected_fill}}), size = 1.5, stroke = 0.3) +
  #scale_colour_manual(values = c("none" = "#240046", "clad" = "#DA4167", "blad+clad" = "#FF6D00", "blad" = "#119DA4")) +
  #labs(title = paste(var_select, "expression")) +
  #ylab(label = "Normalised log counts") +
  #xlab(label = "Time (days)") 

#ggsave(paste("Figures/", Sys.Date(), "-", de_list$Variable, "-", as.character(folder_id),"-", correlation_id, "/", gsub("/", "", var_select), ".png", sep = ""), g, width = 10, height = 7, units = 'cm')}
    }}
```






