---
title: "set-up"
author: "Giulia"
date: "27/09/2021"
output:
---

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c("ComplexHeatmap", "S4Vectors", "sva", "SummarizedExperiment", "FELLA", "phyloseq", "metagenomeSeq", "microbiome", "decontam", "MultiAssayExperiment", "MOFA2", "limma", "biomaRt", "edgeR", "org.Hs.eg.db", "GeneTonic", "clusterProfiler", "DESeq2", "TMixClust"))
devtools::install_github("microsud/dysbiosisR")
```


# Libraries

```{r, echo=FALSE, results='hide'}
suppressPackageStartupMessages({ 

# Data wrangling and visualisation
library(tidyverse)
library(ggforce)
library(ggplotify) #turn plot() output into ggplot
library(ggsci)
library(knitr)
library(magrittr)
library(ggpubr)
library(gsubfn)
#library(viridis)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggalluvial)
library(dendsort)
library(tinytex)
library(ggVennDiagram)
#library(reshape2)
library(ggrepel)
library(gplots)
library(gridExtra)
library(janitor)
library(circlize) # chorddiagram
library(gt) # tables
library(ggdendro) # dendrograms
library(strex) # stringr manipulation
library(scales) # rescale from 0 to 1
library(readxl)

## Networks
library(igraph)
#library(intergraph)
#library(GGally) # networks
#library(sna)

# Analysis and statistics  
library(statmod)
library(lme4)
library(lmerTest)
library(longpower)
library(cluster)
library(factoextra)
library(fpc)
library(rstatix)
library(broom)
library(S4Vectors)
library(data.table)
library(devtools)
#library(strucchange) #time split
library(sjmisc)
library(ggeffects) # curve modelling

## ms pre-processing
#library(pmp)
#library(MSPrep)
#library(impute)
#library(missForest)
library(sva) # batch correction
library(SummarizedExperiment)
#library(KEGGREST)
library(FELLA)
#library(vsn)

## microbiome
#library(doParallel)
library(phyloseq)
library(metagenomeSeq)
library(MASS)
library(vegan)
library(microbiome)
library(decontam)
#library(syncomR) # Community stability
#library(dysbiosisR)  

# Mofa
library(MultiAssayExperiment)
library(psych)


  ## rnaseq
library(limma)
library(biomaRt)
#library(variancePartition)
library(formatR)
library(splines)
library(edgeR)
library(org.Hs.eg.db) # biomart
library(hgu133plus2.db) # Affimetrix
library(GeneTonic)
library(msigdbr) # molecular signature database (immune)
#library(caret) # ML
#library(MLeval) # ML
#library(caTools) # ML
#library(MLmetrics) # ML
#library(nnet) # ML
library(clusterProfiler) # pathway analysis needs tidytree installed
library(DESeq2)
library(ashr) # DE shrinkage for DESeq2
#library(topconfects) # select confident DE genes
#library(qusage) # pathway analysis
library(mgcv) # GAM
#library(tradeSeq) # GAM
#library(SingleCellExperiment) # tradeSeq
#library(clusterExperiment) # tradeseq
library(TMixClust)
  
library(SeuratObject)
library(Seurat)
  
#library(survival) # COX factor modelling
#library(survminer) # COX factor modelling visualisation
library(corrplot) # correlation plot
library(reticulate) # load because masked by phyloseq
library(MOFA2) # needs to be last to avoid collisions
  
})
```

# Theme

```{r}
theme <- theme_classic() + 
         theme(plot.title = element_text(face = "bold", size = 11),
               strip.background = element_rect(fill = "white", linewidth = 0.5),
               strip.text = element_text(face = "bold", colour = "black"),
               legend.position = "right", legend.justification = "left")

phylum_colours <- c("#f26a8d", "#3F3D89", "#8339B8", "#CC6AD8", "#EA9010", "#1B8293", "#80ed99", "#38B2FB", "#5C6378")
names(phylum_colours) <- c("Proteobacteria", "Bacteroidota", "Firmicutes", "Actinobacteriota", "Fusobacteriota", "Campylobacterota", "Patescibacteria", "Ascomycota", "Basidiomycota")

clad_phylum_colours <- c( "#ffbe0b", "#78aafa", "darkgrey", "#003995", "#f26a8d", "#80ed99", "#005F16")
names(clad_phylum_colours) <- c("Proteobacteria", "Bacteroidota", "Firmicutes", "Actinobacteriota", "Fusobacteriota", "Campylobacterota", "Patescibacteria")

yellowblue <- brewer.pal(n = 6, name = "YlGnBu")

path_palette <- c("Carboxylic acids and derivatives" = "#1c5ed9",
                            "Organooxygen compounds" = "darkgrey",
                            "Sphingolipids" = "#bd81f5",
                            "Glycerophospholipids" = "#5aeece",
                            "Fatty acyls" = "#8678f0",
                            "Glycerolipids" = "#a456a0",
                            "Purine, pyridines and derivatives" = "#38B2FB",
                            "Sterol lipids and phenol ethers" = "#aecf8e",
                            "Benzene and substituted derivatives" = "#aecf8e",
                            "Organic sulfuric, hydroxy acids and derivatives" = "#23a16d",
                            "Other" = "#EAEAEA")
```

