---
title: "Stable cohort - Molecules associations post-transplant"
author: "Giulia Iacono"
date: Sys.Date()
output: html_document
---

# Stable cohort - Molecules associations post-transplant
# Molecules

```{r}
stable_small_molecules_metadata <- small_molecules_metadata %>% filter(Record.ID %in% stable_patients)
dim(stable_small_molecules_metadata) #116 163
stable_small_molecules <- small_molecules[, c(stable_small_molecules_metadata$Patient_days)]
dim(stable_small_molecules) #985 116

# stable_small_molecules_metadata %>% group_by(Record.ID) %>% dplyr::count() #18
```

# PCA

```{r}
a <- data.frame(adonis2(formula = as.formula("t(stable_small_molecules) ~Days_interval+Transplant_indication+Total_ant_num+GIT_ischemic_time+Record.ID"), 
                    data = stable_small_molecules_metadata, 
                    distance = "eu", 
                    dfun = vegdist, 
                    permutations = 999,
                    na.action = na.omit,
                    parallel = 6,
                    by = "terms"))

annotation <- paste(paste("SumOfSqs =", round(a["Days_interval",]$SumOfSqs, 3)),
              paste("R2 =", round(a["Days_interval",]$R2, 3)),
              paste("p =", round(a["Days_interval",]$Pr..F., 3)), sep = " ; " )

pca.list <- pca_process(stable_small_molecules, stable_small_molecules_metadata)

p <-
  
  ggplot(pca.list$data.PCA, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(fill = Days_interval, colour = Days_interval), linewidth = 0.2, geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
  geom_point(aes(fill = Days_interval), shape = 21, size = 2) +
  theme +
  labs(title = 'Molecules ordination',
       caption = annotation,
       fill = "Time post-transplant (months)",
       colour = "Time post-transplant (months)",
       x = paste0('PC1 ', round(pca.list$varexp[1], 2), '%'),
       y = paste0('PC2 ', round(pca.list$varexp[2], 2), '%')) +
  plot_fill(Days_interval = TRUE) +
  plot_colours(Days_interval = TRUE)

ggsave(paste0(analysis_path, "transcriptomics-small-molecules/pca/molecules-pcoa-months-grouped.png"), p, width = 13, height = 7, units = 'cm', dpi = 600)
```


# *Differential abundance over months

```{r}
# molecules_list <- list()
design <- model.matrix(~ns(Months, df = 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time, stable_small_molecules_metadata) # same for tx indication
dupcor <- duplicateCorrelation(stable_small_molecules, design = design, block = stable_small_molecules_metadata$Record.ID)
vfit <- lmFit(stable_small_molecules, design, block = stable_small_molecules_metadata$Record.ID, correlation = dupcor$consensus)
efit <- eBayes(vfit)
summary(decideTests(efit, lfc = 0)) 
#topTreat(efit, coef = "ns(Months, df = 3)2", number = 400)

test_variable <- "months"
folder_id <- paste0('limma/', test_variable, '/molecules/')
dir.create(folder_id)

molecules_list <- limma_molecules_process(list = molecules_list, 
                                    data = stable_small_molecules, 
                                    metadata = stable_small_molecules_metadata,
                                    coef_name = c("ns(Months, df = 3)1", "ns(Months, df = 3)2", "ns(Months, df = 3)3"), 
                                    test_var1 = factor(c("Decreased", "Increased")), 
                                    title = test_variable, 
                                    efit = efit, 
                                    folder_id = folder_id, 
                                    logFC_threshold = 0)

stable_molecules_combined_df <- rbind(molecules_list$months$`ns(Months, df = 3)1`$sig_hits, 
                                      molecules_list$months$`ns(Months, df = 3)2`$sig_hits,
                                      molecules_list$months$`ns(Months, df = 3)3`$sig_hits) %>%
  group_by(Molecule) %>%
  slice_min(adj.P.Val, n = 1)
stable_molecules_combined <- stable_molecules_combined_df$Molecule
length(stable_molecules_combined) # 97

molecules_list <- molecules_pathway_analysis(list = molecules_list, 
                                             data_info = small_molecules_info, 
                                             title = test_variable, 
                                             molecules = stable_molecules_combined, 
                                             folder_id = folder_id, fella = fella.hsa)

saveRDS(molecules_list, paste0(analysis_path, "transcriptomics-small-molecules/limma/molecules_list.rds"))

# molecules_list$months$fella_output
```


```{r}
# Find out how DE molecules are named in gem database using database IDs
a <- gem_molecules %>% 
  dplyr::select(NAME, "KEGG" = metKEGGID, "HMDB_Accession" = metHMDBID, "LIPIDMAPS_ID" = metLipidMapsID) %>% 
  distinct() %>%
  mutate(KEGG = ifelse(KEGG == "", NA, KEGG),
         HMDB_Accession = ifelse(HMDB_Accession == "", NA, HMDB_Accession),
         LIPIDMAPS_ID = ifelse(LIPIDMAPS_ID == "", NA, LIPIDMAPS_ID) ) %>%
  filter(!c(is.na(KEGG) & is.na(HMDB_Accession) & is.na(LIPIDMAPS_ID)))

stable_molecules_gem_names <- molecules_list$months$molecules_info %>%
  mutate(HMDB_Accession = gsub(";.*", "", HMDB_Accession),
         KEGG = gsub(";.*", "", KEGG),
         LIPIDMAPS_ID = gsub(";.*", "", LIPIDMAPS_ID), 
         KEGG = ifelse(KEGG == "", NA, KEGG),
         HMDB_Accession = ifelse(HMDB_Accession == "", NA, HMDB_Accession),
         LIPIDMAPS_ID = ifelse(LIPIDMAPS_ID == "", NA, LIPIDMAPS_ID)) %>%
  left_join(a %>% dplyr::select(NAME, HMDB_Accession) %>% na.omit() %>% distinct()) %>%
  left_join(a %>% dplyr::select(NAME, KEGG) %>% na.omit() %>% distinct()) %>%
  left_join(a %>% dplyr::select(NAME, LIPIDMAPS_ID) %>% na.omit() %>% distinct())
```


### TMixClust

```{r}
molecules_combined_heat <- as.data.frame(molecules_list$months$`ns(Months, df = 3)2`$heat_all[stable_molecules_combined, ] ) 

data <- 
  molecules_combined_heat %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Patient_days") %>%
  left_join(stable_small_molecules_metadata[, c("Patient_days", "Months_grouped")]) %>%
  dplyr::select(!c("Patient_days")) %>%
  group_by(Months_grouped) %>%
  summarise(across(where(is.numeric), mean)) %>%
  column_to_rownames("Months_grouped") %>%
  t() %>%
  as.data.frame()

cluster_obj <- list()
for (i in 2:3){
  n_clust <- i
  cluster_obj[[n_clust]] <- TMixClust(data, nb_clusters = n_clust)
}
saveRDS(cluster_obj, paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/cluster_obj_interval.rds"))
# cluster_obj <- readRDS("tmixclust/molecules/cluster_obj_interval.rds")

plot_silhouette(cluster_obj[[2]]) # 2 0.57 | 3 0.53

# plot the time series in each cluster
for (i in 1:2) {
plot_time_series_df(data[which(cluster_obj[[2]]$em_cluster_assignment == i), ], plot_title = paste("cluster",i))
}

cluster_ms <- list()
for (i in 1:2) {
cluster_ms[[i]] <- rownames(data[which(cluster_obj[[2]]$em_cluster_assignment == i), ]) 
}

length(unique(unlist(cluster_ms))) # 97
length(cluster_ms[[1]]) # 63 down
length(cluster_ms[[2]]) # 34 up

cluster_ms_list <- list()
for (i in 1:2) {
cluster_ms_list[[i]] <- data[which(cluster_obj[[2]]$em_cluster_assignment == i), ] %>% mutate("Cluster" = i)
}
```

#### Discrete

```{r}
# Time cluster plots
cluster_ms_df <- bind_rows(cluster_ms_list) %>%
  rownames_to_column("Molecule") %>%
  pivot_longer(-c("Molecule", "Cluster"), names_to = "Months_grouped", values_to = "Expression") %>%
  mutate(Cluster_description = case_when(Cluster == 2 ~ "Glycerophospholipid metabolism", 
                                         Cluster == 1 ~ "Amino and carboxylic acid metabolism"),
         Cluster_description_de = case_when(Cluster == 2 ~ "Glycerophospholipids\nDA molecules: 34", 
                                            Cluster == 1 ~ "Amino and carboxylic acids\nDA molecules: 63"),
         Months_grouped = factor(Months_grouped, c("<1.5", "1.5-3", "4-6", "7-9", "10-12", "13-18", ">18")))

p <- 
  
  ggplot(cluster_ms_df, aes(x = Months_grouped, y = Expression)) +
  geom_smooth(aes(group = Molecule), method = "lm", formula = y ~ ns(x, 3), se = FALSE, colour = "lightgrey", linewidth = 0.3) +
  geom_smooth(aes(group = as.character(Cluster), colour = as.character(Cluster)), method = "lm", formula = y ~ ns(x, 3), linewidth = 0.8) +
  labs(x = "Time post-transplant (months)", y = "Scaled log abundance trajectory") +
  theme + 
  theme(legend.position = "none", axis.text.x = element_text(size = 7)) +
  facet_grid(Cluster_description_de~., switch = "y") +
  geom_vline(xintercept = 4, lty = "dashed", linewidth = 0.6, colour = "darkgrey") +
  scale_colour_manual(values = c("#f78436", "#150578")) +
  scale_x_discrete(expand = c(0,0.5))
ggsave(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/met-lip-clusters.pdf", p, width = 9, height = 12, units = 'cm'))

ms_cluster_assignment <- cluster_ms_df %>% dplyr::select(Molecule, Cluster, Cluster_description) %>% distinct() %>% as.data.frame() %>%
  left_join(molecules_list$months$molecules_info ) %>% arrange(Molecule) 
# write.csv(ms_cluster_assignment, "tmixclust/molecules/ms-cluster-assignment.csv", row.names = FALSE)

ms_classes <- read.csv(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/stable/stable-molecules-heatmap.csv"))
ms_cluster_assignment <- ms_cluster_assignment %>% 
  left_join(ms_classes[, c("Molecule", "Class", "Rename")]) %>%
  left_join(stable_molecules_combined_df) %>%
  mutate(Cluster_description = factor(Cluster_description, levels = c("Amino and carboxylic acid metabolism", "Glycerophospholipid metabolism"))) %>%
  arrange(Cluster_description)  %>% 
  as.data.frame() %>%
  arrange(Cluster_description, Class) %>%
  mutate(Class = factor(Class, levels = c(setdiff(unique(Class), c("Other")), "Other")) )
rownames(ms_cluster_assignment) <- ms_cluster_assignment$Molecule
```

#### Continuous

```{r}
data <- molecules_combined_heat %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Patient_days") %>%
  pivot_longer(!Patient_days, values_to = "Abundance", names_to = "Molecule") %>%
  left_join(stable_small_molecules_metadata[, c ("Patient_days", "Months")]) %>%
  left_join(ms_cluster_assignment) %>%
  left_join(cluster_ms_df[, c("Cluster_description", "Cluster_description_de")] %>% distinct())
  
p <- 
  
  ggplot(data, aes(x = Months, y = Abundance)) +
  geom_smooth(aes(group = Molecule), method = "lm", formula = y ~ ns(x, 3), se = FALSE, colour = "lightgrey", linewidth = 0.3) + # genes
  geom_smooth(aes(group = as.character(Cluster), colour = as.character(Cluster)), method = "lm", formula = y ~ ns(x, 3), linewidth = 0.8) +  
  labs(x = "Time post-transplant\n(Months)", y = "Scaled log abundance", title = "DA molecules trajectory") +
  theme + 
  theme(legend.position = "none", strip.text = element_text(size = 10), plot.title = element_text(size = 9)) +
  facet_grid(Cluster_description_de~., switch = "y") +
  scale_x_continuous(expand = c(0,0)) +
  scale_colour_manual(values = c("#f78436", "#150578"))

ggsave(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/stable/ms-clusters-cont.png"), p, width = 6, height = 11, units = 'cm', dpi = 600)
```

### Pathways

```{r}
stable_path_list <- list()
c <- c("Amino and carboxylic acids", "Glycerophospholipids")

for (i in c(1:2)){ # for each cluster separately

compounds <- ms_cluster_assignment %>% filter(Cluster == i) %>% pull(KEGG) %>% gsub(";.*", "", .) %>% na.omit()
path <- defineCompounds(compounds = compounds, data = fella.hsa)
path_analysis <- enrich(compounds = path@userinput@metabolites, data = fella.hsa, methods = listMethods(), approx = "normality")

path_analysis_table <- generateResultsTable(object = path_analysis, method = "hypergeom", data = fella.hsa) 
colnames(path_analysis_table) <- c("Pathway", "Pathway_name", "CompoundHits", "CompoundsInPathway", "p_value")
path_analysis_table <- path_analysis_table %>%
    mutate(p_value = as.numeric(p_value),
           padj = p.adjust(p_value, method = 'BH', n = length(p_value)))

path_analysis_hy <- generateResultsGraph(object = path_analysis, method = "hypergeom", data = fella.hsa)
st.hy <- data.frame(as_edgelist(path_analysis_hy))
colnames(st.hy) <- c("name", "Pathway")
st.hy <- st.hy %>% inner_join(data.frame(vertex_attr(path_analysis_hy)), by = "name")

path_analysis_df <- st.hy %>%
  full_join(path_analysis_table, by = "Pathway") %>%
  na.omit() %>%
  dplyr::select("KEGG" = name, Pathway, "Metabolite" = label, Pathway_name, CompoundHits, CompoundsInPathway, p_value, padj)
path_analysis_df$Pathway_name <- gsub(" - Homo sapiens (human)", "", path_analysis_df$Pathway_name, fixed = TRUE)

stable_path_list[[c[i]]] <- as.data.frame(path_analysis_df)

}

# rm(compounds, path, path_analysis, path_analysis_table, path_analysis_hy, st.hy, path_analysis_df)
```

```{r}
# get top pathways
top_aa_pathways <- stable_path_list$`Amino and carboxylic acids` %>% 
  dplyr::select(Pathway_name, CompoundHits, padj) %>% 
  distinct() %>% filter(padj < 0.05) %>% 
  arrange(padj) %>% 
  slice_max(CompoundHits, n = 10) %>%
  pull(Pathway_name)

top_g_pathways <- stable_path_list$Glycerophospholipids %>% 
  dplyr::select(Pathway_name, CompoundHits, padj) %>% 
  distinct() %>% filter(padj < 0.05, CompoundHits > 2) %>% # at least 2 compoundhits
  arrange(padj) %>% 
  slice_max(CompoundHits, n = 10) %>%
  pull(Pathway_name)

d <- 
  
rbind(
stable_path_list$`Amino and carboxylic acids` %>% 
  filter(Pathway_name %in% top_aa_pathways) %>%
  mutate(Cluster = "Amino and carboxylic acids"),  # up

stable_path_list$Glycerophospholipids %>% 
  filter(Pathway_name %in% top_g_pathways) %>%
  mutate(Cluster = "Glycerophospholipids") # down
) %>% 
  mutate(Value = "1") %>%
  mutate(Pathway_name = str_to_sentence(Pathway_name),
         Pathway_name = gsub("Camp", "cAMP", Pathway_name),
         Pathway_name = gsub("Abc", "ABC", Pathway_name),
         Pathway_name = gsub("Ras", "RAS", Pathway_name),
         Pathway_name = gsub("trna", "tRNA", Pathway_name),
         Pathway_name = gsub("trp", "TRP", Pathway_name)
         )
  
p <- 
  
ggplot(d, aes(x = Value, y = reorder(Pathway_name, -log(padj)))) +
  geom_point(aes(fill = -log(padj), size = CompoundHits), shape = 21) +
  theme +
  theme(legend.position = "right", legend.justification = "left", legend.box = "vertical", legend.title = element_text(face = "bold")) +
    theme(axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_line(size = 0),
        strip.text = element_text(size = 8),
        #axis.text.y = element_text(margin = margin(r = 100), hjust = 0)
        ) +
  labs(title = paste("Pathways"),
       x = " ",
       y = " ",
       fill = "-Log(p-adj)",
       size = "DA molecules") +
  scale_fill_gradient(low = "#150578", high = "#f78436") +
  facet_grid(Cluster~., space = "free", scales = "free")

ggsave(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/stable/cluster-all-pathways.png"), p, width = 13, height = 11, units = 'cm', dpi = 600)  
```


### Combined Heatmap

```{r}
molecules_combined_heat <- as.data.frame(molecules_list$months$`ns(Months, df = 3)2`$heat_all[stable_molecules_combined, ] ) 

m <- ms_cluster_assignment %>% 
  #filter(Heatmap_annotation == TRUE) %>% 
  mutate(Cluster_description = gsub(" metabolism", "", Cluster_description, fixed = TRUE),
         Cluster_description = factor(Cluster_description)) %>%
  arrange(Cluster_description, Class)

ms_combined_heat_intervals <- molecules_combined_heat %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Patient_days") %>%
  left_join(stable_small_molecules_metadata[, c("Patient_days", "Months_grouped")]) %>%
  dplyr::select(!c("Patient_days")) %>%
  group_by(Months_grouped) %>%
  summarise(across(where(is.numeric), mean)) %>%
  column_to_rownames("Months_grouped") %>%
  t() %>%
  as.data.frame()

ms_combined_heat_intervals <- ms_combined_heat_intervals[ms_cluster_assignment$Molecule, ] 

df <- stable_small_molecules_metadata[, c("Months_grouped", "Immunosuppression")] %>% remove_rownames() %>%  
  distinct() %>% arrange(Months_grouped) %>%
  rename(`Time post-transplant (months)` = Months_grouped)
rownames(df) <- df$`Time post-transplant (months)`

identical(rownames(ms_combined_heat_intervals), rownames(ms_cluster_assignment)) # TRUE
identical(colnames(ms_combined_heat_intervals), rownames(df)) # TRUE

# Abbreviated molecules
rownames(ms_combined_heat_intervals) <- ms_cluster_assignment$Rename

colHm <- colorRamp2(breaks = seq(-1, 1, length = 7), colors=c('#2166ac','#67a9cf','#d1e5f0', '#f7f7f7', '#fddbc7', '#ef8a62', '#CD0514'))
```

```{r}
# Annotation
row_position <- ms_combined_heat_intervals %>% 
  rownames_to_column("Rename") %>% 
  mutate(Row = c(1:nrow(.))) %>% 
  dplyr::select(Rename, Row) %>% 
  filter(Rename %in% m$Rename)

text_annotation <- rowAnnotation(text_annotation = anno_mark(at = row_position %>% pull(Row) %>% as.numeric(), 
                                                   labels = row_position %>% pull(Rename), 
                                                   labels_gp = gpar(fontsize = 12, fontface = "bold"), which = "row"))

class_annotation <- annotate_heatmap_row(metadata = ms_cluster_assignment, 
                                        row_id = Rename, 
                                        row_or = ms_cluster_assignment$Rename, 
                                        annotations = c("Class"), 
                                        path_palette = path_palette)
```

```{r} 
h <- Heatmap(as.matrix(ms_combined_heat_intervals), 
             col = colHm,
        name = "z-score",
        column_title = "DA molecules",
        column_title_gp = gpar(fontface = "bold"),
        show_column_names = FALSE,
        show_row_names = FALSE,
        show_parent_dend_line = FALSE,
        cluster_row_slices = FALSE,
        cluster_rows = FALSE,
        top_annotation = annotate_heatmap(df, c("Immunosuppression", "Time post-transplant (months)")),
        column_order = levels(df$`Time post-transplant (months)`),
        row_split = ms_cluster_assignment$Cluster_description,
        row_title_gp = gpar(fontface = "bold") )

#h_list <- h + class_annotation + text_annotation

p <-
  
  draw(h,
     merge_legend = TRUE, 
     heatmap_legend_side = "left")

png(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/stable/combined-heat-interval.png"), width = 18, height = 25, units = "cm", res = 600)
p
invisible(dev.off())
```

### Alluvium

```{r}
p2 <- 
  
  ggplot(m %>% mutate(Rename = factor(Rename, levels = unique(Rename))), aes(axis1 = Rename, axis2 = Class)) +
  scale_x_discrete(expand = c(.05, .05)) + 
  geom_alluvium(aes(fill = Class), fill = "grey60") +
  geom_stratum(size = 0.1, width = 0.2, alpha = 0.3) +
  geom_text(stat = "stratum", aes(label = after_stat(stratum)), size = 2) +
  theme_void() +
  theme(legend.position = "none",
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        axis.text.y = element_blank())

ggsave(paste0(analysis_path, "transcriptomics-small-molecules/tmixclust/molecules/stable/alluvium.png"), width = 4, height = 25, units = "cm", dpi = 600)
```


### Class percentages

```{r}
p <- 
  
  ms_cluster_assignment %>% 
  group_by(Cluster_description) %>%
  dplyr::count(Class) %>%
  arrange(n, .by_group = TRUE) %>%
  mutate(Class_per = n/sum(n),
         label_y = cumsum(0.5 * cumsum(Class_per)),
         Cluster_description = gsub("Amino and carboxylic acids", "Amino  and\ncarboxylic acids", Cluster_description)) %>%

  ggplot(aes(x = Cluster_description, y = Class_per, fill = Class)) +
  geom_bar(width = 0.9, stat = "identity", position = "fill") +
  #geom_text(aes(x = fct_rev(as.character(Cluster)), y = label_y, label = scales::percent(Pathway_per)), size = 2) +
  theme +
  theme(axis.text.y = element_text(size = 10)) +
  labs(title = "Cluster class composition", x = "Cluster", y = "Percentage") +
  #coord_flip() +
  scale_fill_manual(values = path_palette)

ggsave("tmixclust/molecules/cluster-per.pdf", p, width = 15, height = 7, units = 'cm')


ms_cluster_assignment %>% 
  filter(Class == "Glycerophospholipids") %>%
  group_by(Cluster_description) %>%
  dplyr::count(SUB_CLASS) %>%
  arrange(n, .by_group = TRUE) %>%
  mutate(Class_per = n/sum(n),
         label_y = cumsum(0.5 * cumsum(Class_per)),
         Cluster_description = gsub("Amino acids and carboxylic acids", "Amino acids and\ncarboxylic acids", Cluster_description)) %>%

  ggplot(aes(x = Cluster_description, y = Class_per, fill = SUB_CLASS)) +
  geom_bar(width = 0.9, stat = "identity", position = "fill") +
  #geom_text(aes(x = fct_rev(as.character(Cluster)), y = label_y, label = scales::percent(Pathway_per)), size = 2) +
  theme +
  theme(axis.text.y = element_text(size = 10)) +
  labs(title = "Cluster class composition", x = "Cluster", y = "Percentage")
```


