---
title: "Biomarker cohort - Microbial associations post-transplant"
author: "Giulia Iacono"
date: Sys.Date()
output: html_document
---

# Biomarker cohort - Microbial associations post-transplant

```{r}
#bact <- readRDS("/Users/giac0002/Desktop/R/Biomarker2/Multiomics/Preprocessing/Microbiome/Data/16S/bact.rds")
detection <- 15
prevalence <- 0.25

bact.p <- phyloseq::subset_samples(bact, Record.ID %in% biomarker_patients) %>% core(detection = detection, prevalence = prevalence) 
otu_table(bact.p) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(bact.p), p = 0.5)), taxa_are_rows = TRUE)
groups.bact.logcss <- microbiome::transform(bact.p, transform = 'log') 
rm(bact.p) 

length(unique(sample_data(groups.bact.logcss)$Record.ID)) # 26
dim(otu_table(groups.bact.logcss)) # 54 156
```

```{r}
#fungi <- readRDS("/Users/giac0002/Desktop/R/Biomarker2/Multiomics/Preprocessing/Microbiome/Data/ITS/fungi.rds")
detection <- 15
prevalence <- 0.01

fungi.p <- phyloseq::subset_samples(fungi, Record.ID %in% biomarker_patients) %>% core(detection = detection, prevalence = prevalence)  
otu_table(fungi.p) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(fungi.p), p = 0.5)), taxa_are_rows = TRUE)
groups.fungi.logcss <- microbiome::transform(fungi.p, transform = 'log') 
rm(fungi.p) 

length(unique(sample_data(groups.fungi.logcss)$Record.ID)) # 24
dim(otu_table(groups.fungi.logcss)) # 79 107

data.frame(sample_data(groups.fungi.logcss)) %>% group_by(Group) %>% dplyr::count(Record.ID) %>% dplyr::count(Group)
```


# PCoA

```{r}
# Bacteria
permanova_list <- list()
p_list <- list()
c_list <- list("<1.5 months" = c("<1.5"), "1.5-6 months" = c("1.5-3", "4-6"), "7-12 months" = c("7-9", "10-12"), ">13 months" = c("13-18", ">18"))

for (i in c(1:4)){
  
      time <- unname(unlist(c_list[i]))
      data <- phyloseq::subset_samples(groups.bact.logcss, Months_grouped %in% time) 
      
      # adonis
      a <- adonis2(formula = as.formula(paste('phyloseq::distance(data, method = "wunifrac") ~', 'Group+Months+Transplant_indication+Total_ant_num+GIT_ischemic_time+Record.ID')), 
                    data = data.frame(sample_data(data)), 
                    distance = 'wunifrac', 
                    dfun = vegdist, 
                    permutations = 999,
                    parallel = 6)
  
      permanova_list[[i]] <- data.frame(a) %>% mutate(Test = names(c_list)[i])
      
      annotation <- paste(paste("SumOfSqs =", round(permanova_list[[i]]["Group", ]$SumOfSqs, 3)),
              paste("R2 =", round(permanova_list[[i]]["Group", ]$R2, 3)),
              paste("p =", round(permanova_list[[i]]["Group", ]$Pr..F., 3)), sep = " ; " )
    
      # plot
      ord <- ordinate(data, method="PCoA", distance="unifrac", weighted=TRUE)
      d <- plot_ordination(data, ord)

      p_list[[i]] <-
        
      plot_ordination(data, ord) + 
        stat_ellipse(aes(fill = Group, colour = Group), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
        geom_point(aes(fill = Group), shape = 21, size = 2) +
        theme +
        theme(plot.title = element_text(vjust = -2, hjust = 0.5, size = 10), plot.caption = element_text(size = 8)) +
        labs(title = names(c_list)[i],
             caption = annotation,
             colour = "Group",
             fill = "Group",
             x = gsub("Axis.1", "PCoA1", d$labels$x),
             y = gsub("Axis.2", "PCoA2", d$labels$y)) +
        plot_fill(Group = TRUE) +
        plot_colours(Group = TRUE)
      
}

p <- ggarrange(plotlist = p_list, ncol = 4, nrow = 1, common.legend = TRUE, legend = "right")
p <- annotate_figure(p, fig.lab = c("Bacterial ordination"), fig.lab.face = "bold", fig.lab.size = 11)
ggsave("clad-microbial-associations/bact-pcoa-h.pdf", p, width = 26, height = 7, units = 'cm')

# Differences are only at >12M and not prior, no interactions over time
```


```{r}
# Fungi
permanova_list <- list()
p_list <- list()
c_list <- list("<1.5 months" = c("<1.5"), "1.5-6 months" = c("1.5-3", "4-6"), "7-12 months" = c("7-9", "10-12"), ">13 months" = c("13-18", ">18"))

for (i in c(1:4)){
  
      time <- unname(unlist(c_list[i]))
      data <- phyloseq::subset_samples(groups.fungi.logcss, Months_grouped %in% time) 
      
      # adonis
      a <- adonis2(formula = as.formula(paste('phyloseq::distance(data, method = "wunifrac") ~', 'Group+Months+Transplant_indication+Total_antf_num+GIT_ischemic_time+Record.ID')), 
                    data = data.frame(sample_data(data)), 
                    distance = 'wunifrac', 
                    dfun = vegdist, 
                    permutations = 999,
                    parallel = 6)
  
      permanova_list[[i]] <- data.frame(a) %>% mutate(Test = names(c_list)[i])
      
      annotation <- paste(paste("SumOfSqs =", round(permanova_list[[i]]["Group", ]$SumOfSqs, 3)),
              paste("R2 =", round(permanova_list[[i]]["Group", ]$R2, 3)),
              paste("p =", round(permanova_list[[i]]["Group", ]$Pr..F., 3)), sep = " ; " )
    
      # plot
      ord <- ordinate(data, method="PCoA", distance="unifrac", weighted=TRUE)
      d <- plot_ordination(data, ord)

      p_list[[i]] <-
        
      plot_ordination(data, ord) + 
        stat_ellipse(aes(fill = Group, colour = Group), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
        geom_point(aes(fill = Group), shape = 21, size = 2) +
        theme +
        theme(plot.title = element_text(vjust = -2, hjust = 0.5, size = 10), plot.caption = element_text(size = 8)) +
        labs(title = names(c_list)[i],
             caption = annotation,
             colour = "Group",
             fill = "Group",
             x = gsub("Axis.1", "PCoA1", d$labels$x),
             y = gsub("Axis.2", "PCoA2", d$labels$y)) +
        plot_fill(Group = TRUE) +
        plot_colours(Group = TRUE) 
      
}

p <- ggarrange(plotlist = p_list, ncol = 4, nrow = 1, common.legend = TRUE, legend = "right")
p <- annotate_figure(p, fig.lab = c("Fungal ordination"), fig.lab.face = "bold", fig.lab.size = 11)
ggsave("clad-microbial-associations/fungi-pcoa-h.pdf", p, width = 26, height = 7, units = 'cm')
```

# Shannon diversity

```{r}
# Bacteria
d <- data.frame(sample_data(groups.bact.logcss))
# check time
model_full <- lmer(Diversity ~ Group*ns(Months, 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time + (1 | Record.ID),
  data = d,
  REML = FALSE
)

model_reduced <- lmer(Diversity ~ Group+ns(Months, 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time + (1 | Record.ID),
  data = d,
  REML = FALSE
)

anova(model_reduced, model_full) # pval for time interaction 0.768 so use reduced

summary(model_reduced) #pval for Group 0.37103

emm <- emmeans(model_reduced, ~ Months | Group, at = list(Months = c(1, 6)) )
confint(contrast(emm, method = "revpairwise"), level = 0.95)

# (0.535 increase, 95% CI 0.185–0.885, p = 0.768)

p <-
  
ggplot(d, aes(x = Months, y = Diversity)) +
  geom_smooth(aes(group = Record.ID), method = "lm", formula = y ~ ns(x, 2), se = FALSE, colour = "lightgrey", linewidth = 0.3) +
  geom_point(aes(fill = Group), shape = 21) + 
  geom_smooth(aes(group = Group, colour = Group), method = "lm", formula = y ~ ns(x, 3), linewidth = 0.8) +
  theme +
  theme(legend.position = "none") +
  labs(x = "Months post-transplant", 
       y = "Shannon Index",
       shape = "Transplant indication",
       fill = "Group",
       title = "Bacterial Diversity") +
  plot_fill(Group = TRUE) +
  plot_colours(Group = TRUE) 

ggsave("clad-microbial-associations/bact-shannon.pdf", p, width = 8, height = 7, units = 'cm')
```

```{r}
# Fungi
d <- data.frame(sample_data(groups.fungi.logcss))
# check time
model_full <- lmer(Diversity ~ Group*ns(Months, 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time + (1 | Record.ID),
  data = d,
  REML = FALSE
)

model_reduced <- lmer(Diversity ~ Group+ns(Months, 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time + (1 | Record.ID),
  data = d,
  REML = FALSE
)

anova(model_reduced, model_full) # pval for time interaction 0.05677 so use reduced

summary(model_reduced) #pval for Group 0.74042

emm <- emmeans(model_reduced, ~ Months | Group, at = list(Months = c(1, 6)) )
confint(contrast(emm, method = "revpairwise"), level = 0.95)

# (0.342 increase, 95% CI 0.0406–0.643, p = 0.05677)

p <-
  
ggplot(d, aes(x = Months, y = Diversity)) +
  geom_smooth(aes(group = Record.ID), method = "lm", formula = y ~ ns(x, 2), se = FALSE, colour = "lightgrey", linewidth = 0.3) +
  geom_point(aes(fill = Group), shape = 21) + 
  geom_smooth(aes(group = Group, colour = Group), method = "lm", formula = y ~ ns(x, 3), linewidth = 0.8) +
  theme +
  theme(legend.position = "none") +
  labs(x = "Months post-transplant", 
       y = "Shannon Index",
       shape = "Transplant indication",
       fill = "Group",
       title = "Fungal Diversity") +
  plot_fill(Group = TRUE) +
  plot_colours(Group = TRUE) 

ggsave("clad-microbial-associations/fungi-shannon.pdf", p, width = 8, height = 7, units = 'cm')
```

# Limma between groups

```{r}
# unique(tax_table(groups.bact.logcss)[,6])
data <- groups.bact.logcss
metadata <- data.frame(sample_data(data)) 
input_data <- data.frame(otu_table(data), check.names = FALSE)
identical(metadata$Patient_days, colnames(input_data))

design <- model.matrix(~Group + ns(Months, df = 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time, metadata) 
corfit <- duplicateCorrelation(input_data, design = design, block = metadata$Record.ID)
fit <- lmFit(input_data, design = design, block = metadata$Record.ID, correlation = corfit$consensus)
efit <- eBayes(fit)
summary(decideTests(efit, lfc = 0, p.value = 1)) # No DA species for Group 

topTable(efit, coef = "GroupCLAD", number = Inf, adjust = 'BH') %>% 
  as.data.frame() %>% 
  mutate(Direction = case_when(logFC > 0 ~ "Increased", logFC < 0 ~ "Decreased")) %>% 
  rownames_to_column("ASV") %>% 
  filter(abs(logFC) > 0, adj.P.Val < 0.1)
```

# Antibiotics

```{r}
antibiotics <- master_metadata$meds_metadata %>%
  dplyr::select(Patient_days, Total_ant_num, blactamother, `sulfonamide-benzylpyrimidine`, macrolide) %>%
  separate(Patient_days, c("Record.ID", "Days"), sep = "_") %>%
  mutate(Days = as.numeric(Days) , # ignore NA message
         Months = round(Days/30.417, 1),
         Record.ID = as.numeric(Record.ID), 
         #blactamother = ifelse(blactamother > 0, "Yes", "No"),
         #`sulfonamide-benzylpyrimidine` = ifelse(`sulfonamide-benzylpyrimidine` > 0, "Yes", "No"),
         #macrolide = ifelse(macrolide > 0, "Yes", "No"),
         CLAD_diagnosis = "CLAD diagnosis") %>%
  filter(Record.ID %in% biomarker_patients) %>%
  left_join(clad_assessment) %>%
  mutate(Days_interval = ifelse(Days <= 183, "<6 months", ">6 months"),
         Months_grouped = case_when(Days <= 42 ~ "<1.5",
                                    Days > 42 & Days <= 92 ~ "1.5-3",
                                    Days > 92 & Days <= 183 ~ "4-6",
                                    Days > 183 & Days <= 273 ~ "7-9",
                                    Days > 273 & Days <= 365 ~ "10-12",
                                    Days > 365 & Days <= 547 ~ "13-18",
                                    Days > 547 ~ ">18"), 
         Months_grouped = factor(Months_grouped, c("<1.5", "1.5-3", "4-6", "7-9", "10-12", "13-18", ">18")),
         days_to_clad = clad_day_clinical - Days,
         Status = ifelse(days_to_clad > 0 | is.na(days_to_clad), "(before)", "(after)"),
         Group_status = paste(Group, Status),
         Group_status = factor(Group_status, levels = c("CLAD-free (before)", "CLAD (before)", "CLAD (after)")),
         Group_interval = paste(Group, Months_grouped)) %>%
  filter(is.na(days_to_clad) | days_to_clad >= -30, Months < 26) %>%
  arrange(Group, Months_grouped) %>%
  mutate(Group_interval = factor(Group_interval, levels = unique(Group_interval)) ) %>%
  group_by(Group, Group_interval, Record.ID) %>%
  summarise(total_ant_mean = mean(Total_ant_num),
            blactamother_mean = mean(blactamother),
            sulfonamide_mean = mean(`sulfonamide-benzylpyrimidine`),
            macrolide_mean = mean(macrolide) ) 

# Average per patient
p <- 
  
  ggplot(antibiotics, aes(x = Group_interval, y = total_ant_mean)) +
    geom_boxplot(aes(fill = Group), alpha = 0.2) +
    geom_path(aes(group = Record.ID), colour = "lightgrey") +
    geom_point() +
    stat_compare_means(label = "p.signif", 
                     method = "wilcox.test",
                     ref.group = "CLAD <1.5") +
    labs(title = "Antibiotics usage", y = "Average per patient", x = "Group", fill = "Group") +
    theme +
    theme(legend.position = "none",
          axis.text.x = element_text(size = 7)) +
    plot_fill(Group = TRUE) 

ggsave("clad-microbial-associations/antibiotics.pdf", p, width = 22, height = 7, units = 'cm')
```

# Antifungals

```{r}
antifungals <- master_metadata$meds_metadata %>%
  dplyr::select(Patient_days, Total_antf_num) %>%
  separate(Patient_days, c("Record.ID", "Days"), sep = "_") %>%
  mutate(Days = as.numeric(Days) , # ignore NA message
         Months = round(Days/30.417, 1),
         Record.ID = as.numeric(Record.ID)) %>%
  filter(Record.ID %in% biomarker_patients) %>%
  left_join(clad_assessment) %>%
  mutate(Days_interval = ifelse(Days <= 183, "<6 months", ">6 months"),
         days_to_clad = clad_day_clinical - Days,
         Status = ifelse(days_to_clad > 0 | is.na(days_to_clad), "(before)", "(after)"),
         Group_interval = paste(Group, Days_interval),
         Group_interval = factor(Group_interval, levels = c("CLAD-free <6 months", "CLAD-free >6 months", "CLAD <6 months", "CLAD >6 months" )) 
         ) %>%
  filter(is.na(days_to_clad) | days_to_clad >= -30, Months < 26) %>%
  group_by(Group, Group_interval, Record.ID) %>%
  summarise(total_antf_mean = mean(Total_antf_num)) 

# Average per patient
p <- 
  
  ggplot(antifungals, aes(x = Group_interval, y = total_antf_mean)) +
    geom_boxplot(aes(fill = Group), alpha = 0.2) +
    geom_path(aes(group = Record.ID), colour = "lightgrey") +
    geom_point() +
    stat_compare_means(label = "p.signif", 
                     method = "wilcox.test",
                     ref.group = "CLAD >6 months") +
    labs(title = "Antifungal usage", y = "Average per patient", x = "Group", fill = "Group") +
    theme +
    theme(legend.position = "none",
          axis.text.x = element_text(size = 7)) +
    plot_fill(Group = TRUE) 

ggsave("clad-microbial-associations/antifungals.pdf", p, width = 12, height = 7, units = 'cm')
```

# Taxonomy

```{r}
data.df <- groups.bact.logcss %>%
  phyloseq::tax_glom(taxrank = 'Genus', NArm = T) %>%
  microbiome::transform(transform = 'compositional') %>% 
  phyloseq::psmelt()

p <-
  
  ggplot(data.df, aes(x = Time_interval_combined, y = Abundance, NArm = F)) + 
    geom_bar(aes(fill = Phylum, colour = Phylum), linewidth = 0.1, stat = 'identity', position = "fill") + 
    theme +
    labs(x = "Time interval",
         y = "Relative abundance", 
         title = "Taxonomy composition") +
    theme(#legend.key.size = unit(0.3, "cm"),
          axis.text = element_text(size = 8),
          legend.text = element_text(size = 9),
          strip.text = element_text(size = 8)) +
    guides(fill = guide_legend(ncol = 1)) +
    #facet_grid(.~Time_interval_combined, scales = "free", space = "free") +
    scale_fill_manual(values = clad_phylum_colours) +
    scale_colour_manual(values = clad_phylum_colours)

ggsave("clad-microbial-associations/bact-tax-phylum.pdf", p, width = 14, height = 8, units = 'cm')
```

