---
title: "biomarker-rna-associations"
output: html_document
date: "2025-02-27"
---

# Transcriptomics

```{r}
biomarker_rna_metadata <- rna_corrected$samples %>% filter(Record.ID %in% biomarker_patients)
biomarker_rna <- rna_corrected_normalised$E[, biomarker_rna_metadata$Patient_days]
identical(biomarker_rna_metadata$Patient_days, colnames(biomarker_rna)) #TRUE
dim(biomarker_rna) # 127
biomarker_rna_metadata %>% group_by(Record.ID) %>% dplyr::count() # 26
```

# PCA

```{r}
a <- data.frame(adonis2(formula = as.formula("t(assay(vsd)) ~Days_interval+Transplant_indication+Total_ant_num+Record.ID"), 
                    data = colData(vsd), 
                    distance = "eu", 
                    dfun = vegdist, 
                    permutations = 999,
                    na.action = na.omit,
                    parallel = 6))

annotation <- paste0("SumOfSqs=", round(a["Group",]$SumOfSqs, 3), "\n",
              "R2=", round(a["Group",]$R2, 3), "\n",
              "p=", round(a["Group",]$Pr..F., 3) )

pca.list <- pca_process(as.data.frame(assay(vsd)), as.data.frame(colData(vsd)))

p <-
  
  ggplot(pca.list$data.PCA, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(fill = Time_interval_combined, colour = Time_interval_combined), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
  geom_point(aes(fill = Time_interval_combined, shape = Group), size = 2) +
  theme +
  annotate(geom = "text", x = -150, y = 110, label = annotation, size = 3, hjust = 0, vjust = 1) +
  labs(title = 'Transcriptomics PCA',
       shape = "Group",
       colour = "Time interval",
       fill = "Time interval",
       x = paste0('PC1 ', round(pca.list$varexp[1], 2), '%'),
       y = paste0('PC2 ', round(pca.list$varexp[2], 2), '%')) +
  scale_shape_manual(values = c(21, 24)) +
  plot_fill(Time_interval_combined = TRUE) +
  plot_colours(Time_interval_combined = TRUE)

ggsave("pca/rna/pca-interval.pdf", p, width = 12, height = 7, units = 'cm')
```


```{r}
permanova_list <- list()
p_list <- list()

dds <- DESeqDataSetFromMatrix(countData = biomarker_rna,
                              colData = biomarker_rna_metadata,
                              design = ~1)  
vsd <- vst(dds, blind=FALSE)

for (i in c(1:2)){
  
      time <- unique(colData(dds)$Days_interval)[i]
      
      m <- colData(dds) %>% filter(Days_interval == time)
      data <- assay(vsd)[, m$Patient_days]
      
      # adonis
      a <- data.frame(adonis2(formula = as.formula("t(data) ~Group+Months+Transplant_indication+Record.ID"), 
                      data = m, 
                      distance = "eu", 
                      permutations = 999,
                      na.action = na.omit,
                      parallel = 6))
  
      permanova_list[[i]] <- data.frame(a) %>% mutate(Test = time)
      
      d <- permanova_list[[i]]["Group", ]
      annotation <- paste0("SumOfSqs=", round(d$SumOfSqs, 3), "\n",
                    "R2=", round(d$R2, 3), "\n",
                    "p=", round(d$Pr..F., 3) )

      # plot
      pca.list <- pca_process(data, m)

      p_list[[i]] <-
        
          ggplot(pca.list$data.PCA, aes(x = PC1, y = PC2)) +
            stat_ellipse(aes(fill = Group, colour = Group), geom = 'polygon', type = 't', level = 0.95, alpha = 0.2) + 
            geom_point(aes(fill = Group), shape = 21, size = 2) +
            theme +
            theme(plot.title = element_text(vjust = -2, size = 10)) +
            annotate(geom = "text", x = -200, y = 200, label = annotation, size = 3, hjust = 0, vjust = 1) +
            labs(title = time,
                 colour = "Group",
                 fill = "Group",
                 x = paste0('PC1 ', round(pca.list$varexp[1], 2), '%'),
                 y = paste0('PC2 ', round(pca.list$varexp[2], 2), '%')) +
                  plot_fill(Group = TRUE) +
                  plot_colours(Group = TRUE) +
                  scale_y_continuous(limits = c(-200, 200)) +
                  scale_x_continuous(limits = c(-200, 200))
      
}

p <- ggarrange(p_list[[1]], p_list[[2]], ncol = 2, nrow = 1, common.legend = TRUE, legend = "bottom")
p <- annotate_figure(p, fig.lab = c("Transcriptomics PCA"), fig.lab.face = "bold", fig.lab.size = 11)
ggsave("pca/rna/pca-groups.pdf", p, width = 16, height = 7, units = 'cm')

# Differences between groups over time after 12M only, not at 3M, significant interaction of groups with time 
# Differences between groups at 12M are part of a shared early inflammatory signature that dissipates after X months, but resurfaces in CLAD patients after 12M
```

# *DESEQ2 Differential expression group + months

```{r}
# deseq_list$`group+months` <- NULL
test_variable <- "group+months"

m <- biomarker_rna_metadata %>% filter(!is.na(GIT_ischemic_time))
d <- biomarker_rna[, m$Patient_days]

de_matrix <- model.matrix(~Group + ns(Months, df = 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time, m) 
deseq_list[[test_variable]][["design"]] <- "Group + ns(Months, df = 3) + Transplant_indication + Total_ant_num + GIT_ischemic_time"

# view(bind_rows(deseq_list$`group+months`$de_summary)) 
deseq_list$`group+months`$GroupCLAD$res_df %>% view()
```

```{r}
# Run
dds <- DESeqDataSetFromMatrix(countData = d,
                              colData = m,
                              design = de_matrix)  
dds <- DESeq(dds, minReplicatesForReplace = 3) 
mcols(dds)$maxCooks <- apply(assays(dds)[["cooks"]], 1, max) # add max cooks column
deseq_list[[test_variable]][["dds"]] <- dds
#dds <- deseq_list$`group*months`$dds

vsd <- vst(dds, blind=FALSE) 
deseq_list[[test_variable]][["vsd"]] <- vsd

# resultsNames(dds)

for (i in c(2)){ 
  
  name <- resultsNames(dds)[i]
  res <- lfcShrink(dds, coef = name, lfcThreshold = 0, type = "ashr", parallel = TRUE, BPPARAM = SnowParam(6) )
  deseq_list[[test_variable]][[name]][["res"]] <- res
  
  # summary(res, alpha = 0.05)
  logFC_threshold <- 0
  # cutoff at max qf(.99, 5, ncol(dds) - 5)
  #cutoff <- qf(.99, 5, ncol(dds) - 5)
  
  c <- data.frame("maxCooks" = mcols(dds)$maxCooks, "Gene" = names(mcols(dds)$maxCooks))
    
  res.df <- as.data.frame(res) %>% 
    rownames_to_column("Gene") %>%
    left_join(c) %>%
    mutate(Direction = case_when(
        padj < 0.05 & log2FoldChange <= -logFC_threshold ~ "Decreased",
        padj < 0.05 & log2FoldChange >= logFC_threshold ~ "Increased",
        padj >= 0.05 | abs(log2FoldChange) < logFC_threshold ~ 'n.s.',
        is.na(padj) == TRUE | is.na(pvalue) == TRUE ~ "NA")) 
  deseq_list[[test_variable]][[name]][["res_df"]] <- res.df
  
  #dconfects <- deseq2_confects(dds, name=name, step=0.01, cooksCutoff=Inf, independentFiltering=FALSE)
  #deseq_list[[test_variable]][[name]]["dconfects"] <- dconfects
  #dconfects$table <- dconfects$table %>% rename(Gene = name) %>% dplyr::select(rank, index, confect, filtered, Gene)
  #res.df <- res.df %>% left_join(dconfects$table)

  sig.results.df <- res.df %>% filter(padj < 0.05 & abs(log2FoldChange) >= logFC_threshold)  
  deseq_list[[test_variable]][[name]][["sig_results_df"]] <- sig.results.df
  deseq_list[[test_variable]][[name]][["heat_sig"]] <- t(scale(t(as.data.frame( assay(vsd)[rownames(assay(vsd)) %in% sig.results.df$Gene,]))))
  
  if (nrow(sig.results.df) == 0){
    next 
  }

  g <- sig.results.df %>% group_by(Direction) %>% mutate(Genes = paste0(Gene, collapse = ", ")) %>% dplyr::select(Direction, Genes) %>% distinct()
  
  deseq_list[[test_variable]][["de_summary"]][[name]] <- sig.results.df %>% 
    group_by(Direction) %>% 
    dplyr::count() %>% 
    ungroup() %>%    
    left_join(g, by = "Direction") %>%
    mutate(Name = name) 
  
  top.results.df <- res.df %>% filter(padj < 0.01 & abs(log2FoldChange) >= logFC_threshold) 
  deseq_list[[test_variable]][[name]][["top_results_df"]] <- top.results.df
   if (nrow(top.results.df) > 0){
      deseq_list[[test_variable]][[name]][["heat_top"]] <- t(scale(t(as.data.frame( assay(vsd)[rownames(assay(vsd)) %in% top.results.df$Gene,] )))) 
    }
}

saveRDS(deseq_list, "deseq/deseq_list.rds")
#deseq_list <- readRDS("deseq/deseq_list.rds")
```

```{r}
biomarker_rna_combined_df <- deseq_list$`group+months`$GroupCLAD$sig_results_df %>%
  filter(abs(log2FoldChange) >= 1) 
biomarker_rna_combined <- biomarker_rna_combined_df$Gene
length(biomarker_rna_combined) # 222
```

```{r}
# Filter for DE genes in GEM
data <- rna_corrected$genes %>% filter(SYMBOL %in% biomarker_rna_combined)

gem_list <- list()
for (i in c(1:length(biomarker_rna_combined))) {
  gem_list[[i]] <- gem_reactions$GENE.ASSOCIATION %like% data$ENSEMBL[i] 
}
a <- Matrix::colSums(do.call(rbind, gem_list))
index <- a > 0
biomarker_gem_reactions_de <- gem_reactions[index, ] %>% remove_rownames()

# Replace ENSEMBL with SYMBOL
d <- as.list(biomarker_gem_reactions_de$GENE.ASSOCIATION)
g <- rna_corrected$genes %>% filter(SYMBOL %in% biomarker_rna_combined) %>% pull(SYMBOL)
names(g) <- rna_corrected$genes %>% filter(SYMBOL %in% biomarker_rna_combined) %>% pull(ENSEMBL)

etos_list <- list()
for (i in c(1:length(d))){
  etos_list[[i]] <- d[[i]] %>% str_replace_all(g) %>% str_split(., pattern = " or ") %>% unlist() %>% str_subset(., "ENSG", negate = TRUE) %>% paste(., sep = "", collapse = ",")
}

biomarker_gem_reactions_de <- cbind(biomarker_gem_reactions_de, "GENES" = do.call(rbind, etos_list)) %>% relocate(GENES, .before = GENE.ASSOCIATION) 

biomarker_gem_reactions_de <- biomarker_gem_reactions_de %>%
  mutate(Metabolism = ifelse(grepl("Amino sugar|Arginine|Cysteine|Glycine|Histidine|Phenylalanine|Tryptophan|Tyrosine|Protein", SUBSYSTEM), 
                             "Amino acid and peptide", 
                             ifelse(grepl("Biopterin|Pantothenate|Nicotinate|Riboflavin|Retinol|Vitamin", SUBSYSTEM), 
                                    "Vitamins and other amino compounds", 
                                    ifelse(grepl("Purine|Pyrimidine|Nucleotide", SUBSYSTEM), 
                                           "Nucleotide",
                                          ifelse(grepl("Fatty|fatty", SUBSYSTEM), "Fatty acid", 
                                                  ifelse(grepl("Drug|Estrogen|Serotonin|Sulfur|Xenobiotics|Transport|Miscellaneous|Pool|Isolated", SUBSYSTEM), "Miscellaneous", 
                                                         ifelse(grepl("Glycolysis|Fructose|Galactose|Pentose|Starch|Tricarboxylic|Pyruvate", SUBSYSTEM), 
                                                                "Cellular energy",
                                                                ifelse(grepl("Chondroitin|Keratan", SUBSYSTEM), 
                                                                       "Extracellular matrix", 
                                                                       ifelse(grepl("Glycerolipid|Glycerophospholipid|Sphingolipid", SUBSYSTEM), "Complex lipid", 
                                                                              ifelse(grepl("Steroid|Arachidonic|Eicosanoid|Leukotriene|Linoleate|Omega|Prostaglandin", SUBSYSTEM), "Eicosanoid and steroid", ""))) )))))))

saveRDS(biomarker_gem_reactions_de, "gem/biomarker_gem_reactions_de.rds")
#biomarker_gem_reactions_de <- readRDS("gem/biomarker_gem_reactions_de.rds")

#unique(biomarker_gem_reactions_de$GENES) %>% sort() %>% str_split(., pattern = ",") %>% unlist() %>% unique()
#unique(biomarker_gem_reactions_de$SUBSYSTEM) %>% sort() %>% write.csv(., "gem/reactions.csv")
```

```{r}
biomarker_metabolism <- biomarker_gem_reactions_de %>%
  filter(grepl(paste(as.character(na.omit(stable_molecules_gem_names$NAME)), collapse = "|"), EQUATION))
write.csv(biomarker_metabolism %>% arrange(GENES), "gem/biomarker_metabolism.csv")
```

```{r}
dset <- biomarker_rna_combined_df 
gene.list <- list()
d <- read.csv("genes-database/genes-database.csv") 

for (i in 1:length(dset$Gene)){
  gene <- geneinfo_2_html(dset$Gene[i])
  gene <- sub(".*GeneCards database: <a href = ", "", gene)
  gene <- sub(" target.*", "", gene)
  gene <- gsub('"', '', gene)
  gene.list[[i]] <- gene
}
g <- dplyr::bind_cols(gene.list) %>% t() %>% as.data.frame %>% remove_rownames()
gene_links <- data.frame("Link" = g$V1, 
                         "Gene" = dset$Gene) %>% 
  left_join(d) %>% 
  right_join(biomarker_rna_combined_df) %>%
  arrange(Gene)
write.csv(gene_links, "deseq/group+months/gene-links.csv", row.names = FALSE)

# rm(gene.list, gene, dset, d, g)
```

## Heatmap with intervals

```{r}
rna_combined_heat <- as.data.frame(deseq_list$`group+months`$GroupCLAD$heat_sig) 

gene_cluster_assignment_filt <- deseq_list$`group+months`$GroupCLAD$sig_results_df %>% 
  filter(Gene %in% biomarker_rna_combined) %>%
  as.data.frame() %>%
  #left_join(b_cluster_rna_df) %>%
  #arrange(Direction, Cluster) %>%
  mutate(Direction = factor(Direction)#,
         #Cluster = factor(Cluster),
         #Direction_cluster = factor(Direction_cluster)
         )
rownames(gene_cluster_assignment_filt) <- gene_cluster_assignment_filt$Gene

m <- read.csv("deseq/group+months/biomarker-rna-heatmap.csv") 

rna_combined_heat_intervals <- rna_combined_heat %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Patient_days") %>%
  left_join(biomarker_rna_metadata[, c("Patient_days", "Months_grouped", "Group")]) %>%
  dplyr::select(!c("Patient_days")) %>%
  group_by(Group, Months_grouped) %>%
  summarise(across(where(is.numeric), mean)) %>%
  ungroup() %>%
  mutate(Group_month = paste(Group, Months_grouped, sep = "_")) %>%
  dplyr::select(!c("Group", "Months_grouped")) %>%
  column_to_rownames("Group_month") %>%
  t() %>%
  as.data.frame()

rna_combined_heat_intervals <- rna_combined_heat_intervals[m$Gene, ] 

df <- biomarker_rna_metadata[, c("Months_grouped", "Immunosuppression", "Group")] %>% 
  distinct() %>% 
  arrange(Group, Months_grouped) %>%
  mutate(Group_month = paste(Group, Months_grouped, sep = "_"),
         Group_month = factor(Group_month, levels = unique(Group_month))) %>%
  rename(`Time post-transplant (months)` = Months_grouped)
rownames(df) <- df$Group_month

identical(rownames(rna_combined_heat_intervals), m$Gene) # TRUE
identical(colnames(rna_combined_heat_intervals), rownames(df))

# Annotation
#text_list_small <- read.csv("deseq/group+months/biomarker-rna-heatmap.csv") %>% filter(Heatmap_annotation == TRUE & Bold == FALSE) %>% pull(Gene)
text_list_big <- read.csv("deseq/group+months/biomarker-rna-heatmap.csv") %>% filter(sc_heatmap_annotation == TRUE) %>% pull(Gene)

row_position <- rna_combined_heat_intervals %>% 
  rownames_to_column("Genes") %>% 
  mutate(Row = c(1:nrow(.))) %>% 
  dplyr::select(Genes, Row)

#row_position_small <- row_position %>%
#  filter(Genes %in% text_list_small) 

row_position_big <- row_position %>%
  filter(Genes %in% text_list_big)

#text_annotation_small <- anno_mark(at = row_position_small %>% pull(Row) %>% as.numeric(), 
#                                                   labels = row_position_small %>% pull(Genes), 
#                                                   labels_gp = gpar(fontsize = 8),
#                                   which = "row")

text_annotation_big <- anno_mark(at = row_position_big %>% pull(Row) %>% as.numeric(), 
                                                   labels = row_position_big %>% pull(Genes), 
                                                   labels_gp = gpar(fontsize = 12, fontface = "bold"),
                                 which = "row")

text_annotation <- rowAnnotation(#text_annotation_small = text_annotation_small,
                                 text_annotation_big = text_annotation_big)

colHm <- colorRamp2(breaks = seq(min(rna_combined_heat_intervals, na.rm = T), max(rna_combined_heat_intervals, na.rm = T), length = 7), colors=c('#2166ac','#67a9cf','#d1e5f0', '#f7f7f7', '#fddbc7', '#ef8a62', '#CD0514'))
```

```{r}
p <- 
  
  draw(Heatmap(as.matrix(rna_combined_heat_intervals), 
        col = colHm,
        name = "z-score",
        column_title = "DE genes",
        column_title_gp = gpar(fontface = "bold"),
        row_title_gp = gpar(fontface = "bold"),
        show_column_names = FALSE,
        show_row_names = FALSE,
        show_parent_dend_line = FALSE,
        cluster_row_slices = FALSE,
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        top_annotation = annotate_heatmap(df, c("Immunosuppression", "Time post-transplant (months)", "Group")),
        right_annotation = text_annotation,
        column_order = levels(df$Group_month),
        row_split = m %>% pull(Direction) ),
     padding = unit(c(0, 0, 0, 0), "cm"),
     merge_legend = TRUE,
     heatmap_legend_side = "left")

png("deseq/group+months/biomarker-heat-interval-all.png", width = 18, height = 25, units = "cm", res = 500)
p
invisible(dev.off())
```

```{r}
b <- biomarker_rna_metadata[, c("Record.ID", "Group", "Months", "Patient_days")] %>% mutate(Record.ID = as.factor(Record.ID)) %>% arrange(Group, Record.ID, Months)
m <- rna_combined_heat %>% dplyr::select(b$Patient_days)

p <- 
  
  draw(Heatmap(as.matrix(m), 
        col = colHm,
        name = "z-score",
        column_title = "DE genes",
        column_title_gp = gpar(fontface = "bold"),
        row_title_gp = gpar(fontface = "bold"),
        show_column_names = FALSE,
        show_row_names = FALSE,
        show_parent_dend_line = FALSE,
        cluster_row_slices = FALSE,
        cluster_rows = TRUE,
        show_row_dend = FALSE,
        top_annotation = annotate_heatmap(b, c("Record.ID", "Group")),
        column_order = b$Patient_days
        ),
     padding = unit(c(0, 0, 0, 0), "cm"),
     merge_legend = TRUE,
     heatmap_legend_side = "left")

png("deseq/group+months/biomarker-heat-interval-all.png", width = 18, height = 25, units = "cm", res = 500)
p
invisible(dev.off())
```


## Pathways

```{r}
# deseq_list$`group+months`$cluster_genes <- NULL
  #d <- b_cluster_rna_df %>% filter(Direction_cluster == unique(b_cluster_rna_df$Direction_cluster)[i]) %>% pull(Gene)
  
  gene.df <- rna_corrected$genes %>% filter(SYMBOL %in% biomarker_rna_combined) 

  ego <- enrichGO(gene = gene.df$ENTREZID,
                  OrgDb         = org.Hs.eg.db,
                  ont           = "BP",
                  pAdjustMethod = "BH",
                  pvalueCutoff  = 0.2,
                  qvalueCutoff  = 0.2,
                  readable = TRUE)
  ego@result$log_adjpvalue <- abs(log10(ego@result$p.adjust))
  if (nrow(ego@result) > 0){
    if (length(unique(ego@result$geneID)) < nrow(ego@result)){
      ego@result <- aggregate(ego@result, by = list(ego@result$geneID), FUN = function(x) {
                                                                    if (is.numeric(x))
                                                                      cbind(x[1]) #max(x)
                                                                    else if (is.character(x))
                                                                      cbind(x[1])
                                                                    }) # keep first value only
    }
  }
  deseq_list$`group+months`[["cluster_genes"]][["ego"]] <- ego
  
  ego@result$qvalue <- NA
  ego.simp <- clusterProfiler::simplify(ego, cutoff = 0.35, by="p.adjust")
  deseq_list$`group+months`[["cluster_genes"]][["ego_simp"]] <- ego.simp

# rm(ego, ego.simp, gene.df)

deseq_list$`group+months`$cluster_genes$ego_simp@result %>% arrange(p.adjust) %>% filter(p.adjust < 0.05, Count >= 6) 
```

```{r}
biomarker_rna_pathways <- deseq_list$`group+months`$cluster_genes$ego_simp@result %>% 
  arrange(p.adjust) %>% 
  filter(p.adjust < 0.05, Count > 7) %>%
  mutate(Description = str_to_sentence(Description),
         Description = gsub("dna-binding", "DNA-binding", Description))

p <- 
  
ggplot(biomarker_rna_pathways, aes(x = 1, y = reorder(Description, -log(p.adjust)))) +
  geom_point(aes(fill = -log(p.adjust), size = Count), shape = 21) +
  theme +
  theme(legend.position = "right", legend.justification = "left", legend.box = "vertical", legend.title = element_text(face = "bold")) +
    theme(axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_line(size = 0)
        ) +
  labs(title = paste("Pathways"),
       x = " ",
       y = " ",
       fill = "-Log(p-adj)",
       size = "DE genes") +
  scale_fill_gradient(low = "#150578", high = "#f78436") 

ggsave("deseq/group+months/biomarker-pathways.pdf", p, width = 13, height = 9, units = 'cm')  
ggsave("deseq/group+months/biomarker-pathways.png", p, width = 13, height = 9, units = 'cm', dpi = 500)  
```



# *Limma differential expression group + months

```{r}
design <- model.matrix(~Transplant_indication + Total_ant_num + GIT_ischemic_time + Gender + ns(Months, df = 3) + Group, biomarker_rna_metadata)
dupcor <- duplicateCorrelation(biomarker_rna, design = design, block = biomarker_rna_metadata$Record.ID)
vfit <- lmFit(biomarker_rna, design, block = biomarker_rna_metadata$Record.ID, correlation = dupcor$consensus)
efit <- eBayes(vfit)
summary(decideTests(efit, p.value = 0.2, lfc = 0)) 
# topTreat(efit, coef = "GroupCLAD", n = 100)
```



# Needed

A model that can handle time as continuous (not glmmnet)
A model that can handle negative binomial raw data (DESeq2 performs best compared with eg Limma)
A model that can adjust for patient ID (random option preferred)
A model that can handle splines is preferred (or tip points can be tested after for hits only)

Need to use batch corrected data to avoid confounding genes

Dream
- Yes to continuous variables
- No to negative binomial
- Yes to multiple random effects
- No to splines
- No to hits after Combat

Limma
- Yes to continuous variables
- No to negative binomial
- Yes to 1 only random effect
- Yes to splines
- No to hits after Combat

DESeq2
- Yes to continuous variables
- Yes to negative binomial
- No to random effects
- Yes to splines
- Yes to hits after Combat

Glmmseq
- No to continuous variables
- Yes to negative binomial, can be adapted to lim
- Yes to random effects
- No to splines



