---
title: "biomarker-mofa"
output: html_document
date: "2025-02-27"
---

# MOFA 

```{r}
# RNA
select_var <- names(sort(apply(biomarker_rna, 1, var), decreasing=TRUE))[1:2000]
mofa_rna <- biomarker_rna[select_var, ]
# intersect(select_var, biomarker_rna_combined_df$Gene)

# MOLECULES
select_var <- names(sort(apply(biomarker_small_molecules, 1, var), decreasing=TRUE))[1:500]
mofa_molecules <- biomarker_small_molecules[select_var, ]

# BACTERIA
detection <- 15
prevalence <- 0.1

bact.p <- phyloseq::subset_samples(bact, Record.ID %in% biomarker_patients) %>% core(detection = detection, prevalence = prevalence) 
otu_table(bact.p) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(bact.p), p = 0.5)), taxa_are_rows = TRUE)
groups.bact.logcss <- microbiome::transform(bact.p, transform = 'log') 
mofa_bacteria <- otu_table(groups.bact.logcss)
mofa_bacteria_metadata <- data.frame(sample_data(groups.bact.logcss))
rm(bact.p, groups.bact.logcss) 

hist(as.vector(mofa_rna), breaks=500)
hist(as.vector(as.matrix(mofa_molecules)), breaks=100)
hist(as.vector(as.matrix(mofa_bacteria)), breaks=100) 
```

```{r}
features <- c("Record.ID", "Patient_days", "Group", "Days", "Months", "Months_grouped", "Gender", "Age", "Transplant_indication", "Total_ant_num", "GIT_ischemic_time", "days_to_clad")

mofa_metadata <- biomarker_rna_metadata[, c(features)] %>% 
  full_join(biomarker_small_molecules_metadata[, features]) %>%
  full_join(mofa_bacteria_metadata[, features]) %>%
  filter(Patient_days %in% unique(c(colnames(mofa_rna), colnames(mofa_molecules), colnames(mofa_bacteria)))) %>%
  arrange(Months_grouped, Group) %>%
  mutate(Group_months_grouped = paste(Group, Months_grouped, sep = "_"),
        Group_months_grouped = factor(Group_months_grouped, levels = unique(Group_months_grouped)),
        CLAD_progression = ifelse(Group == "CLAD-free", "CLAD-free", 
                                     ifelse(days_to_clad < 365 | is.na(days_to_clad), "CLAD<365", "CLAD>365")),
        CLAD_progression = factor(CLAD_progression, levels = c("CLAD-free", "CLAD>365", "CLAD<365")),
        Group = factor(Group, levels = c("CLAD-free", "CLAD")))

rownames(mofa_metadata) <- mofa_metadata$Patient_days # important or cannot map
dim(mofa_metadata) #172  14
```

```{r}
mofa_object <- MultiAssayExperiment(list(Transcripts = as.matrix(mofa_rna),
                                      Molecules = as.matrix(mofa_molecules),
                                      Bacteria = as.matrix(mofa_bacteria)), 
                  colData = mofa_metadata)
mofa_object <- create_mofa(mofa_object) 
mofa_object <- set_covariates(mofa_object, covariates = c("Months"))

p <- plot_data_overview(mofa_object, show_dimensions = TRUE, show_covariate = TRUE,
                   colors = c("#56CFE1", "#FF6348FF", "blue")) + 
  theme +
  theme(plot.title = element_text(face = "bold"),
              axis.text.x = element_blank(),
        legend.position = "bottom") +
    labs(title = "MOFA omics datasets", x = "Sample", y = "Omics")
ggsave(paste0(analysis_path, "biomarker-mofa/datasets.pdf"), p, width = 12, height = 8, units = 'cm')
```

## Train

```{r}
# mofa_list <- list()

# Set Data options
data.opts <- get_default_data_options(mofa_object)

# Model options
model.opts <- get_default_model_options(mofa_object)

# Training options 
train.opts <- get_default_training_options(mofa_object)

drop_factor_threshold <- 0.05
maxiter <- 10000
convergence_mode <- "slow"
train.opts$drop_factor_threshold <- drop_factor_threshold
train.opts$maxiter <- maxiter
train.opts$convergence_mode <- convergence_mode

# Mefisto options
mefisto.opts <- get_default_mefisto_options(mofa_object)

# Set object
m <- prepare_mofa(object = mofa_object,
  data_options = data.opts,
  model_options = model.opts,
  training_options = train.opts, 
  mefisto_options = mefisto.opts
  ) 

name <- paste("rna-mol-bact", drop_factor_threshold, maxiter, convergence_mode, sep = "-")
p <- run_mofa(m, outfile = paste0("biomarker-mofa/models/", name, "-preimputed.hdf5"), use_basilisk = TRUE) # cannot save remotely or fatal error
#p <- load_model(paste0("biomarker-mofa/models/", name, "-preimputed.hdf5"))
mofa_imputed <- impute(p)
rownames(mofa_imputed@samples_metadata) <- mofa_imputed@samples_metadata$Patient_days
saveRDS(mofa_imputed, paste0(analysis_path, "biomarker-mofa/models/", name, "-imputed.rds"))

mofa_list[[name]] <- mofa_imputed
saveRDS(mofa_list, paste0(analysis_path, "biomarker-mofa/models/mofa_list.rds"))
# mofa_list <- readRDS("biomarker-mofa/models/mofa_list.rds")
```

## Test with LMM

```{r}
# linear_model_list <- list()
# p_list <- list()

for (i in 1:3) {
  
  drop_factor_threshold <- c(0.01, 0.02, 0.05)[i]
  name <- paste("rna-mol-bact", drop_factor_threshold, maxiter, convergence_mode, sep = "-")
  
  mofa_imputed <- mofa_list[[name]]
  
  metadata <- mofa_imputed@samples_metadata
  latent_factors <- as.data.frame(get_factors(mofa_imputed)$group1)
  
  a <- fit_lmm_and_extract_pvals(latent_factors, metadata, c("Group", "Months", "Transplant_indication", "Age", "Gender",
                                                             "Total_ant_num", "GIT_ischemic_time", "Group_months_grouped",
                                                             "Group*Months + Transplant_indication + Total_ant_num + GIT_ischemic_time",
                                                             "Group + Months + Transplant_indication + Total_ant_num + GIT_ischemic_time")) 
  linear_model_list[[name]] <- a %>%
    mutate(drop_factor_threshold = drop_factor_threshold,
           Type = ifelse(grepl("+", variable, fixed = TRUE), ".", ".."),
           Significance = case_when(Adjusted_p_value >= 0.05 ~ "",
                                        Adjusted_p_value < 0.05 & Adjusted_p_value >= 0.01 ~ "*",
                                        Adjusted_p_value < 0.01 & Adjusted_p_value >= 0.001 ~ "**",
                                        Adjusted_p_value < 0.001 & Adjusted_p_value >= 0.0001 ~ "***",
                                        Adjusted_p_value < 0.0001 ~ "****"),
           variable = gsub("_", " ", variable),
           variable = str_squish(variable),
           Factor_order = as.numeric(gsub("Factor", "", Factor))) %>%
    arrange(Factor_order) %>%
    mutate(Factor = factor(Factor, levels = unique(Factor)))
}
saveRDS(linear_model_list, paste0(analysis_path, "biomarker-mofa/models/linear_model_list.rds"))
```

```{r}
for (i in 1:3) {
  
  drop_factor_threshold <- c(0.01, 0.02, 0.05)[i]
  name <- paste("rna-mol-bact", drop_factor_threshold, maxiter, convergence_mode, sep = "-")
  
  p_list[[name]] <-
  
    ggplot(linear_model_list[[name]], aes(y = variable, x = as.factor(Factor_order))) +
      geom_tile(aes(fill = estimate, colour = estimate, alpha = abs(estimate))) +
      geom_text(aes(label = Significance), vjust = 0.5, hjust = 0.5, size = 2) +
      theme_bw() + 
      theme(plot.subtitle = element_text(size = 8),
            axis.text.y = element_text(size = 8),
            strip.text = element_blank(),
            panel.grid.major = element_blank() ) +
      guides(alpha = "none") +
      labs(subtitle = paste("Drop factor threshold", drop_factor_threshold),
           colour = "LMM estimate",
           fill = "LMM estimate",
           y = "Variable", x = "Factor") +
      facet_grid(Type~., scales = "free", space = "free") +
      scale_fill_gradient2(low = "#56CFE1" , mid = "white", high = "#FF6348FF") +
      scale_colour_gradient2(low = "#56CFE1" , mid = "white", high = "#FF6348FF") 
}

p <- ggarrange(plotlist = p_list, common.legend = TRUE, legend = "bottom", align = "h")
p <- annotate_figure(p, top = text_grob("MOFA LMM Factor testing", face = "bold", size = 11), 
                     fig.lab = "Formula = Factor ~ Variable + Age + Gender + (1|Record.ID)",
                     fig.lab.pos = "bottom.right")
ggsave(paste0(analysis_path, "biomarker-mofa/lmm-testing-list.pdf"), p, width = 35, height = 17, units = 'cm')
```

```{r}
# For Factor 2
lmm_out_table <- do.call(rbind, linear_model_list) %>% 
  filter(Factor == "Factor2") %>% 
  mutate_if(is.numeric, round, digits = 4) %>%
  dplyr::select(Factor, variable, "Threshold" = drop_factor_threshold, "Estimate" = estimate, p.value, "adj_pval" = Adjusted_p_value, "Significance_adj_pval" = Significance) %>% 
  arrange(variable, Threshold) 
write.csv(lmm_out_table, "biomarker-mofa/models/out_table_factor2.csv", row.names = FALSE)

mofa_imputed <- mofa_list$`rna-mol-bact-0.05-10000-slow`
```

## Total variance

```{r}
# Total variance explained by omics
r2t <- reshape2::melt(mofa_imputed@cache$variance_explained$r2_total)

p <- ggplot(r2t, aes(fill=Var1, y=value, x=Var1)) +
  scale_fill_manual(values=c("#56CFE1", "#FF6348FF")) + 
  geom_bar(position='stack', stat='identity', col='black', cex=0.25, alpha = 0.9) +
  theme +
  theme(legend.position='none') +
  labs(x='Omics dataset', y='Variance explained (%)', title = "Omics views")

ggsave("biomarker-mofa/total-var.pdf", p, width = 6, height = 6, units = 'cm')

# Variance per factor
r2f <- reshape2::melt(mofa_imputed@cache$variance_explained$r2_per_factor[[1]]) %>% 
  mutate(Var1 = as.factor(as.numeric(gsub("Factor", "", Var1))))

p <-
  
  ggplot(r2f, aes(fill=Var2, y=value, x=Var1)) +
  scale_fill_manual(values=c("#56CFE1", "#FF6348FF")) + 
  geom_bar(position='stack', stat='identity', col='black', cex=0.25, alpha = 0.9) +
  theme +
  labs(x = 'Factor', 
       y = 'Variance explained (%)', 
       title = "MOFA omics factors",
       fill = "Omics")

ggsave("biomarker-mofa/var-per-factor.pdf", p, width = 12, height = 6, units = 'cm')
```

## Factors and covariates

```{r}
latent_factors <- data.frame(get_factors(mofa_imputed)$group1) %>% 
  rownames_to_column("Patient_days") %>%
  left_join(mofa_metadata)
```

```{r}
# Days
latent_factors <- data.frame(get_factors(mofa_imputed)$group1) %>% 
  rownames_to_column("Patient_days") %>%
  left_join(mofa_metadata) %>%
  arrange(Record.ID, Months)

  p <-
    
  ggplot(latent_factors, aes(x = Months, y = Factor2)) +
    geom_smooth(aes(group = Record.ID), method = "lm", formula = y ~ ns(x, 2), se = FALSE, colour = "lightgrey", linewidth = 0.3) +
    geom_smooth(aes(group = Group, colour = Group), method = "lm", formula = y ~ ns(x, 3), linewidth = 0.8) +
    geom_point(aes(fill = Group), shape = 21) +
    labs(title = "Factor 2", y = "Value", x = "Months post-transplant", colour = "Group", fill = "Group") +
    theme +
    theme(legend.position = "right") +
    plot_fill(Group = TRUE) +
    plot_colours(Group = TRUE)
  
  ggsave("biomarker-mofa/factor2-months-lm.pdf", p, width = 10, height = 7, units = 'cm')
```

## Weights

```{r}
fact_weights_list <- list()

for (i in c(1:2)){
  fact_weights_list[[i]] <- get_weights(mofa_imputed, as.data.frame = T, factors = i) %>%
    group_by(view) %>%
    mutate(sign = case_when(value > 0 ~ "+",
                          value <= 0 ~ "-"),
           Feature = feature,
           feature = gsub("_Proteins|_Transcripts", "", feature)) 
}

saveRDS(fact_weights_list, "biomarker-mofa/weights/factor_weights_list.rds")
```

```{r}
# Factor 2 features
p <- fact_weights_list[[2]]  %>%
  group_by(view) %>%
  slice_max(abs(value), n = 20) %>% 

ggplot(aes(x = value, y = reorder(feature, abs(value)))) +
  geom_col(width = 0.5, fill = "lightgrey") +
  geom_point(aes(fill = view), size = 2, shape = 21) +
  theme +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 7)) +
  labs(title = paste("Factor 2 features"),
       y = "Feature",
       x = "Weight value") +
  facet_grid(view~., scales = "free", space = "free") +
  scale_fill_manual(values = c("Transcripts" = "#56CFE1", "Proteins" = "#FF6348"))

  ggsave("biomarker-mofa/weights/factor2.pdf", p, width = 8, height = 15, units = 'cm')
```

```{r}
mofa_rna_features <- fact_weights_list[[2]] %>% filter(view == "Transcripts") %>% slice_max(abs(value), n = 50) %>% pull(Feature) %>% as.character()
mofa_molecules_features <- fact_weights_list[[2]] %>% filter(view == "Molecules") %>% slice_max(abs(value), n = 50) %>% pull(Feature) %>% as.character()

intersect(mofa_rna_features, ml_rna_features)
intersect(mofa_molecules_features, ml_molecules_features)
```

