#Load Packages

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
my_packages <- c("Seurat","SeuratWrappers","SeuratObject","SeuratDisk","scCustomize","Matrix","here","data.table","dplyr","readxl","scattermore","ggplot2","RColorBrewer","SingleCellExperiment","tidyverse","ggpubr","BPCells")
suppressMessages(lapply(my_packages, require, character.only = TRUE)) 
set.seed(9166)

library(readxl)    
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}

`%notin%` <- Negate(`%in%`)

options(Seurat.object.assay.version = "v3")

```

## Get stats per library

```{r}

sample.id <- readxl::read_xlsx(here("./Metadata.xlsx"))$Sample

#Stats <- fread(paste0("./Export_Files/11-LLB-Healthy_S2_L002/GeneFull_Ex50pAS/summary.csv")) %>% dplyr::pull("V1")

Stat_oi = "Median Reads per Cell"
Numbers <- c()
for (X in sample.id) {
  temp <- fread(here(paste0("./Export_Files/",X,"/GeneFull_Ex50pAS/Summary.csv"))) %>%
    column_to_rownames(var = "V1")
  temp <- temp[Stat_oi,] %>% as.numeric()
  Numbers <- c(Numbers,temp)
}

```

#SoupX Script

```{r}

library(celda)
library(DropletUtils)
library(SoupX)

# Define the base folder path and sub-folders
base_fp <- here("./Export_Files/")

# list each samples directory name
sample.id <- list.dirs(paste0(base_fp), recursive = F)
  
  # Loop through the different sequencing runs to perform the SoupX removal of ambient RNA
for (seq_id in sample.id) {
    message("Sample: ", which(sample.id == seq_id),"/",length(sample.id)," : ", seq_id)
    # Read in the filtered matrix (putative cells) and raw matrix (with empty droplets, a.k.a. "soup")
    # We get two objects of type dgCMatrix
    filt_matrix <- Read10X(data.dir = paste0(seq_id, '/GeneFull_Ex50pAS/', 'filtered'))
    raw_matrix <- Read10X(data.dir = paste0(seq_id, '/GeneFull_Ex50pAS/', 'raw'))
    # Create a Seurat object from the sparse matrix
    srat <- CreateSeuratObject(counts = filt_matrix)
    # Create the "SoupChannel" needed by SoupX
    soup_channel <- SoupChannel(raw_matrix, filt_matrix)
  
    # Create the clusters (in order to define marker genes) required by SoupX
    srat <- SCTransform(srat, verbose = FALSE)
    srat <- RunPCA(srat, verbose = FALSE)
    srat <- RunUMAP(srat, dims = 1:30, verbose = FALSE)
    srat <- FindNeighbors(srat, dims = 1:30, verbose = FALSE)
    srat <- FindClusters(srat, verbose = TRUE)
    # Add clusters to setClusters; setDR is useful for visualisation
    meta <- srat@meta.data
    umap <- srat@reductions$umap@cell.embeddings
    soup_channel <- setClusters(soup_channel, setNames(meta$seurat_clusters, rownames(meta)))
    soup_channel <- setDR(soup_channel, umap)
    # With defined clusters, run the main SoupX function, calculating ambient RNA profile
    # Genes with the highest background expression are typically those for ribosomal proteins
    soup_channel <- autoEstCont(soup_channel)
    # Use roundToInt to ensure you return an output integer matrix
    adj_matrix <- adjustCounts(soup_channel, roundToInt = TRUE)
    # Write the directory with corrected read counts
    DropletUtils::write10xCounts(paste0(seq_id, '/soupX_corrected'), adj_matrix)
}

```

#PreProcessing (Low-Pass Filtering + Object Assembly)

##STEP 1: Basic Pre-Filtering and QC Metric creation, Object creation

```{r}

library(DoubletFinder)
library(foreach)
library(doParallel)

my_packages <- c("Seurat","DoubletFinder","foreach","doParallel","dplyr","data.table","here","Matrix","scuttle","SingleCellExperiment")

Resolution = 1
min.pc2 = 50
RNA_Ambient = "soupX_corrected"

# list each samples directory name
SRR <- readxl::read_xlsx(here("./Metadata.xlsx"))$Sample
# Define the base folder path
base_fp <- here("Export_Files")

Metadata <- readxl::read_xlsx(here("./Metadata.xlsx"))

#DR = 0.0007665*RC

#Setup parallel processing
cores=4
cl <- makeCluster(cores)
registerDoParallel(cl)
writeLines(c(""), "log.txt")

comb <- function(x, ...) {
    lapply(seq_along(x),
    function(i) c(x[[i]], lapply(list(...), function(y) y[[i]])))
}

# Loop through the different sequencing runs, add metadata
List <- foreach(seq_id = SRR, .combine='comb', .multicombine = TRUE, .init=list(Seurat = list(), Pre_Stats = list()), .packages=my_packages, .verbose = T) %dopar% {
  set.seed(9166)
  cat(c("CYCLE_START:",seq_id,"\n"), file="log.txt", append=TRUE)
  #print(seq_id)
  #Read in soupX corrected values for each study, add metadata, append
  matrix <- Read10X(data.dir = here(base_fp, paste0(seq_id), RNA_Ambient), gene.column = 2)
  srat <- CreateSeuratObject(counts = matrix
                             #, min.cells = 3
                             )
  #Get spliced and unspliced counts per cell and % (low % intronic = apoptotic cell)
  spliced <- ReadMtx(mtx = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','spliced.mtx.gz'),
                     cells = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','barcodes.tsv.gz'),
                     features = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','features.tsv.gz'))
  spliced_counts <- colSums(spliced) %>% as.numeric()
  unspliced <- ReadMtx(mtx = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','unspliced.mtx.gz'),
                       cells = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','barcodes.tsv.gz'),
                       features = here(base_fp, paste0(seq_id),'Velocyto', 'filtered','features.tsv.gz'))
  unspliced_counts <- colSums(unspliced) %>% as.numeric()
  percent_unspliced = (unspliced_counts/(spliced_counts+unspliced_counts))*100
  rm(spliced, unspliced, matrix)
  gc()
  #Append QC Metrics to Use
  srat[["total_unspliced"]] <- unspliced_counts
  srat[["total_spliced"]] <- spliced_counts
  srat[["percent.unspliced"]] <- percent_unspliced
  srat[["percent.mt"]] <- PercentageFeatureSet(object = srat, pattern = "^MT-")
  srat[["percent.rb"]] <- PercentageFeatureSet(object = srat, pattern = "^RPS|^RPL")
  srat[["percent.hb"]] <- PercentageFeatureSet(object = srat, pattern = "^HB[^(P)]")
  srat[['logit_rb']] <- qlogis(((srat$nCount_RNA * srat$percent.rb / 100) + 1)/(srat$nCount_RNA + 2))
  srat[['logit_hb']] <- qlogis(((srat$nCount_RNA * srat$percent.hb / 100) + 1)/(srat$nCount_RNA + 2))
  srat[["log10GenesPerUMI"]] <- log10(srat$nFeature_RNA) / log10(srat$nCount_RNA) #Complexity
  #Do basic filtering prior to doublet detection
  
  sc_srat <- as.SingleCellExperiment(srat)
  is.mito <- grep("MT-", rownames(sc_srat))
  per.cell <- perCellQCMetrics(sc_srat, subsets=list(Mito=is.mito))
  per.cell$unspliced_percent <- srat$percent.unspliced
  per.cell$Complexity <- srat$log10GenesPerUMI
  
  low.total <- isOutlier(per.cell$sum, type="lower", log=TRUE)
  Thresh <- attr(low.total, "threshold")[1]
  qc.lib <- srat@meta.data$nCount_RNA > Thresh 
  srat[["Scuttle_nCount_RNA"]] <- as.numeric(Thresh)
  low.total <- isOutlier(per.cell$detected, type="lower", log=TRUE)
  Thresh <- attr(low.total, "threshold")[1]
  qc.nexprs <- srat@meta.data$nFeature_RNA > Thresh 
  srat[["Scuttle_nFeature_RNA"]] <- as.numeric(Thresh)
  low.total <- isOutlier(per.cell$Complexity, type="lower", log=TRUE)
  Thresh <- attr(low.total, "threshold")[1]
  qc.complex <- srat@meta.data$log10GenesPerUMI > Thresh
  srat[["Scuttle_Complexity"]] <- as.numeric(Thresh)
  low.total <- isOutlier(per.cell$subsets_Mito_percent, type="higher", log=FALSE)
  Thresh <- attr(low.total, "threshold")[2]
  qc.mito <- srat@meta.data$percent.mt < Thresh
  srat[["Scuttle_percent.mt"]] <- as.numeric(Thresh)
  low.total <- isOutlier(per.cell$unspliced_percent, type="lower", log=FALSE)
  Thresh <- attr(low.total, "threshold")[1]
  srat[["Scuttle_unspliced_percent_low"]] <- as.numeric(Thresh)
  low.total <- isOutlier(per.cell$unspliced_percent, type="higher", log=FALSE)
  Thresh <- attr(low.total, "threshold")[2]
  qc.unspliced <- srat@meta.data$percent.unspliced < Thresh
  srat[["Scuttle_unspliced_percent_high"]] <- as.numeric(Thresh)
  
  qc.lib <- srat@meta.data$nCount_RNA > 1000
  qc.nexprs <- srat@meta.data$nFeature_RNA > 300
  qc.mito <- srat@meta.data$percent.mt < 50
  qc.complex <- srat@meta.data$log10GenesPerUMI > 0.8
  
  discardAll <- qc.lib & qc.nexprs & qc.mito & qc.complex
  #table(discardAll)
  
  #Append Other Metadata
  srat[["Sample_Align"]] <- Metadata[Metadata$Sample == seq_id,]$Sample
  srat[["Disease"]] <- Metadata[Metadata$Sample == seq_id,]$Disease
  
  #Get stats
  stats_pre <- srat@meta.data
  
  #Filter
  srat$discardAll = discardAll
  srat <- subset(x = srat, subset = discardAll == "TRUE")
  rm(sc_srat)
  gc()
  
  #End, write
  cat(c("CYCLE_END:",seq_id,"\n"), file="log.txt", append=TRUE)
  list(seq_id = srat, seq_id = stats_pre)
}
stopCluster(cl)

names(List[[1]]) <- SRR %>% gsub("-","_",.)
names(List[[2]]) <- SRR %>% gsub("-","_",.)

rm(list = setdiff(ls(),c("List","`%notin%`","RNA_Ambient","Resolution")))

saveRDS(List[[2]],here(paste0("./Processed_Data/CLAD_Human_",RNA_Ambient,"_",Resolution,"_Step1_Pre_Stats.rds")))

List <- Merge_Seurat_List(list_seurat = List[[1]], project = "CLAD", add.cell.ids = names(List[[1]]))

saveRDS(List,here(paste0("./Processed_Data/CLAD_Human_",RNA_Ambient,"_",Resolution,"_Step1.rds")))

```

### QC Plots

```{r}

QC_Data <- readRDS(here("./Processed_Data/CLAD_Human_soupX_corrected_1_Step1_Pre_Stats.rds"))
QC_Data <- do.call(rbind, QC_Data)

QC_metric <- c("nCount_RNA","nFeature_RNA","percent.mt","log10GenesPerUMI")
names(QC_metric) <- c("Scuttle_nCount_RNA","Scuttle_nFeature_RNA","Scuttle_percent.mt","Scuttle_Complexity")
Manual_QC <- c(3000, 1000, 25, 0.8)

w = 12
h = 8
c = 4

Plots <- list()
for (i in c(1:4)) {
  
  Name <- as.character(QC_metric)[i]
  QC_Data$Plot <- QC_Data[,paste0(Name)]
  scuttle_data <- data.frame(Sample_Align = unique(QC_Data$Sample_Align), 
                             data = unique(QC_Data[,paste0(names(QC_metric)[i])]))
  
  g <- ggplot(QC_Data, aes(x=Plot)) +
    geom_histogram(fill = "#58508d", color = "black", linewidth = 0.2, bins = 50) +
    geom_vline(xintercept = Manual_QC[i], color = "#005a74", linetype = "dashed", linewidth = 0.75) +
    geom_vline(data = scuttle_data, aes(xintercept = data), color = "#ff2323", linetype = "dashed", linewidth = 0.75) +
    scale_y_continuous(expand = c(0,0)) + 
    theme_minimal() +
    theme(axis.text = element_text(size = 12, colour = "black"),
          axis.text.x = element_text(angle = 90, vjust = 0.5),
          axis.title = element_text(size = 16, colour = "black", face = "bold"),
          panel.border = element_rect(color = "black", linewidth = 1.5, fill = NA)) +
    facet_wrap(~Sample_Align, ncol = c) +
    labs(x = paste0(Name), y = "Count")
  
  if (i %in% c(1,2)) {
    g <- g + scale_x_continuous(expand = c(0,0), trans = "log10")
  } else {
    g <- g + scale_x_continuous(expand = c(0,0))
  }
  Plots[[paste0(i)]] <- g
  
  #ggsave(paste0("./Downstream_analysis/QC_Plots/Pre_",as.character(QC_metric)[i],".pdf"), g, width = 7, height = 4, dpi = 300, units = "in")
  ggsave(here(paste0("./Downstream_Analysis/QC_Plots/Pre_",as.character(QC_metric)[i],".tiff")), g, width = w, height = h, dpi = 300, units = "in")
  
}

#Plots[["1"]]
#table(QC_Data$Match)
#nCount_RNA vs nFeature_RNA + nCount_RNA vs percent.mt:

g <- ggplot(QC_Data, aes(x=nCount_RNA, y = nFeature_RNA)) +
    geom_scattermore() + 
    scale_y_continuous(expand = c(0,0)) + scale_x_continuous(expand = c(0,0), trans = "log10") + 
    theme_minimal() +
    theme(axis.text = element_text(size = 12, colour = "black"),
          axis.text.x = element_text(angle = 90, vjust = 0.5),
          axis.title = element_text(size = 16, colour = "black", face = "bold"),
          panel.border = element_rect(color = "black", linewidth = 1.5, fill = NA)) +
    facet_wrap(~Sample_Align, ncol = c)

ggsave(here(paste0("./Downstream_Analysis/QC_Plots/Pre_nCount_nFeature.tiff")), g, width = w, height = h, dpi = 300, units = "in")

g <- ggplot(QC_Data, aes(x=nCount_RNA, y = percent.mt)) +
    geom_scattermore() + 
    scale_y_continuous(expand = c(0,0)) + scale_x_continuous(expand = c(0,0), trans = "log10") + 
    theme_minimal() +
    theme(axis.text = element_text(size = 12, colour = "black"),
          axis.text.x = element_text(angle = 90, vjust = 0.5),
          axis.title = element_text(size = 16, colour = "black", face = "bold"),
          panel.border = element_rect(color = "black", linewidth = 1.5, fill = NA)) +
    facet_wrap(~Sample_Align, ncol = c)

ggsave(here(paste0("./Downstream_Analysis/QC_Plots/Pre_nCount_mt.tiff")), g, width = w, height = h, dpi = 300, units = "in")

```

# HLCA Setup

```{r}

#prep for HLCA:

Merged_Temp <- readRDS("./Processed_Data/CLAD_Human_soupX_corrected_1_Step1.rds")
Merged_Temp[["RNA"]] <- JoinLayers(Merged_Temp[["RNA"]])
genes <- fread("./Export_Files/21_361_DT/GeneFull_Ex50pAS/filtered/features.tsv.gz", header = F)

counts <- GetAssayData(Merged_Temp, layer="counts", assay="RNA")
rownames(counts) <- genes$V1

HLCA_genes <- fread(here("./HLCA_reference_model_gene_order_ids_and_symbols.csv"))
HLCA_keep <- counts@Dimnames[[1]] %in% HLCA_genes$gene_id
counts <- counts[HLCA_keep,]
Merged_Temp <- CreateSeuratObject(counts=counts, meta.data = Merged_Temp@meta.data)
rm(counts)
saveRDS(Merged_Temp,"./Processed_Data/CLAD_Step1_Post_Filt_HLCA.rds")
Merged_Temp <- readRDS("./Processed_Data/CLAD_Step1_Post_Filt_HLCA.rds")

#Create .h5ad file (ensure you have an appropriate python venv set up)

library(reticulate)
#conda_list()
use_condaenv("CellRank")
library(anndata)

adata <- AnnData(X = t(as.matrix(Merged_Temp@assays$RNA$counts)),
                 obs = Merged_Temp@meta.data
                )

write_h5ad(adata, paste0("./Processed_Data/CLAD_HLCA_Input.h5ad"))

```

## STEP 2.1: Filter Basic

```{r}

rm(list=ls())

Merged_Temp <- readRDS("./Processed_Data/CLAD_Human_soupX_corrected_1_Step1.rds")
Merged_Temp[["RNA"]] <- JoinLayers(Merged_Temp[["RNA"]])
Merged_Temp <- subset(Merged_Temp, subset = nFeature_RNA > 1000 & nCount_RNA > 3000 & percent.mt < 25 & log10GenesPerUMI > 0.8)

saveRDS(Merged_Temp,"./Processed_Data/CLAD_Human_soupX_corrected_1_Step2pre.rds")

```

## STEP 2.2: Doublet Filtering

```{r}

library(DoubletFinder)
library(foreach)
library(doParallel)

comb <- function(x, ...) {
  lapply(seq_along(x),
    function(i) c(x[[i]], lapply(list(...), function(y) y[[i]])))
}

SRR <- readxl::read_xlsx(here("./Metadata.xlsx"))$Sample

cores=2
cl <- makeCluster(cores)
registerDoParallel(cl)
writeLines(c(""), "log.txt")
min.pc2 = 50
Resolution = 1

my_packages <- c("Seurat","DoubletFinder","foreach","doParallel","dplyr","data.table","here","Matrix","scuttle","SingleCellExperiment")

#DR = 0.0007665*RC
List <- foreach(seq_id = SRR, .combine='comb', .packages=my_packages, .multicombine = TRUE, .init=list(Seurat = list(), bcmvn_df = list()), .verbose = T) %dopar% {
  set.seed(9166)
  Merged_Temp <- readRDS("./Processed_Data/CLAD_Human_soupX_corrected_1_Step2pre.rds")
  options(Seurat.object.assay.version = "v3")
  srat = subset(Merged_Temp, subset = Sample_Align == paste0(seq_id))
  rm(Merged_Temp)
  cat(c("CYCLE_DOUBLET_BEGIN:",seq_id,"\n"), file="log.txt", append=TRUE)
  #Doublet Calculations
  #Doublet.temp = NormalizeData(srat, normalization.method = "LogNormalize", scale.factor = 10000)
  #Doublet.temp = FindVariableFeatures(Doublet.temp, verbose = F)
  #Doublet.temp = ScaleData(Doublet.temp, vars.to.regress = c("nFeature_RNA", "percent.mt"), verbose = F)
  #Doublet.temp = RunPCA(Doublet.temp, verbose = F, npcs = 50)
  Doublet.temp <- SCTransform(srat)
  Doublet.temp <- RunPCA(Doublet.temp)
  #Find N signficant PCs
  stdv <- Doublet.temp[["pca"]]@stdev
  sum.stdv <- sum(Doublet.temp[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] - 
                       percent.stdv[2:length(percent.stdv)]) > 0.1), 
              decreasing = T)[1] + 1
  min.pc <- min(co1, co2)
  srat[["min_pc"]] <- min.pc
  min.pc = min.pc2
  cat(c("Continue:",seq_id,"\n"), file="log.txt", append=TRUE)
  #Continue
  Doublet.temp = RunUMAP(Doublet.temp, dims = 1:min.pc, verbose = F)
  Doublet.temp <- FindNeighbors(object = Doublet.temp, dims = 1:min.pc)              
  Doublet.temp <- FindClusters(object = Doublet.temp, resolution = Resolution)
  #Param Optimise
  sweep.res <- paramSweep(seu = Doublet.temp, PCs = 1:min.pc, sct=TRUE) 
  sweep.stats <- summarizeSweep(sweep.res, GT = FALSE) 
  bcmvn <- find.pK(sweep.stats)
  # Optimal pK is the max of the bimodality coefficent (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- bcmvn.max$pK
  optimal.pk <- as.numeric(levels(optimal.pk))[optimal.pk]
  #Linear dependency for retrieved cells (estimated) vs doublet rate, calculate
  nExp <- round(ncol(srat) * (0.0007665*ncol(srat)/100))
  #Homotypic doublet proportion estimate
  annotations <- Doublet.temp@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.adj <- round(nExp * (1 - homotypic.prop))
  Doublet.temp <- doubletFinder(Doublet.temp, PCs = 1:min.pc, pK = optimal.pk, nExp = nExp.adj, sct = TRUE)
  index <- length(colnames(Doublet.temp@meta.data))
  #srat[["Doublets_optimalpk"]] <- optimal.pk
  #srat[["Doublets_MeanBC"]] <- bcmvn.max$MeanBC
  #srat[["Doublets_VarBC"]] <- bcmvn.max$VarBC
  #srat[["Doublets_BCmetric"]] <- bcmvn.max$BCmetric
  #srat[["Doublets_co1"]] <- co1
  #srat[["Doublets_co2"]] <- co2
  #srat[["Doublets_nExp"]] <- nExp
  #srat[["Doublets_nExp.adj"]] <- nExp.adj
  #srat[["homotypic_prop"]] <- homotypic.prop
  srat[["Doublets"]] <- Doublet.temp@meta.data[,index]
  srat[["Doublets_pANN"]] <- Doublet.temp@meta.data[,index-1]
  #End
  cat(c("CYCLE_END:",seq_id,"\n"), file="log.txt", append=TRUE)
  list(seq_id = srat, seq_id = bcmvn)
}
stopCluster(cl)

saveRDS(List[[2]], "./Processed_Data/CLAD_Human_soupX_corrected_1_Step2_dd_1_50_bcmvn.rds")
List <- Merge_Seurat_List(list_seurat = List[[1]], project = "CLAD")
saveRDS(List, "./Processed_Data/CLAD_Human_soupX_corrected_1_Step2_dd_1_50.rds")

```

##STEP 2.3: Do Basic Clustering

```{r}

rm(list=ls())

Merged_Temp <- readRDS("./Processed_Data/CLAD_Human_soupX_corrected_1_Step2_dd_1_50.rds")
Merged_Temp[["RNA"]] <- JoinLayers(Merged_Temp[["RNA"]])

Merged_Temp <- subset(Merged_Temp, subset = Doublets == "Singlet")

counts <- GetAssayData(Merged_Temp, layer="counts", assay="RNA")
genes.filter <- rowSums(counts > 0)
genes.filter <- genes.filter > 10
counts <- counts[genes.filter,]
metadata <- Merged_Temp@meta.data

#Merged_Temp <- CreateSeuratObject(counts=counts, meta.data = metadata)
rm(list = setdiff(ls(),c("metadata","counts")))
gc()

Merged_Temp <- CreateSeuratObject(counts = counts, meta.data = metadata)

rm(counts)
rm(metadata)

Merged_Temp <- SCTransform(Merged_Temp, assay = "RNA", variable.features.n = 5000)
Merged_Temp <- Merged_Temp %>% RunPCA(npcs = 50) %>% RunUMAP(dims = 1:50) %>% FindNeighbors(dims = 1:50) %>% FindClusters(resolution = seq(1,10,1))

saveRDS(Merged_Temp,paste0("./Processed_Data/CLAD_Post_Filter1_soupX_corrected_1_Step2_3.rds"))

```

# Add HLCA Annotations

```{r}

Merged_Temp <- readRDS("./Processed_Data/CLAD_Post_Filter1_soupX_corrected_1_Step2_3.rds")

HLCA_Meta <- fread("./Processed_Data/CLAD_Comb_Metadata.csv") %>% column_to_rownames(var = "V1")
HLCA_Meta <- HLCA_Meta[HLCA_Meta$ref_or_query == "query",]
rownames(HLCA_Meta) <- paste0("_",rownames(HLCA_Meta))
HLCA_Meta <- HLCA_Meta[colnames(Merged_Temp),]

Merged_Temp$HLCA_Lvl1 <- HLCA_Meta$ann_level_1_transferred_label_unfiltered
Merged_Temp$HLCA_Lvl2 <- HLCA_Meta$ann_level_2_transferred_label_unfiltered
Merged_Temp$HLCA_Lvl3 <- HLCA_Meta$ann_level_3_transferred_label_unfiltered
Merged_Temp$HLCA_Lvl4 <- HLCA_Meta$ann_level_4_transferred_label_unfiltered
Merged_Temp$HLCA_Lvl5 <- HLCA_Meta$ann_level_5_transferred_label_unfiltered

Merged_Temp$HLCA_Lvl1_Filt <- HLCA_Meta$ann_level_1_transferred_label
Merged_Temp$HLCA_Lvl2_Filt <- HLCA_Meta$ann_level_2_transferred_label
Merged_Temp$HLCA_Lvl3_Filt <- HLCA_Meta$ann_level_3_transferred_label
Merged_Temp$HLCA_Lvl4_Filt <- HLCA_Meta$ann_level_4_transferred_label
Merged_Temp$HLCA_Lvl5_Filt <- HLCA_Meta$ann_level_5_transferred_label

Merged_Temp$HLCA_Lvl1_Uncert <- HLCA_Meta$ann_level_1_transfer_uncert
Merged_Temp$HLCA_Lvl2_Uncert <- HLCA_Meta$ann_level_2_transfer_uncert
Merged_Temp$HLCA_Lvl3_Uncert <- HLCA_Meta$ann_level_3_transfer_uncert
Merged_Temp$HLCA_Lvl4_Uncert <- HLCA_Meta$ann_level_4_transfer_uncert
Merged_Temp$HLCA_Lvl5_Uncert <- HLCA_Meta$ann_level_5_transfer_uncert

```

## STEP 3: Batch Correction

```{r}

rm(list = setdiff(ls(),"Merged_Temp"))

# Split datasets and process prior to integration
DefaultAssay(Merged_Temp) <- 'RNA'
Merged_Temp <- DietSeurat(Merged_Temp, assays = 'RNA')

# Identify genes with useful names
gene_names <- rownames(Merged_Temp)
genes_to_keep <- grep("^(ENSG|LINC)", gene_names, invert = TRUE, value = TRUE)
Merged_Temp <- subset(Merged_Temp, features = genes_to_keep)

# Split by sample, and run pre-processing steps
Merged_Temp[['RNA']] <- split(Merged_Temp[['RNA']], f = Merged_Temp$Sample_Align)
Merged_Temp <- Merged_Temp %>% 
  SCTransform(assay = "RNA", variable.features.n = 5000) %>% 
  RunPCA(dims = 1:50) %>% RunUMAP(dims = 1:50)

# Integrate datasets using CCA calculations
options(future.globals.maxSize = 8000 * 1024^2)
Merged_Temp <- IntegrateLayers(Merged_Temp,
                               method = CCAIntegration,
                               normalization.method = 'SCT',
                               k.anchor = 5,
                               new.reduction = 'integrated_cca')

#UMAP
Merged_Temp <- RunUMAP(Merged_Temp, reduction = "integrated_cca", reduction.name = "umap_cca", dims = c(1:50))

# Find neighbours and clusters
Merged_Temp <- FindNeighbors(Merged_Temp, reduction = 'integrated_cca', dims = 1:50)
Merged_Temp <- FindClusters(Merged_Temp, resolution = seq(1, 10, 1))

# Plot
DimPlot(Merged_Temp, reduction = "umap_cca", label = T, group.by = "SCT_snn_res.10") + NoLegend()
DimPlot(Merged_Temp, reduction = "umap_cca", label = T, group.by = "HLCA_Lvl4", repel = T) + NoLegend()

saveRDS(Merged_Temp, "./Processed_Data/CLAD_Step3pre.rds")

```

# Assign manual annotation

```{r}

Merged_Temp <- readRDS("./Processed_Data/CLAD_Step3pre.rds")

HLCA_Assign <- table(Merged_Temp$SCT_snn_res.10, Merged_Temp$HLCA_Lvl5) %>% as.data.frame() %>% pivot_wider(., names_from = "Var2", values_from = "Freq") %>% column_to_rownames("Var1") %>% t()

Idents(object = Merged_Temp) <- "SCT_snn_res.10"
new.cluster.ids <- case_when(
    Merged_Temp$SCT_snn_res.10 %in% c(100,105,77,72,30,128,56,88,123) ~ "ATI",
    Merged_Temp$SCT_snn_res.10 %in% c(115,117,48,6,79,49,66,16,10,0,38,47,101) ~ "ATII",
    Merged_Temp$SCT_snn_res.10 %in% c(95) ~ "Club Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(114) ~ "Basal Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(71,97) ~ "Ciliated Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(67,113) ~ "Mast Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(96,82,39) ~ "Venous",
    Merged_Temp$SCT_snn_res.10 %in% c(116) ~ "Systemic Venous",
    Merged_Temp$SCT_snn_res.10 %in% c(1,31,21,2,75) ~ "gCap",
    Merged_Temp$SCT_snn_res.10 %in% c(23,93,32,44) ~ "aCap",
    Merged_Temp$SCT_snn_res.10 %in% c(68,53,106,81) ~ "Arterial",
    Merged_Temp$SCT_snn_res.10 %in% c(132,60,36) ~ "Lymphatics",
    Merged_Temp$SCT_snn_res.10 %in% c(87) ~ "Pericytes",
    Merged_Temp$SCT_snn_res.10 %in% c(42,91) ~ "SMCs",
    Merged_Temp$SCT_snn_res.10 %in% c(104,41) ~ "Alv. Fibr.",
    Merged_Temp$SCT_snn_res.10 %in% c(34,70,124,129) ~ "Adv. Fibr.",
    Merged_Temp$SCT_snn_res.10 %in% c(18,59,28,74,25,58,69,86) ~ "NK Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(76,13,35,51,57,33,112,99,24,19,7) ~ "CD8+ T-cells",
    Merged_Temp$SCT_snn_res.10 %in% c(125,22,110,73,63,14,65) ~ "CD4+ T-cells",
    Merged_Temp$SCT_snn_res.10 %in% c(17,54,107,62,5,85) ~ "cMono",
    Merged_Temp$SCT_snn_res.10 %in% c(11,3,12,61,29,90,27,15) ~ "ncMono",
    Merged_Temp$SCT_snn_res.10 %in% c(43) ~ "DCs",
    Merged_Temp$SCT_snn_res.10 %in% c(45,89,108) ~ "Int-Mφ",
    Merged_Temp$SCT_snn_res.10 %in% c(46,109,26,120,94,80,8,50,55,9,119) ~ "Alv-Mφ",
    Merged_Temp$SCT_snn_res.10 %in% c(78,52,64,20,4,37,40) ~ "Mo-AMφ",
    Merged_Temp$SCT_snn_res.10 %in% c(130,118,126) ~ "Plasma Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(102,83,131,92) ~ "B Cells",
    Merged_Temp$SCT_snn_res.10 %in% c(111,135,121,84,122,98,133,134,103,127) ~ "JUNK"
)
Merged_Temp$Manual_Anno <- factor(new.cluster.ids)

Merged_Temp <- subset(Merged_Temp, subset = Manual_Anno != "JUNK")

```

## STEP 4: Re-normalise + Correct

```{r}

DefaultAssay(Merged_Temp) <- 'RNA'
Merged_Temp <- DietSeurat(Merged_Temp, assays = 'RNA')

Merged_Temp <- Merged_Temp %>% 
  SCTransform(assay = "RNA", variable.features.n = 5000) %>% 
  RunPCA(dims = 1:50) %>% RunUMAP(dims = 1:50)

options(future.globals.maxSize = 8000 * 1024^2)
Merged_Temp <- IntegrateLayers(Merged_Temp, method = CCAIntegration, normalization.method = 'SCT', k.anchor = 5, new.reduction = 'integrated_cca')

Merged_Temp <- RunUMAP(Merged_Temp, reduction = "integrated_cca", reduction.name = "umap_cca", dims = c(1:50))

Merged_Temp <- FindNeighbors(Merged_Temp, reduction = 'integrated_cca', dims = 1:50)
Merged_Temp <- FindClusters(Merged_Temp, resolution = seq(1, 10, 1))

Merged_Temp <- subset(Merged_Temp, subset = SCT_snn_res.10 %notin% c("120"))

DefaultAssay(Merged_Temp) <- 'RNA'
Merged_Temp <- DietSeurat(Merged_Temp, assays = 'RNA')

Merged_Temp <- Merged_Temp %>% 
  SCTransform(assay = "RNA", variable.features.n = 5000) %>% 
  RunPCA(dims = 1:50) %>% RunUMAP(dims = 1:50)

options(future.globals.maxSize = 8000 * 1024^2)
Merged_Temp <- IntegrateLayers(Merged_Temp, method = CCAIntegration, normalization.method = 'SCT', k.anchor = 5, new.reduction = 'integrated_cca')

Merged_Temp <- RunUMAP(Merged_Temp, reduction = "integrated_cca", reduction.name = "umap_cca", dims = c(1:50))

Merged_Temp <- FindNeighbors(Merged_Temp, reduction = 'integrated_cca', dims = 1:50)
Merged_Temp <- FindClusters(Merged_Temp, resolution = seq(1, 10, 1))

```

### Fix Annotations

```{r}

Merged_Temp$Manual_Anno2 <- as.character(Merged_Temp$Manual_Anno)
Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(67,54),"gCap", Merged_Temp$Manual_Anno2)

Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(12,43,10,47,35,36,57),"Alv-Mφ", Merged_Temp$Manual_Anno2)
Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(1,48,20),"cMono", Merged_Temp$Manual_Anno2)
Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(32,22,28,9,61,26),"ncMono", Merged_Temp$Manual_Anno2)
Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(50,46),"DCs", Merged_Temp$Manual_Anno2)
Merged_Temp$Manual_Anno2 <- ifelse(Merged_Temp$SCT_snn_res.5 %in% c(41,62),"Int-Mφ", Merged_Temp$Manual_Anno2)

Merged_Temp$Manual_Anno2 <- factor(Merged_Temp$Manual_Anno2, levels = 
                                     c("aCap","gCap","Arterial","Venous","Systemic Venous","Lymphatics",
                                       "Pericytes","SMCs","Adv. Fibr.", "Alv. Fibr.",
                                       "ATII","ATI","Basal Cells","Club Cells","Ciliated Cells",
                                       "Mast Cells","B Cells","Plasma Cells",
                                       "NK Cells","CD8+ T-cells","CD4+ T-cells",
                                       "cMono","ncMono","DCs","Int-Mφ","Mo-AMφ","Alv-Mφ"))

```

# Score Giulia's Signature

```{r}

library(readxl)    
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}

DE_Features <- read_excel_allsheets("./de_features.xlsx")
DE_Features <- DE_Features[[4]]

Markers <- DE_Features$Gene

Merged_Temp <- PrepSCTFindMarkers(Merged_Temp)

Merged_Temp <- AddModuleScore(Merged_Temp, features = list(Markers), name = "Signature", nbin = 25)

saveRDS(Merged_Temp, "Processed_Data/CLAD_Step4Done.rds")
Merged_Temp <- readRDS("Processed_Data/CLAD_Step4Done.rds")

#DimPlot(Merged_Temp, group.by = "Manual_Anno", label = T, reduction = "umap_cca") + NoLegend()

```

## UMAP 

```{r}
UMAP <- Merged_Temp@reductions$umap_cca@cell.embeddings %>% as.data.frame()
colnames(UMAP) <- c("umap_1","umap_2")
UMAP$Sample <- Merged_Temp$Sample_Align
UMAP$Disease <- Merged_Temp$Disease
UMAP$Annotation <- Merged_Temp$Manual_Anno2

UMAP2 <- UMAP %>% group_by(Annotation) %>% dplyr::select(umap_1, umap_2) %>% summarize_all(median)
arr <- list(x = min(UMAP$umap_1), y = min(UMAP$umap_2), x_len = (max(UMAP$umap_1)-min(UMAP$umap_1))*0.2, y_len = (max(UMAP$umap_2)-min(UMAP$umap_2))*0.2)

Cells <- c("aCap","gCap","Arterial","Venous","Systemic Venous","Lymphatics",
                                       "Pericytes","SMCs","Adv. Fibr.", "Alv. Fibr.",
                                       "ATII","ATI","Basal Cells","Club Cells","Ciliated Cells",
                                       "Mast Cells","B Cells","Plasma Cells",
                                       "NK Cells","CD8+ T-cells","CD4+ T-cells",
                                       "cMono","ncMono","DCs","Int-Mφ","Mo-AMφ","Alv-Mφ")

Cells_Colors <- c("#5276c9","#1c5ed9","#768bb5","#6692d9","#2674c7","grey75",
                  "#6dee6f","#23a16d","#78e9ab","#3cb049",
                  "#508991","#00d8e1","#6f85c7","#00bfe8","#5aeece",
                  "#637eb3","#aecf8e","#4f6e31",
                  "#d26c67","#7f241b","#b1634d",
                  "#6f64c7","#bd81f5","#d9b4f0","#8678f0","#8953c2","#a456a0")

p <- 
  
  ggplot(UMAP, aes(x=umap_1, y=umap_2, fill=Annotation)) +
  geom_point(size = 1.75, alpha = 0.8, shape=21, stroke = 0.3, color="grey5") +
  theme_minimal() +
  theme(panel.grid = element_blank(), 
        axis.text = element_blank(),
        axis.title = element_blank(),
        title = element_text(size = 20, face = "bold", hjust = 0.5),
        legend.position = "none") + 
  scale_fill_manual(values = Cells_Colors) + 
  labs(title = "Khatri et al. 2023", x = "UMAP 1", y = "UMAP 2") + 
  ggrepel::geom_label_repel(data = UMAP2, aes(label = Annotation), seed = 1234, max.overlaps = Inf, size = 9, 
                            segment.size = 1, segment.curvature = -0.1, segment.ncp = 3, segment.angle = 20, box.padding = 1) +
  annotate("segment", 
           x = arr$x, xend = arr$x + c(arr$x_len, 0), 
           y = arr$y, yend = arr$y + c(0, arr$y_len), 
           arrow = arrow(type = "closed", length = unit(10, 'pt')), linewidth = 1.5, lineend = c("round","round")) +
  annotate("text", x = min(UMAP$umap_1)+2.5, y = min(UMAP$umap_2)-1, label = "UMAP 1", size = 6) +
  annotate("text", x = min(UMAP$umap_1)-1, y = min(UMAP$umap_2)+2.5, label = "UMAP 2", size = 6, angle = 90)

ggsave("Figure_Plots/UMAP_Full.png", p, width = 13, height = 8, dpi = 400, units = "in")
```

## Plot Signature

```{r}
p1 <- FeaturePlot(Merged_Temp, features = "Signature1", label = F, repel = F, pt.size = 1, reduction = "umap_cca") +
    scale_colour_gradientn(colours = c("#768bb5","#f7f7f7","#bd81f5"), 
                           values = c(0,0.25,1), limits = c(-0.1,0.31), oob=scales::squish) + theme(axis.text = element_blank(),
              axis.line = element_blank(),
              axis.ticks = element_blank(),
              axis.title = element_blank()) +
  labs(title = "", color = "Signature\nScore")

ggsave("./Downstream_analysis/Figure_Plots/UMAP_Sig_Score.png", p1, width = 9, height = 9, dpi = 400)
```

### Ridge Plots

```{r}
p <- 
  
  RidgePlot(Merged_Temp, group.by = "Manual_Anno2", features = "Signature1") + 
  NoLegend() +
  scale_fill_manual(values = Cells_Colors) +
  geom_vline(xintercept = 0, lty = "dashed") +
  labs(x = "Score", y = "Cell", title = "Transcriptomics signature") +
  theme(axis.text = element_text(face = "bold"),
        axis.title.x = element_text(hjust = 0.5),
        axis.title.y = element_text(hjust = 0.5)) 

ggsave("Figure_Plots/Sig_Score_Ridge.png", p, width = 5, height = 6, dpi = 300)
```

## Plot Signature Sub-Clusters Heatmap

```{r}

#Idents(Merged_Temp) <- "Manual_Anno"
#B <- DotPlot(Merged_Temp, features = rownames(Merged_Temp)[rownames(Merged_Temp) %in% DE_Features[DE_Features$Endothelium_glycocalyx == TRUE,]$Gene], split.by = "Disease")
#C <- B[["data"]]
#D <- pivot_wider(C[,c(3:5)], values_from = "avg.exp.scaled", names_from = "id") #%>% column_to_rownames(var = "features.plot")
#D <- na.omit(D)

Idents(Merged_Temp) <- "Manual_Anno2"
B <- DotPlot(Merged_Temp, features = DE_Features$Gene, scale = T)
C <- B[["data"]]
saveRDS(C,"Signature_Cells_Expression_Stats_Pre.rds")
C <- readRDS("Signature_Cells_Expression_Stats_Pre.rds")

#Genes_Keep <- C %>% dplyr::group_by(features.plot) %>% dplyr::slice_max(pct.exp, n=1) %>% 
#  dplyr::filter(pct.exp > 10) %>% dplyr::pull(features.plot) %>% droplevels() %>% as.character()
#C <- C[C$features.plot %in% Genes_Keep,]
#saveRDS(C,"Signature_Cells_Expression_Stats.rds")
#C <- readRDS("Signature_Cells_Expression_Stats.rds")

D <- pivot_wider(C[,c(3:5)], values_from = "avg.exp.scaled", names_from = "id") #%>% column_to_rownames(var = "features.plot")
D <- na.omit(D)
D <- D %>% column_to_rownames(var = "features.plot") %>% as.matrix()

GOI <- read.csv("biomarker-rna-heatmap.csv") %>% 
  #filter(sc_heatmap_annotation == TRUE) %>% 
  pull(Gene)

ha = rowAnnotation(foo = anno_mark(at = c(which(rownames(D) %in% GOI)), 
                                   labels = rownames(D)[which(rownames(D) %in% GOI)] %>% as.character(),
                                   link_gp = gpar(colour = "black", lwd= 1),
                                   labels_gp = gpar(colour = "black", fontsize = 12)))

col_ha <- HeatmapAnnotation("Cell class" = cells_df$Cell_type, col = list("Cell class" = cell_type))

set.seed(2)
ht1 <- Heatmap(D, 
        col = colorRampPalette(c("#336194", "grey5","#97c9a2"))(100),
        show_row_names = FALSE, row_title = NULL, 
        column_title = "DE genes",
        column_title_gp = gpar(fontface = "bold"),
        name = "Scaled\nRelative\nExpression", 
        column_names_gp = gpar(fontface = "bold", fontsize = 12),
        column_km = 5, column_gap = unit(1, "mm"),
        row_km = 5, row_gap = unit(1, "mm"),
        border_gp = gpar(col = "black", lwd = 2),
        #heatmap_legend_param = list(title_gp = gpar(fontface = "bold"),
        #                            legend_height = unit(4,"cm"),
        #                            legend_width = unit(2,"cm")#,
                                    #legend_gp = gpar(col = "black", lwd = 2.5)
        #                            ),
        #row_dend_gp = gpar(lwd = 2), column_dend_gp = gpar(lwd = 2),
        #clustering_distance_rows = "euclidean", clustering_distance_columns = "euclidean", 
        #clustering_method_rows = "complete", clustering_method_columns = "complete",
        top_annotation = col_ha,
        right_annotation = ha
        )

set.seed(2)
png("Figure_Plots/Sig_25pct_Expr_Cluster_Cells-name.png", width=20, height=60, units = "cm", res = 400)
ht = draw(ht1, merge_legend = TRUE, heatmap_legend_side = "right")
invisible(dev.off())
set.seed(2)
pdf("Figure_Plots/Sig_25pct_Expr_Cluster_Cells-name.pdf", width=12, height=40)
ht = draw(ht1, merge_legend = TRUE, heatmap_legend_side = "right")
invisible(dev.off())
```

## DE_Results

```{r}
DE_Results <- list()
for (i in levels(Merged_Temp$Manual_Anno2)) {
  message(i)
  Sub <- subset(Merged_Temp, subset = Manual_Anno == paste0(i))
  Temp <- FindMarkers(Sub, group.by = "Disease", ident.1 = "CLAD", ident.2 = "NDC", min.pct = 0.25, logfc.threshold = 0.5, recorrect_umi = FALSE)
  Temp <- Temp %>% dplyr::filter(p_val_adj < 1e-5) %>% dplyr::mutate(Cell = paste0(i)) %>% rownames_to_column("Gene") %>%
    dplyr::mutate(Signature = ifelse(Gene %in% unique(DE_Features$Gene),"Signature","Not_Signature"))
  DE_Results[[paste0(i)]] <- Temp
}

DE_Results <- do.call("rbind", DE_Results)

saveRDS(DE_Results, "./Downstream_Analysis/DE_Results_WRS.rds")

DE_Results <- list()
for (i in levels(Merged_Temp$Manual_Anno2)) {
  message(i)
  Sub <- subset(Merged_Temp, subset = Manual_Anno == paste0(i))
  Sub <- AggregateExpression(Sub, assays = "SCT", return.seurat = T, group.by = c("Disease","Sample_Align"))
  Temp <- FindMarkers(Sub, group.by = "Disease", ident.1 = "CLAD", ident.2 = "NDC", test.use = "DESeq2")
  Temp <- Temp %>% dplyr::filter(p_val_adj < 0.05) %>% dplyr::mutate(Cell = paste0(i)) %>% rownames_to_column("Gene") %>%
    dplyr::mutate(Signature = ifelse(Gene %in% unique(DE_Features$Gene),"Signature","Not_Signature"))
  DE_Results[[paste0(i)]] <- Temp
}

DE_Results <- do.call("rbind", DE_Results)

saveRDS(DE_Results, "./Downstream_Analysis/DE_Results_Aggregate_DESEQ2.rds")
```


```{r}
Cells <- c("aCap","gCap","Arterial","Venous","Systemic Venous","Lymphatics",
                                       "Pericytes","SMCs","Adv. Fibr.", "Alv. Fibr.",
                                       "ATII","ATI","Basal Cells","Club Cells","Ciliated Cells",
                                       "Mast Cells","B Cells","Plasma Cells",
                                       "NK Cells","CD8+ T-cells","CD4+ T-cells",
                                       "cMono","ncMono","DCs","Int-Mφ","Mo-AMφ","Alv-Mφ")

Cells_Colors <- c("#5276c9","#1c5ed9","#768bb5","#6692d9","#2674c7","grey75",
                  "#6dee6f","#23a16d","#78e9ab","#3cb049",
                  "#508991","#00d8e1","#6f85c7","#00bfe8","#5aeece",
                  "#637eb3","#aecf8e","#4f6e31",
                  "#d26c67","#7f241b","#b1634d",
                  "#6f64c7","#bd81f5","#d9b4f0","#8678f0","#8953c2","#a456a0")

cells_df <- data.frame("Cells" = Cells, 
                       "Cells_Colours" = Cells_Colors,
                       "Cell_type" = c(rep("Arterial & Lymphatics", 6), rep("Stromal", 4), rep("Epithelial", 5),
                                       rep("Myeloid", 1), rep("B & T cells", 5), rep("Myeloid", 6)) )
rownames(cells_df) <- cells_df$Cells

cell_type <- c("Arterial & Lymphatics" = "#1c5ed9", "B & T cells" = "#637eb3", "Myeloid" = "#bd81f5", "Stromal" = "#23a16d", "Epithelial" = "#5aeece")
```


### Violins

### Ridge Plots

